{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example: Polysistice Kidney Disease use case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39\\\\python39.zip', 'C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39\\\\DLLs', 'C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39\\\\lib', 'C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv', '', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages\\\\win32', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages\\\\win32\\\\lib', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages\\\\Pythonwin', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\src']\n"
     ]
    }
   ],
   "source": [
    "new_path = r\"C:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\examples\\usecases\"\n",
    "\n",
    "# Setting up the working directory\n",
    "import os\n",
    "import sys\n",
    "\n",
    "src_path = os.path.abspath(os.path.join(\"..\", \"..\", \"..\", \"src\"))\n",
    "if src_path not in sys.path:\n",
    "    sys.path.append(src_path)\n",
    "\n",
    "print(sys.path)  # Verify the correct src path is included"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules\n",
    "import pickle\n",
    "\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "from pyBiodatafuse import human_homologs, id_mapper, testcodeopenai, utils\n",
    "from pyBiodatafuse.annotators import (\n",
    "    bgee,\n",
    "    disgenet,\n",
    "    intact,\n",
    "    kegg,\n",
    "    minerva,\n",
    "    molmedb,\n",
    "    opentargets,\n",
    "    pubchem,\n",
    "    stringdb,\n",
    "    wikipathways,\n",
    ")\n",
    "from pyBiodatafuse.constants import (\n",
    "    BGEE_GENE_EXPRESSION_LEVELS_COL,\n",
    "    DISGENET_DISEASE_COL,\n",
    "    ENSEMBL_HOMOLOG_COL,\n",
    "    KEGG_COL,\n",
    "    MINERVA,\n",
    "    MOLMEDB_PROTEIN_COMPOUND_COL,\n",
    "    OPENTARGETS_DISEASE_COMPOUND_COL,\n",
    "    OPENTARGETS_GENE_COMPOUND_COL,\n",
    "    OPENTARGETS_GO_COL,\n",
    "    OPENTARGETS_REACTOME_COL,\n",
    "    PUBCHEM_COMPOUND_ASSAYS_COL,\n",
    "    STRING_PPI_COL,\n",
    "    WIKIPATHWAYS,\n",
    ")\n",
    "from pyBiodatafuse.graph import generator\n",
    "from pyBiodatafuse.utils import combine_sources, combine_with_homologs, create_harmonized_input_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type 'exit' to quit.\n",
      "\n",
      "ChatGPT: Test received! How can I assist you today?\n",
      "\n"
     ]
    }
   ],
   "source": [
    "testcodeopenai.main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Entity resolution using BridgeDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1. Load the input list and convert it to a dataframe\n",
    "\n",
    "Small set of mouse data used for debugging."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEST Mice use case =\n",
    "genes_of_interest = \"\"\"ENSMUSG00000026295\n",
    "ENSMUSG00000022877\n",
    "ENSMUSG00000020914\n",
    "ENSMUSG00000024747\n",
    "ENSMUSG00000032081\n",
    "ENSMUSG00000004035\n",
    "ENSMUSG00000072949\n",
    "ENSMUSG00000028970\n",
    "ENSMUSG00000028937\n",
    "ENSMUSG00000075044\"\"\"\n",
    "\n",
    "# TEST Mice Ensembl\n",
    "genes_of_interest = \"\"\"ENSMUSG00000067274\n",
    "ENSMUSG00000000001\n",
    "ENSMUSG00000030619\n",
    "ENSMUSG00000027490\n",
    "ENSMUSG00000022472\n",
    "ENSMUSG00000059552\"\"\"\n",
    "\n",
    "# genes_of_interest = \"\"\"ENSG00000141510\"\"\"\n",
    "\n",
    "# TEST Rat Ensembl\n",
    "# genes_of_interest = \"\"\"ENSRNOG00060027926\n",
    "# ENSRNOG00055005387\n",
    "# ENSRNOG00060018596\n",
    "# ENSRNOG00060011358\n",
    "# ENSRNOG00055009275\n",
    "# \"\"\"\n",
    "\n",
    "# TEST Human HGNC\n",
    "# genes_of_interest = \"\"\"CHRNG\n",
    "# DMD\n",
    "# AHR\n",
    "# SCN4A\n",
    "# LC25A1\n",
    "# HTR3A\"\"\"\n",
    "\n",
    "# TEST Human Ensembl\n",
    "# genes_of_interest = \"\"\"ENSG00000072080\n",
    "# ENSG00000113905\n",
    "# ENSG00000131747\n",
    "# ENSG00000165092\n",
    "# ENSG00000110245\n",
    "# ENSG00000213366\n",
    "# ENSG00000184227\n",
    "# ENSG00000085563\n",
    "# ENSG00000097021\n",
    "# ENSG00000149742\"\"\"\n",
    "\n",
    "gene_list = genes_of_interest.split(\"\\n\")\n",
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mouse use case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from pyBiodatafuse import data_loader, id_mapper\n",
    "\n",
    "# data_input = data_loader.create_df_from_dea(\"data/full_de_genes_treated_vs_untreated_plus_cpm_fc.csv\")\n",
    "# data_filtered = data_input[data_input['DE'].isin([1, -1])]\n",
    "# features_filtered = data_filtered['identifier']\n",
    "\n",
    "# gene_list = features_filtered.tolist()\n",
    "# print(len(gene_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000030619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000027490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000022472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000059552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier\n",
       "0  ENSMUSG00000067274\n",
       "1  ENSMUSG00000000001\n",
       "2  ENSMUSG00000030619\n",
       "3  ENSMUSG00000027490\n",
       "4  ENSMUSG00000022472\n",
       "5  ENSMUSG00000059552"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_input = pd.DataFrame(gene_list, columns=[\"identifier\"])\n",
    "data_input.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUSG00000067274', 'ENSMUSG00000000001', 'ENSMUSG00000030619', 'ENSMUSG00000027490', 'ENSMUSG00000022472', 'ENSMUSG00000059552']\n"
     ]
    }
   ],
   "source": [
    "print(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2. Query BridgeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>10524718</td>\n",
       "      <td>Affy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0071353</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0070180</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0022626</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:1904628</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0022625</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0042277</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0098794</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q5M8R8</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>101213_at</td>\n",
       "      <td>Affy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0005840</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0007595</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q18247798</td>\n",
       "      <td>Wikidata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0036464</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>4916236</td>\n",
       "      <td>Affy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0002181</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>11837</td>\n",
       "      <td>WikiGenes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:1990904</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>S4R1N1</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>4551704</td>\n",
       "      <td>Affy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>4456156</td>\n",
       "      <td>Affy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>GO:0003735</td>\n",
       "      <td>Gene Ontology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>5390295</td>\n",
       "      <td>Affy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            identifier identifier.source              target   target.source\n",
       "0   ENSMUSG00000067274           Ensembl  ENSMUSG00000067274         Ensembl\n",
       "1   ENSMUSG00000067274           Ensembl            10524718            Affy\n",
       "2   ENSMUSG00000067274           Ensembl          GO:0016020   Gene Ontology\n",
       "3   ENSMUSG00000067274           Ensembl          GO:0071353   Gene Ontology\n",
       "4   ENSMUSG00000067274           Ensembl          GO:0070180   Gene Ontology\n",
       "5   ENSMUSG00000067274           Ensembl          GO:0022626   Gene Ontology\n",
       "6   ENSMUSG00000067274           Ensembl          GO:1904628   Gene Ontology\n",
       "7   ENSMUSG00000067274           Ensembl          GO:0022625   Gene Ontology\n",
       "8   ENSMUSG00000067274           Ensembl          GO:0042277   Gene Ontology\n",
       "9   ENSMUSG00000067274           Ensembl          GO:0098794   Gene Ontology\n",
       "10  ENSMUSG00000067274           Ensembl              Q5M8R8  Uniprot-TrEMBL\n",
       "11  ENSMUSG00000067274           Ensembl           101213_at            Affy\n",
       "12  ENSMUSG00000067274           Ensembl          GO:0005840   Gene Ontology\n",
       "13  ENSMUSG00000067274           Ensembl          GO:0007595   Gene Ontology\n",
       "14  ENSMUSG00000067274           Ensembl           Q18247798        Wikidata\n",
       "15  ENSMUSG00000067274           Ensembl          GO:0036464   Gene Ontology\n",
       "16  ENSMUSG00000067274           Ensembl             4916236            Affy\n",
       "17  ENSMUSG00000067274           Ensembl          GO:0002181   Gene Ontology\n",
       "18  ENSMUSG00000067274           Ensembl               11837       WikiGenes\n",
       "19  ENSMUSG00000067274           Ensembl          GO:1990904   Gene Ontology\n",
       "20  ENSMUSG00000067274           Ensembl              S4R1N1  Uniprot-TrEMBL\n",
       "21  ENSMUSG00000067274           Ensembl             4551704            Affy\n",
       "22  ENSMUSG00000067274           Ensembl             4456156            Affy\n",
       "23  ENSMUSG00000067274           Ensembl          GO:0003735   Gene Ontology\n",
       "24  ENSMUSG00000067274           Ensembl             5390295            Affy"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mouse usecase\n",
    "input_species = \"Mouse\"\n",
    "\n",
    "bridgedb_df, bridgedb_metadata = id_mapper.bridgedb_xref(\n",
    "    identifiers=data_input,\n",
    "    input_species=input_species,\n",
    "    input_datasource=\"Ensembl\",\n",
    "    output_datasource=\"All\",\n",
    ")\n",
    "\n",
    "# TEST Human data\n",
    "# bridgedb_df, bridgedb_metadata = id_mapper.bridgedb_xref(\n",
    "#     identifiers=data_input,\n",
    "#     input_species=\"Human\",\n",
    "#     input_datasource=\"Ensembl\",\n",
    "#     output_datasource=\"All\",\n",
    "# )\n",
    "\n",
    "bridgedb_df.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Homologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>Ensembl_homologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000177103'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000165092'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000213366'}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000100426           Ensembl  ENSMUSG00000100426       Ensembl   \n",
       "1  ENSMUSG00000032087           Ensembl  ENSMUSG00000032087       Ensembl   \n",
       "2  ENSMUSG00000024747           Ensembl  ENSMUSG00000024747       Ensembl   \n",
       "3  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813       Ensembl   \n",
       "4  ENSMUSG00000004035           Ensembl  ENSMUSG00000004035       Ensembl   \n",
       "\n",
       "                   Ensembl_homologs  \n",
       "0                [{'homolog': nan}]  \n",
       "1  [{'homolog': 'ENSG00000177103'}]  \n",
       "2  [{'homolog': 'ENSG00000165092'}]  \n",
       "3                [{'homolog': nan}]  \n",
       "4  [{'homolog': 'ENSG00000213366'}]  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_homologs_df, ensembl_metadata = human_homologs.get_homologs(bridgedb_df=bridgedb_df)\n",
    "ensembl_homologs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>Ensembl_homologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000177103'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000165092'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000213366'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000160282'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000146039'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000128944'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000134690'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000159164'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000044816</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000044816</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000188674'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000003038</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000003038</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000015451</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000015451</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000224389'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000026295</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000026295</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000072080'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000056749</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000056749</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000165030'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000016756</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000016756</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000027452</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027452</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000154930'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000027577</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027577</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000101204'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000029368</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000029368</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000163631'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000026205</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000026205</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000213901'}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            identifier identifier.source              target target.source  \\\n",
       "0   ENSMUSG00000100426           Ensembl  ENSMUSG00000100426       Ensembl   \n",
       "1   ENSMUSG00000032087           Ensembl  ENSMUSG00000032087       Ensembl   \n",
       "2   ENSMUSG00000024747           Ensembl  ENSMUSG00000024747       Ensembl   \n",
       "3   ENSMUSG00000091813           Ensembl  ENSMUSG00000091813       Ensembl   \n",
       "4   ENSMUSG00000004035           Ensembl  ENSMUSG00000004035       Ensembl   \n",
       "5   ENSMUSG00000001155           Ensembl  ENSMUSG00000001155       Ensembl   \n",
       "6   ENSMUSG00000021336           Ensembl  ENSMUSG00000021336       Ensembl   \n",
       "7   ENSMUSG00000027331           Ensembl  ENSMUSG00000027331       Ensembl   \n",
       "8   ENSMUSG00000028873           Ensembl  ENSMUSG00000028873       Ensembl   \n",
       "9   ENSMUSG00000038486           Ensembl  ENSMUSG00000038486       Ensembl   \n",
       "10  ENSMUSG00000044816           Ensembl  ENSMUSG00000044816       Ensembl   \n",
       "11  ENSMUSG00000003038           Ensembl  ENSMUSG00000003038       Ensembl   \n",
       "12  ENSMUSG00000015451           Ensembl  ENSMUSG00000015451       Ensembl   \n",
       "13  ENSMUSG00000026295           Ensembl  ENSMUSG00000026295       Ensembl   \n",
       "14  ENSMUSG00000056749           Ensembl  ENSMUSG00000056749       Ensembl   \n",
       "15  ENSMUSG00000016756           Ensembl  ENSMUSG00000016756       Ensembl   \n",
       "16  ENSMUSG00000027452           Ensembl  ENSMUSG00000027452       Ensembl   \n",
       "17  ENSMUSG00000027577           Ensembl  ENSMUSG00000027577       Ensembl   \n",
       "18  ENSMUSG00000029368           Ensembl  ENSMUSG00000029368       Ensembl   \n",
       "19  ENSMUSG00000026205           Ensembl  ENSMUSG00000026205       Ensembl   \n",
       "\n",
       "                    Ensembl_homologs  \n",
       "0                 [{'homolog': nan}]  \n",
       "1   [{'homolog': 'ENSG00000177103'}]  \n",
       "2   [{'homolog': 'ENSG00000165092'}]  \n",
       "3                 [{'homolog': nan}]  \n",
       "4   [{'homolog': 'ENSG00000213366'}]  \n",
       "5   [{'homolog': 'ENSG00000160282'}]  \n",
       "6   [{'homolog': 'ENSG00000146039'}]  \n",
       "7   [{'homolog': 'ENSG00000128944'}]  \n",
       "8   [{'homolog': 'ENSG00000134690'}]  \n",
       "9   [{'homolog': 'ENSG00000159164'}]  \n",
       "10  [{'homolog': 'ENSG00000188674'}]  \n",
       "11                [{'homolog': nan}]  \n",
       "12  [{'homolog': 'ENSG00000224389'}]  \n",
       "13  [{'homolog': 'ENSG00000072080'}]  \n",
       "14  [{'homolog': 'ENSG00000165030'}]  \n",
       "15                [{'homolog': nan}]  \n",
       "16  [{'homolog': 'ENSG00000154930'}]  \n",
       "17  [{'homolog': 'ENSG00000101204'}]  \n",
       "18  [{'homolog': 'ENSG00000163631'}]  \n",
       "19  [{'homolog': 'ENSG00000213901'}]  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_homologs_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000177103', 'ENSG00000165092', 'ENSG00000213366', 'ENSG00000160282', 'ENSG00000146039', 'ENSG00000128944', 'ENSG00000134690', 'ENSG00000159164', 'ENSG00000188674', 'ENSG00000224389', 'ENSG00000072080', 'ENSG00000165030', 'ENSG00000154930', 'ENSG00000101204', 'ENSG00000163631', 'ENSG00000213901', 'ENSG00000169679', 'ENSG00000213886', 'ENSG00000273604', 'ENSG00000182272', 'ENSG00000144035', 'ENSG00000138778', 'ENSG00000149742', 'ENSG00000131747', 'ENSG00000160097', 'ENSG00000117399', 'ENSG00000100321', 'ENSG00000073111', 'ENSG00000108830', 'ENSG00000127220', 'ENSG00000109107', 'ENSG00000144395', 'ENSG00000151287', 'ENSG00000162897', 'ENSG00000087076', 'ENSG00000100526', 'ENSG00000165507', 'ENSG00000159228', 'ENSG00000120054', 'ENSG00000149554', 'ENSG00000184545', 'ENSG00000170430', 'ENSG00000123427', 'ENSG00000120800', 'ENSG00000113905', 'ENSG00000169245', 'ENSG00000072571', 'ENSG00000101412', 'ENSG00000136689', 'ENSG00000102837', 'ENSG00000112299', 'ENSG00000168268', 'ENSG00000097021', 'ENSG00000144035', 'ENSG00000181778', 'ENSG00000161800', 'ENSG00000110921', 'ENSG00000170312', 'ENSG00000184428', 'ENSG00000137804', 'ENSG00000106927', 'ENSG00000085563', 'ENSG00000157456', 'ENSG00000146918', 'ENSG00000178828', 'ENSG00000184227', 'ENSG00000111665', 'ENSG00000110245', 'ENSG00000141682', 'ENSG00000173237', 'ENSG00000166483']\n"
     ]
    }
   ],
   "source": [
    "homologs = (\n",
    "    ensembl_homologs_df[ENSEMBL_HOMOLOG_COL]\n",
    "    .apply(\n",
    "        lambda x: (\n",
    "            x[0][\"homolog\"] if isinstance(x, list) and len(x) > 0 and \"homolog\" in x[0] else None\n",
    "        )\n",
    "    )\n",
    "    .dropna()\n",
    "    .tolist()\n",
    ")\n",
    "\n",
    "print(homologs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Query homologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:3782384</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:2150309</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>114873</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>A0A1L1SQZ7</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E9QPR7</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q4VA61</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>A0A1L1SQ53</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>26358</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>O35945</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>B2RTL5</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:1347050</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>F6Z9B9</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:3648740</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>436059</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>68312</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YVP5</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YVP6</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YZ29</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E9QAC8</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q80W21</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:1915562</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            identifier identifier.source              target   target.source\n",
       "0   ENSMUSG00000100426           Ensembl         MGI:3782384             MGI\n",
       "1   ENSMUSG00000100426           Ensembl  ENSMUSG00000100426         Ensembl\n",
       "2   ENSMUSG00000032087           Ensembl         MGI:2150309             MGI\n",
       "3   ENSMUSG00000032087           Ensembl              114873       NCBI Gene\n",
       "4   ENSMUSG00000032087           Ensembl          A0A1L1SQZ7  Uniprot-TrEMBL\n",
       "5   ENSMUSG00000032087           Ensembl              E9QPR7  Uniprot-TrEMBL\n",
       "6   ENSMUSG00000032087           Ensembl              Q4VA61  Uniprot-TrEMBL\n",
       "7   ENSMUSG00000032087           Ensembl          A0A1L1SQ53  Uniprot-TrEMBL\n",
       "8   ENSMUSG00000032087           Ensembl  ENSMUSG00000032087         Ensembl\n",
       "9   ENSMUSG00000024747           Ensembl  ENSMUSG00000024747         Ensembl\n",
       "10  ENSMUSG00000024747           Ensembl               26358       NCBI Gene\n",
       "11  ENSMUSG00000024747           Ensembl              O35945  Uniprot-TrEMBL\n",
       "12  ENSMUSG00000024747           Ensembl              B2RTL5  Uniprot-TrEMBL\n",
       "13  ENSMUSG00000024747           Ensembl         MGI:1347050             MGI\n",
       "14  ENSMUSG00000091813           Ensembl              F6Z9B9  Uniprot-TrEMBL\n",
       "15  ENSMUSG00000091813           Ensembl         MGI:3648740             MGI\n",
       "16  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813         Ensembl\n",
       "17  ENSMUSG00000091813           Ensembl              436059       NCBI Gene\n",
       "18  ENSMUSG00000004035           Ensembl               68312       NCBI Gene\n",
       "19  ENSMUSG00000004035           Ensembl              D3YVP5  Uniprot-TrEMBL\n",
       "20  ENSMUSG00000004035           Ensembl              D3YVP6  Uniprot-TrEMBL\n",
       "21  ENSMUSG00000004035           Ensembl              D3YZ29  Uniprot-TrEMBL\n",
       "22  ENSMUSG00000004035           Ensembl              E9QAC8  Uniprot-TrEMBL\n",
       "23  ENSMUSG00000004035           Ensembl              Q80W21  Uniprot-TrEMBL\n",
       "24  ENSMUSG00000004035           Ensembl         MGI:1915562             MGI"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_species = \"Human\"\n",
    "\n",
    "data_input_hl = pd.DataFrame(homologs, columns=[\"identifier\"])\n",
    "\n",
    "bridgedb_df_hl, bridgedb_metadata_hl = id_mapper.bridgedb_xref(\n",
    "    identifiers=data_input_hl,\n",
    "    input_species=\"Human\",\n",
    "    input_datasource=\"Ensembl\",\n",
    "    output_datasource=\"All\",\n",
    ")\n",
    "\n",
    "bridgedb_df.head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Step-by-step graph generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1. Gene-Disease edges\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15eadd18-9b50-466d-a41d-3deb5fad122c\n"
     ]
    }
   ],
   "source": [
    "load_dotenv(\"disgenet.env\")\n",
    "\n",
    "disgenet_api_key = os.getenv(\"DISGENET_API_KEY\")\n",
    "print(disgenet_api_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying DisGeNET: 100%|██████████| 71/71 [01:09<00:00,  1.02it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_98032\\374358601.py:1: UserWarning: The intermediate_df in DISGENET annotatur should be checked, please create an issue on https://github.com/BioDataFuse/pyBiodatafuse/issues/.\n",
      "  disgenet_df, disgenet_metadata = disgenet.get_gene_disease(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>DISGENET_diseases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>6694</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Retinal dystrophy', 'HPO': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>3161</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Cancer, Breast', 'HPO': Non...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>4171</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Liver cell carcinoma', 'HPO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>5243</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': nan, 'HPO': nan, 'NCI': nan,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>51171</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Severe myopia (&gt; -6.00 diop...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source target target.source  \\\n",
       "0  ENSG00000072080           Ensembl   6694     NCBI Gene   \n",
       "1  ENSG00000072571           Ensembl   3161     NCBI Gene   \n",
       "2  ENSG00000073111           Ensembl   4171     NCBI Gene   \n",
       "3  ENSG00000085563           Ensembl   5243     NCBI Gene   \n",
       "4  ENSG00000087076           Ensembl  51171     NCBI Gene   \n",
       "\n",
       "                                   DISGENET_diseases  \n",
       "0  [{'disease_name': 'Retinal dystrophy', 'HPO': ...  \n",
       "1  [{'disease_name': 'Cancer, Breast', 'HPO': Non...  \n",
       "2  [{'disease_name': 'Liver cell carcinoma', 'HPO...  \n",
       "3  [{'disease_name': nan, 'HPO': nan, 'NCI': nan,...  \n",
       "4  [{'disease_name': 'Severe myopia (> -6.00 diop...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disgenet_df, disgenet_metadata = disgenet.get_gene_disease(\n",
    "    api_key=disgenet_api_key, bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "disgenet_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'disease_name': 'Retinal dystrophy',\n",
       "  'HPO': 'HPO:HP:0000556',\n",
       "  'NCI': 'NCI:C35625',\n",
       "  'OMIM': None,\n",
       "  'MONDO': 'MONDO:0019118',\n",
       "  'ORDO': 'ORDO:71862',\n",
       "  'EFO': None,\n",
       "  'DO': 'DOID:8501',\n",
       "  'MESH': 'MESH:D058499',\n",
       "  'UMLS': 'UMLS:C0854723',\n",
       "  'disease_type': 'disease',\n",
       "  'disease_umlscui': 'C0854723',\n",
       "  'score': 0.4,\n",
       "  'ei': nan,\n",
       "  'el': None}]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disgenet_df[DISGENET_DISEASE_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Disease-Compound edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UMLS_C0006142</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000305</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UMLS_C0006142</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0003869</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UMLS_C1458155</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0003869</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UMLS_C2239176</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000182</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UMLS_C2239176</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000762</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      identifier identifier.source       target target.source\n",
       "0  UMLS_C0006142              UMLS  EFO_0000305           EFO\n",
       "1  UMLS_C0006142              UMLS  EFO_0003869           EFO\n",
       "2  UMLS_C1458155              UMLS  EFO_0003869           EFO\n",
       "3  UMLS_C2239176              UMLS  EFO_0000182           EFO\n",
       "4  UMLS_C2239176              UMLS  EFO_0000762           EFO"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Prepare the input to use DISGENET output as seed for OpenTargets\n",
    "disease_mapping_df = create_harmonized_input_file(disgenet_df, DISGENET_DISEASE_COL, \"EFO\", \"UMLS\")\n",
    "disease_mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing diseases-drug interactions: 100%|██████████| 292/292 [00:07<00:00, 37.11it/s]\n",
      "Mapping PubChem: 100%|██████████| 2129/2129 [06:13<00:00,  5.70it/s]\n",
      "c:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\annotators\\opentargets.py:1133: UserWarning: Not all values in column 'adverse_effect' have the correct type '<class 'list'>'.\n",
      "  check_columns_against_constants(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_disease_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UMLS_C0000786</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_1001255</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1276308', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UMLS_C0002171</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0004192</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1200963', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UMLS_C0002940</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0004264</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1076903', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UMLS_C0002940</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0009659</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL526', 'drugbank_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UMLS_C0002994</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0005532</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL2028850', 'drugba...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      identifier identifier.source       target target.source  \\\n",
       "0  UMLS_C0000786              UMLS  EFO_1001255           EFO   \n",
       "1  UMLS_C0002171              UMLS  EFO_0004192           EFO   \n",
       "2  UMLS_C0002940              UMLS  EFO_0004264           EFO   \n",
       "3  UMLS_C0002940              UMLS  EFO_0009659           EFO   \n",
       "4  UMLS_C0002994              UMLS  EFO_0005532           EFO   \n",
       "\n",
       "                       OpenTargets_disease_compounds  \n",
       "0  [{'chembl_id': 'CHEMBL:CHEMBL1276308', 'drugba...  \n",
       "1  [{'chembl_id': 'CHEMBL:CHEMBL1200963', 'drugba...  \n",
       "2  [{'chembl_id': 'CHEMBL:CHEMBL1076903', 'drugba...  \n",
       "3  [{'chembl_id': 'CHEMBL:CHEMBL526', 'drugbank_i...  \n",
       "4  [{'chembl_id': 'CHEMBL:CHEMBL2028850', 'drugba...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    opentargets_disease_compound_df,\n",
    "    opentargets_disease_compound_metadata,\n",
    ") = opentargets.get_disease_compound_interactions(disease_mapping_df)\n",
    "opentargets_disease_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'chembl_id': 'CHEMBL:CHEMBL1276308',\n",
       "  'drugbank_id': 'DrugBank:DB00834',\n",
       "  'compound_cid': 'pubchem.compound:55245',\n",
       "  'compound_name': 'MIFEPRISTONE',\n",
       "  'clincal_trial_phase': 4.0,\n",
       "  'is_approved': True,\n",
       "  'relation': 'treats',\n",
       "  'adverse_effect_count': 41.0,\n",
       "  'adverse_effect': [{'name': 'abortion incomplete'},\n",
       "   {'name': 'haemorrhage'},\n",
       "   {'name': 'pregnancy'},\n",
       "   {'name': 'endometritis'},\n",
       "   {'name': 'induced abortion failed'},\n",
       "   {'name': 'vaginal haemorrhage'},\n",
       "   {'name': 'anaemia'},\n",
       "   {'name': 'muscle spasms'},\n",
       "   {'name': 'metrorrhagia'},\n",
       "   {'name': 'abortion induced incomplete'},\n",
       "   {'name': 'menorrhagia'},\n",
       "   {'name': 'pain'},\n",
       "   {'name': 'uterine haemorrhage'},\n",
       "   {'name': 'post abortion infection'},\n",
       "   {'name': 'uterine rupture'},\n",
       "   {'name': 'ectopic pregnancy'},\n",
       "   {'name': 'blood potassium decreased'},\n",
       "   {'name': 'syncope'},\n",
       "   {'name': 'endometritis bacterial'},\n",
       "   {'name': 'pelvic inflammatory disease'},\n",
       "   {'name': 'uterine dilation and curettage'},\n",
       "   {'name': 'haemorrhagic anaemia'},\n",
       "   {'name': 'retained products of conception'},\n",
       "   {'name': 'endometrial thickening'},\n",
       "   {'name': 'ruptured ectopic pregnancy'}]},\n",
       " {'chembl_id': 'CHEMBL:CHEMBL606',\n",
       "  'drugbank_id': 'DrugBank:DB00929',\n",
       "  'compound_cid': 'pubchem.compound:5282381',\n",
       "  'compound_name': 'MISOPROSTOL',\n",
       "  'clincal_trial_phase': 4.0,\n",
       "  'is_approved': True,\n",
       "  'relation': 'treats',\n",
       "  'adverse_effect_count': 33.0,\n",
       "  'adverse_effect': [{'name': 'abortion incomplete'},\n",
       "   {'name': 'haemorrhage'},\n",
       "   {'name': 'pregnancy'},\n",
       "   {'name': 'endometritis'},\n",
       "   {'name': 'menorrhagia'},\n",
       "   {'name': 'anaemia'},\n",
       "   {'name': 'muscle spasms'},\n",
       "   {'name': 'induced abortion failed'},\n",
       "   {'name': 'international normalised ratio fluctuation'},\n",
       "   {'name': 'drug hypersensitivity'},\n",
       "   {'name': 'pain'},\n",
       "   {'name': 'rheumatoid arthritis'},\n",
       "   {'name': 'bloody discharge'},\n",
       "   {'name': 'uterine rupture'},\n",
       "   {'name': 'ectopic pregnancy'},\n",
       "   {'name': 'infection'},\n",
       "   {'name': 'pelvic inflammatory disease'},\n",
       "   {'name': 'vaginal haemorrhage'},\n",
       "   {'name': 'metrorrhagia'},\n",
       "   {'name': 'abdominal pain'},\n",
       "   {'name': 'retained products of conception'},\n",
       "   {'name': 'foetal distress syndrome'},\n",
       "   {'name': 'syncope'},\n",
       "   {'name': 'post abortion infection'},\n",
       "   {'name': 'abortion induced incomplete'}]},\n",
       " {'chembl_id': 'CHEMBL:CHEMBL940',\n",
       "  'drugbank_id': 'DrugBank:DB00996',\n",
       "  'compound_cid': 'pubchem.compound:3446',\n",
       "  'compound_name': 'GABAPENTIN',\n",
       "  'clincal_trial_phase': 4.0,\n",
       "  'is_approved': True,\n",
       "  'relation': 'treats',\n",
       "  'adverse_effect_count': 10.0,\n",
       "  'adverse_effect': [{'name': 'drug hypersensitivity'},\n",
       "   {'name': 'somnolence'},\n",
       "   {'name': 'myoclonus'},\n",
       "   {'name': 'drug abuse'},\n",
       "   {'name': 'confusional state'},\n",
       "   {'name': 'sedation'},\n",
       "   {'name': 'dizziness'},\n",
       "   {'name': 'drug titration error'},\n",
       "   {'name': 'toxicity to various agents'},\n",
       "   {'name': 'suicidal ideation'}]}]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_disease_compound_df[OPENTARGETS_DISEASE_COMPOUND_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Compound Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compounds from OpenTargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene-drug interactions: 100%|██████████| 71/71 [00:00<00:00, 147.70it/s]\n",
      "Mapping PubChem: 100%|██████████| 37/37 [00:13<00:00,  2.68it/s]\n",
      "c:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\annotators\\opentargets.py:607: UserWarning: Not all values in column 'adverse_effect_count' have the correct type '<class 'int'>'.\n",
      "  check_columns_against_constants(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_gene_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1086218', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source           target target.source  \\\n",
       "0  ENSG00000072080           Ensembl  ENSG00000072080       Ensembl   \n",
       "1  ENSG00000072571           Ensembl  ENSG00000072571       Ensembl   \n",
       "2  ENSG00000073111           Ensembl  ENSG00000073111       Ensembl   \n",
       "3  ENSG00000085563           Ensembl  ENSG00000085563       Ensembl   \n",
       "4  ENSG00000087076           Ensembl  ENSG00000087076       Ensembl   \n",
       "\n",
       "                          OpenTargets_gene_compounds  \n",
       "0  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  \n",
       "1  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  \n",
       "2  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  \n",
       "3  [{'chembl_id': 'CHEMBL:CHEMBL1086218', 'drugba...  \n",
       "4  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_compound_df, opentargets_compound_metadata = opentargets.get_gene_compound_interactions(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "opentargets_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'chembl_id': nan,\n",
       "  'drugbank_id': nan,\n",
       "  'compound_cid': nan,\n",
       "  'compound_name': nan,\n",
       "  'clincal_trial_phase': nan,\n",
       "  'is_approved': nan,\n",
       "  'relation': nan,\n",
       "  'adverse_effect_count': nan,\n",
       "  'adverse_effect': nan}]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_compound_df[OPENTARGETS_GENE_COMPOUND_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Compounds from PubChem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying PubChem: 100%|██████████| 11/11 [00:02<00:00,  4.57it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_98032\\484186372.py:1: UserWarning: There is no annotation for your input list in PubChem.\n",
      "  pubchem_assay_df, pubchem_assay_metadata = pubchem.get_protein_compound_screened(\n"
     ]
    }
   ],
   "source": [
    "pubchem_assay_df, pubchem_assay_metadata = pubchem.get_protein_compound_screened(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Gene-Pathways edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pathways from WikiPathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying WikiPathways: 100%|██████████| 4/4 [01:42<00:00, 25.58s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>WikiPathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000934</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>72960</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>14317</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'WP:WP435', 'pathway_label': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000001313</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>11858</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000001334</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>384061</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000002870</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>17216</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'WP:WP413', 'pathway_label': '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source  target target.source  \\\n",
       "0  ENSMUSG00000000934           Ensembl   72960     NCBI Gene   \n",
       "1  ENSMUSG00000001155           Ensembl   14317     NCBI Gene   \n",
       "2  ENSMUSG00000001313           Ensembl   11858     NCBI Gene   \n",
       "3  ENSMUSG00000001334           Ensembl  384061     NCBI Gene   \n",
       "4  ENSMUSG00000002870           Ensembl   17216     NCBI Gene   \n",
       "\n",
       "                                        WikiPathways  \n",
       "0  [{'pathway_id': nan, 'pathway_label': nan, 'pa...  \n",
       "1  [{'pathway_id': 'WP:WP435', 'pathway_label': '...  \n",
       "2  [{'pathway_id': nan, 'pathway_label': nan, 'pa...  \n",
       "3  [{'pathway_id': nan, 'pathway_label': nan, 'pa...  \n",
       "4  [{'pathway_id': 'WP:WP413', 'pathway_label': '...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipathways_df, wikipathways_metadata = wikipathways.get_gene_wikipathways(bridgedb_df=bridgedb_df)\n",
    "wikipathways_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pathway_id': nan, 'pathway_label': nan, 'pathway_gene_count': nan}]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipathways_df[WIKIPATHWAYS][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pathways from KEGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C01089</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>6971058</td>\n",
       "      <td>PubChem Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C01089</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>92135</td>\n",
       "      <td>PubChem Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C01089</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C01089</td>\n",
       "      <td>KEGG Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00020</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C00020</td>\n",
       "      <td>KEGG Compound</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C00020</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>6083</td>\n",
       "      <td>PubChem Compound</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier identifier.source   target     target.source\n",
       "0     C01089     KEGG Compound  6971058  PubChem Compound\n",
       "1     C01089     KEGG Compound    92135  PubChem Compound\n",
       "2     C01089     KEGG Compound   C01089     KEGG Compound\n",
       "3     C00020     KEGG Compound   C00020     KEGG Compound\n",
       "4     C00020     KEGG Compound     6083  PubChem Compound"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compounds_of_interest = \"\"\"C01089\n",
    "C00020\n",
    "C02571\n",
    "C00212\n",
    "C00041\n",
    "C00152\n",
    "C00049\n",
    "C00719\n",
    "C00114\n",
    "C00158\n",
    "C00300\n",
    "C01026\n",
    "C00122\n",
    "C00031\n",
    "C00025\n",
    "C00064\n",
    "C00037\n",
    "C00135\n",
    "C00262\n",
    "C00130\n",
    "C00294\n",
    "C00407\n",
    "C00186\n",
    "C00123\n",
    "C00149\n",
    "C00073\n",
    "C00137\n",
    "C00003\n",
    "C00153\n",
    "C00079\n",
    "C00588\n",
    "C00346\n",
    "C04230\n",
    "C00245\n",
    "C00188\n",
    "C00082\n",
    "C00043\n",
    "C00105\n",
    "C00106\n",
    "C00299\n",
    "C00183\"\"\"\n",
    "\n",
    "metabolite_list = compounds_of_interest.split(\"\\n\")\n",
    "data_input = pd.DataFrame(metabolite_list, columns=[\"identifier\"])\n",
    "\n",
    "bridgdb_df_cmp, bridgdb_metadata = id_mapper.bridgedb_xref(\n",
    "    identifiers=data_input,\n",
    "    input_species=\"Mouse\",\n",
    "    input_datasource=\"KEGG Compound\",\n",
    "    output_datasource=\"All\",\n",
    ")\n",
    "bridgdb_df_cmp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mmu:114873\n",
      "\n",
      "\n",
      "mmu:26358\n",
      "mmu:26358\tpath:mmu00830\n",
      "mmu:26358\tpath:mmu01100\n",
      "\n",
      "mmu:436059\n",
      "mmu:436059\tpath:mmu00983\n",
      "\n",
      "mmu:68312\n",
      "mmu:68312\tpath:mmu00480\n",
      "mmu:68312\tpath:mmu00980\n",
      "mmu:68312\tpath:mmu00982\n",
      "mmu:68312\tpath:mmu00983\n",
      "mmu:68312\tpath:mmu01100\n",
      "mmu:68312\tpath:mmu01524\n",
      "mmu:68312\tpath:mmu05200\n",
      "mmu:68312\tpath:mmu05204\n",
      "mmu:68312\tpath:mmu05207\n",
      "mmu:68312\tpath:mmu05208\n",
      "mmu:68312\tpath:mmu05225\n",
      "mmu:68312\tpath:mmu05418\n",
      "\n",
      "mmu:14317\n",
      "mmu:14317\tpath:mmu00340\n",
      "mmu:14317\tpath:mmu00670\n",
      "mmu:14317\tpath:mmu01100\n",
      "\n",
      "mmu:319848\n",
      "\n",
      "\n",
      "mmu:51944\n",
      "\n",
      "\n",
      "mmu:52276\n",
      "\n",
      "\n",
      "mmu:64051\n",
      "mmu:64051\tpath:mmu04512\n",
      "\n",
      "mmu:98303\n",
      "\n",
      "\n",
      "mmu:15331\n",
      "\n",
      "\n",
      "mmu:625018\n",
      "mmu:625018\tpath:mmu04610\n",
      "mmu:625018\tpath:mmu04936\n",
      "mmu:625018\tpath:mmu05133\n",
      "mmu:625018\tpath:mmu05150\n",
      "mmu:625018\tpath:mmu05171\n",
      "mmu:625018\tpath:mmu05322\n",
      "\n",
      "mmu:75396\n",
      "\n",
      "\n",
      "mmu:18030\n",
      "mmu:18030\tpath:mmu04710\n",
      "\n",
      "mmu:12763\n",
      "mmu:12763\tpath:mmu00520\n",
      "mmu:12763\tpath:mmu01250\n",
      "\n",
      "mmu:68738\n",
      "mmu:68738\tpath:mmu00010\n",
      "mmu:68738\tpath:mmu00620\n",
      "mmu:68738\tpath:mmu00630\n",
      "mmu:68738\tpath:mmu00640\n",
      "mmu:68738\tpath:mmu01100\n",
      "mmu:68738\tpath:mmu01200\n",
      "\n",
      "mmu:11438\n",
      "mmu:11438\tpath:mmu04080\n",
      "mmu:11438\tpath:mmu04725\n",
      "mmu:11438\tpath:mmu05033\n",
      "mmu:11438\tpath:mmu05207\n",
      "\n",
      "mmu:11657\n",
      "mmu:11657\tpath:mmu04918\n",
      "\n",
      "mmu:22626\n",
      "\n",
      "\n",
      "mmu:12235\n",
      "mmu:12235\tpath:mmu04110\n",
      "mmu:12235\tpath:mmu04114\n",
      "mmu:12235\tpath:mmu04914\n",
      "\n",
      "mmu:13119\n",
      "mmu:13119\tpath:mmu00071\n",
      "mmu:13119\tpath:mmu00590\n",
      "mmu:13119\tpath:mmu00830\n",
      "mmu:13119\tpath:mmu01100\n",
      "mmu:13119\tpath:mmu03320\n",
      "mmu:13119\tpath:mmu04270\n",
      "mmu:13119\tpath:mmu04750\n",
      "\n",
      "mmu:24108\n",
      "\n",
      "\n",
      "mmu:103551\n",
      "mmu:103551\tpath:mmu03083\n",
      "\n",
      "mmu:330671\n",
      "mmu:330671\tpath:mmu00513\n",
      "mmu:330671\tpath:mmu01100\n",
      "\n",
      "mmu:69049\n",
      "mmu:69049\tpath:mmu00480\n",
      "mmu:69049\tpath:mmu01100\n",
      "\n",
      "mmu:229841\n",
      "mmu:229841\tpath:mmu04814\n",
      "\n",
      "mmu:236293\n",
      "\n",
      "\n",
      "mmu:21973\n",
      "mmu:21973\tpath:mmu01524\n",
      "\n",
      "mmu:384061\n",
      "\n",
      "\n",
      "mmu:107995\n",
      "mmu:107995\tpath:mmu04110\n",
      "mmu:107995\tpath:mmu04114\n",
      "mmu:107995\tpath:mmu04120\n",
      "mmu:107995\tpath:mmu05166\n",
      "mmu:107995\tpath:mmu05203\n",
      "\n",
      "mmu:20972\n",
      "\n",
      "\n",
      "mmu:27384\n",
      "\n",
      "\n",
      "mmu:17216\n",
      "mmu:17216\tpath:mmu03030\n",
      "mmu:17216\tpath:mmu04110\n",
      "\n",
      "mmu:13897\n",
      "mmu:13897\tpath:mmu00983\n",
      "\n",
      "mmu:11858\n",
      "\n",
      "\n",
      "mmu:64296\n",
      "\n",
      "\n",
      "mmu:11676\n",
      "mmu:11676\tpath:mmu00010\n",
      "mmu:11676\tpath:mmu00030\n",
      "mmu:11676\tpath:mmu00051\n",
      "mmu:11676\tpath:mmu01100\n",
      "mmu:11676\tpath:mmu01200\n",
      "mmu:11676\tpath:mmu01230\n",
      "mmu:11676\tpath:mmu04066\n",
      "\n",
      "mmu:78016\n",
      "\n",
      "\n",
      "mmu:75623\n",
      "\n",
      "\n",
      "mmu:64435\n",
      "\n",
      "\n",
      "mmu:66065\n",
      "\n",
      "\n",
      "mmu:72391\n",
      "\n",
      "\n",
      "mmu:213393\n",
      "mmu:213393\tpath:mmu04068\n",
      "\n",
      "mmu:12408\n",
      "mmu:12408\tpath:mmu00590\n",
      "mmu:12408\tpath:mmu00790\n",
      "mmu:12408\tpath:mmu00980\n",
      "mmu:12408\tpath:mmu01100\n",
      "mmu:12408\tpath:mmu05204\n",
      "mmu:12408\tpath:mmu05208\n",
      "\n",
      "mmu:93721\n",
      "\n",
      "\n",
      "mmu:12649\n",
      "mmu:12649\tpath:mmu04110\n",
      "mmu:12649\tpath:mmu04115\n",
      "mmu:12649\tpath:mmu04218\n",
      "mmu:12649\tpath:mmu05166\n",
      "mmu:12649\tpath:mmu05170\n",
      "mmu:12649\tpath:mmu05203\n",
      "\n",
      "mmu:18218\n",
      "mmu:18218\tpath:mmu04010\n",
      "mmu:18218\tpath:mmu04148\n",
      "\n",
      "mmu:17314\n",
      "\n",
      "\n",
      "mmu:100504608\n",
      "\n",
      "\n",
      "mmu:70683\n",
      "\n",
      "\n",
      "mmu:94175\n",
      "\n",
      "\n",
      "mmu:15945\n",
      "mmu:15945\tpath:mmu04060\n",
      "mmu:15945\tpath:mmu04061\n",
      "mmu:15945\tpath:mmu04062\n",
      "mmu:15945\tpath:mmu04620\n",
      "mmu:15945\tpath:mmu04622\n",
      "mmu:15945\tpath:mmu04623\n",
      "mmu:15945\tpath:mmu04657\n",
      "mmu:15945\tpath:mmu04668\n",
      "mmu:15945\tpath:mmu05160\n",
      "mmu:15945\tpath:mmu05164\n",
      "mmu:15945\tpath:mmu05169\n",
      "mmu:15945\tpath:mmu05171\n",
      "\n",
      "mmu:15366\n",
      "mmu:15366\tpath:mmu04512\n",
      "\n",
      "mmu:13555\n",
      "mmu:13555\tpath:mmu01522\n",
      "mmu:13555\tpath:mmu04110\n",
      "mmu:13555\tpath:mmu04137\n",
      "mmu:13555\tpath:mmu04218\n",
      "mmu:13555\tpath:mmu04934\n",
      "mmu:13555\tpath:mmu05160\n",
      "mmu:13555\tpath:mmu05161\n",
      "mmu:13555\tpath:mmu05163\n",
      "mmu:13555\tpath:mmu05165\n",
      "mmu:13555\tpath:mmu05166\n",
      "mmu:13555\tpath:mmu05167\n",
      "mmu:13555\tpath:mmu05169\n",
      "mmu:13555\tpath:mmu05200\n",
      "mmu:13555\tpath:mmu05206\n",
      "mmu:13555\tpath:mmu05207\n",
      "mmu:13555\tpath:mmu05212\n",
      "mmu:13555\tpath:mmu05214\n",
      "mmu:13555\tpath:mmu05215\n",
      "mmu:13555\tpath:mmu05218\n",
      "mmu:13555\tpath:mmu05219\n",
      "mmu:13555\tpath:mmu05220\n",
      "mmu:13555\tpath:mmu05222\n",
      "mmu:13555\tpath:mmu05223\n",
      "mmu:13555\tpath:mmu05224\n",
      "mmu:13555\tpath:mmu05225\n",
      "mmu:13555\tpath:mmu05226\n",
      "\n",
      "mmu:69894\n",
      "\n",
      "\n",
      "mmu:16181\n",
      "mmu:16181\tpath:mmu04060\n",
      "\n",
      "mmu:380924\n",
      "\n",
      "\n",
      "mmu:22361\n",
      "mmu:22361\tpath:mmu00770\n",
      "mmu:22361\tpath:mmu01100\n",
      "\n",
      "mmu:70021\n",
      "\n",
      "\n",
      "mmu:70025\n",
      "mmu:70025\tpath:mmu00062\n",
      "mmu:70025\tpath:mmu01040\n",
      "mmu:70025\tpath:mmu01100\n",
      "\n",
      "mmu:93674\n",
      "mmu:93674\tpath:mmu00480\n",
      "mmu:93674\tpath:mmu01100\n",
      "\n",
      "mmu:226040\n",
      "\n",
      "\n",
      "mmu:26934\n",
      "\n",
      "\n",
      "mmu:17855\n",
      "mmu:17855\tpath:mmu00900\n",
      "mmu:17855\tpath:mmu01100\n",
      "mmu:17855\tpath:mmu04146\n",
      "\n",
      "mmu:12534\n",
      "mmu:12534\tpath:mmu04110\n",
      "mmu:12534\tpath:mmu04114\n",
      "mmu:12534\tpath:mmu04115\n",
      "mmu:12534\tpath:mmu04218\n",
      "mmu:12534\tpath:mmu04540\n",
      "mmu:12534\tpath:mmu04914\n",
      "mmu:12534\tpath:mmu05170\n",
      "mmu:12534\tpath:mmu05203\n",
      "\n",
      "mmu:72960\n",
      "\n",
      "\n",
      "mmu:108907\n",
      "\n",
      "\n",
      "mmu:11699\n",
      "\n",
      "\n",
      "mmu:18669\n",
      "mmu:18669\tpath:mmu02010\n",
      "mmu:18669\tpath:mmu04976\n",
      "mmu:18669\tpath:mmu05206\n",
      "mmu:18669\tpath:mmu05226\n",
      "\n",
      "mmu:12442\n",
      "mmu:12442\tpath:mmu04068\n",
      "mmu:12442\tpath:mmu04110\n",
      "mmu:12442\tpath:mmu04114\n",
      "mmu:12442\tpath:mmu04115\n",
      "mmu:12442\tpath:mmu04218\n",
      "mmu:12442\tpath:mmu04914\n",
      "mmu:12442\tpath:mmu05166\n",
      "mmu:12442\tpath:mmu05170\n",
      "\n",
      "mmu:76044\n",
      "\n",
      "\n",
      "mmu:66825\n",
      "\n",
      "\n",
      "mmu:26897\n",
      "mmu:26897\tpath:mmu00062\n",
      "mmu:26897\tpath:mmu01040\n",
      "mmu:26897\tpath:mmu01100\n",
      "mmu:26897\tpath:mmu04913\n",
      "\n",
      "mmu:14793\n",
      "\n",
      "\n",
      "mmu:11814\n",
      "mmu:11814\tpath:mmu03320\n",
      "mmu:11814\tpath:mmu04979\n",
      "\n",
      "mmu:58801\n",
      "mmu:58801\tpath:mmu01524\n",
      "mmu:58801\tpath:mmu04115\n",
      "mmu:58801\tpath:mmu04210\n",
      "mmu:58801\tpath:mmu04215\n",
      "mmu:58801\tpath:mmu05200\n",
      "mmu:58801\tpath:mmu05203\n",
      "mmu:58801\tpath:mmu05210\n",
      "\n",
      "mmu:69861\n",
      "\n",
      "\n",
      "mmu:22390\n",
      "mmu:22390\tpath:mmu04110\n",
      "mmu:22390\tpath:mmu05170\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>114873</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>26358</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00830', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>436059</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00983', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>68312</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00480', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>14317</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00340', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source  target target.source  \\\n",
       "0  ENSMUSG00000032087           Ensembl  114873     NCBI Gene   \n",
       "1  ENSMUSG00000024747           Ensembl   26358     NCBI Gene   \n",
       "2  ENSMUSG00000091813           Ensembl  436059     NCBI Gene   \n",
       "3  ENSMUSG00000004035           Ensembl   68312     NCBI Gene   \n",
       "4  ENSMUSG00000001155           Ensembl   14317     NCBI Gene   \n",
       "\n",
       "                                       KEGG_pathways  \n",
       "0  [{'pathway_id': nan, 'pathway_label': nan, 'ge...  \n",
       "1  [{'pathway_id': 'path:mmu00830', 'pathway_labe...  \n",
       "2  [{'pathway_id': 'path:mmu00983', 'pathway_labe...  \n",
       "3  [{'pathway_id': 'path:mmu00480', 'pathway_labe...  \n",
       "4  [{'pathway_id': 'path:mmu00340', 'pathway_labe...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df, kegg_metadata = kegg.get_pathways(bridgedb_df)\n",
    "kegg_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pathway_id': 'path:mmu00983',\n",
       "  'pathway_label': 'Drug metabolism - other enzymes - Mus musculus (house mouse)',\n",
       "  'gene_count': 94,\n",
       "  'compounds': [{'KEGG_identifier': 'C00003'},\n",
       "   {'KEGG_identifier': 'C00006'},\n",
       "   {'KEGG_identifier': 'C02320'},\n",
       "   {'KEGG_identifier': 'C02380'},\n",
       "   {'KEGG_identifier': 'C04242'},\n",
       "   {'KEGG_identifier': 'C04646'},\n",
       "   {'KEGG_identifier': 'C05361'},\n",
       "   {'KEGG_identifier': 'C06108'},\n",
       "   {'KEGG_identifier': 'C06837'},\n",
       "   {'KEGG_identifier': 'C07054'},\n",
       "   {'KEGG_identifier': 'C07446'},\n",
       "   {'KEGG_identifier': 'C07447'},\n",
       "   {'KEGG_identifier': 'C07585'},\n",
       "   {'KEGG_identifier': 'C07648'},\n",
       "   {'KEGG_identifier': 'C07649'},\n",
       "   {'KEGG_identifier': 'C11173'},\n",
       "   {'KEGG_identifier': 'C11376'},\n",
       "   {'KEGG_identifier': 'C11736'},\n",
       "   {'KEGG_identifier': 'C12650'},\n",
       "   {'KEGG_identifier': 'C12673'},\n",
       "   {'KEGG_identifier': 'C12739'},\n",
       "   {'KEGG_identifier': 'C13252'},\n",
       "   {'KEGG_identifier': 'C16542'},\n",
       "   {'KEGG_identifier': 'C16543'},\n",
       "   {'KEGG_identifier': 'C16613'},\n",
       "   {'KEGG_identifier': 'C16614'},\n",
       "   {'KEGG_identifier': 'C16615'},\n",
       "   {'KEGG_identifier': 'C16616'},\n",
       "   {'KEGG_identifier': 'C16617'},\n",
       "   {'KEGG_identifier': 'C16618'},\n",
       "   {'KEGG_identifier': 'C16619'},\n",
       "   {'KEGG_identifier': 'C16620'},\n",
       "   {'KEGG_identifier': 'C16621'},\n",
       "   {'KEGG_identifier': 'C16622'},\n",
       "   {'KEGG_identifier': 'C16623'},\n",
       "   {'KEGG_identifier': 'C16624'},\n",
       "   {'KEGG_identifier': 'C16625'},\n",
       "   {'KEGG_identifier': 'C16630'},\n",
       "   {'KEGG_identifier': 'C16631'},\n",
       "   {'KEGG_identifier': 'C16632'},\n",
       "   {'KEGG_identifier': 'C16633'},\n",
       "   {'KEGG_identifier': 'C16634'},\n",
       "   {'KEGG_identifier': 'C16635'},\n",
       "   {'KEGG_identifier': 'C16641'},\n",
       "   {'KEGG_identifier': 'C21748'},\n",
       "   {'KEGG_identifier': 'C21749'},\n",
       "   {'KEGG_identifier': 'C21750'},\n",
       "   {'KEGG_identifier': 'C21751'},\n",
       "   {'KEGG_identifier': 'C21754'},\n",
       "   {'KEGG_identifier': 'C21758'},\n",
       "   {'KEGG_identifier': 'C21759'},\n",
       "   {'KEGG_identifier': 'C21760'}]}]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df[\"KEGG_pathways\"][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>KEGG_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C01089</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C01089</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>{'KEGG_identifier': 'C01089', 'name': '(R)-3-H...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C00020</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C00020</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>{'KEGG_identifier': 'C00020', 'name': 'AMP'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C02571</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C02571</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>{'KEGG_identifier': 'C02571', 'name': 'O-Acety...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00212</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C00212</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>{'KEGG_identifier': 'C00212', 'name': 'Adenosi...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C00041</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>C00041</td>\n",
       "      <td>KEGG Compound</td>\n",
       "      <td>{'KEGG_identifier': 'C00041', 'name': 'L-Alani...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  identifier identifier.source  target  target.source  \\\n",
       "0     C01089     KEGG Compound  C01089  KEGG Compound   \n",
       "1     C00020     KEGG Compound  C00020  KEGG Compound   \n",
       "2     C02571     KEGG Compound  C02571  KEGG Compound   \n",
       "3     C00212     KEGG Compound  C00212  KEGG Compound   \n",
       "4     C00041     KEGG Compound  C00041  KEGG Compound   \n",
       "\n",
       "                                      KEGG_compounds  \n",
       "0  {'KEGG_identifier': 'C01089', 'name': '(R)-3-H...  \n",
       "1       {'KEGG_identifier': 'C00020', 'name': 'AMP'}  \n",
       "2  {'KEGG_identifier': 'C02571', 'name': 'O-Acety...  \n",
       "3  {'KEGG_identifier': 'C00212', 'name': 'Adenosi...  \n",
       "4  {'KEGG_identifier': 'C00041', 'name': 'L-Alani...  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_compound_df = kegg.get_compounds(bridgdb_df_cmp)\n",
    "kegg_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>114873</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'ge...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>26358</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00830', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>436059</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00983', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>68312</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00480', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>14317</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00340', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source  target target.source  \\\n",
       "0  ENSMUSG00000032087           Ensembl  114873     NCBI Gene   \n",
       "1  ENSMUSG00000024747           Ensembl   26358     NCBI Gene   \n",
       "2  ENSMUSG00000091813           Ensembl  436059     NCBI Gene   \n",
       "3  ENSMUSG00000004035           Ensembl   68312     NCBI Gene   \n",
       "4  ENSMUSG00000001155           Ensembl   14317     NCBI Gene   \n",
       "\n",
       "                                       KEGG_pathways  \n",
       "0  [{'pathway_id': nan, 'pathway_label': nan, 'ge...  \n",
       "1  [{'pathway_id': 'path:mmu00830', 'pathway_labe...  \n",
       "2  [{'pathway_id': 'path:mmu00983', 'pathway_labe...  \n",
       "3  [{'pathway_id': 'path:mmu00480', 'pathway_labe...  \n",
       "4  [{'pathway_id': 'path:mmu00340', 'pathway_labe...  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'KEGG_identifier': 'C01089', 'name': '(R)-3-Hydroxybutanoate'}"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_compound_df[\"KEGG_compounds\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data_input_compounds = pd.DataFrame(kegg_identifiers, columns=[\"identifier\"])\n",
    "# data_input_compounds.head()\n",
    "\n",
    "# bridgdb_df, bridgdb_metadata = id_mapper.bridgedb_xref(\n",
    "#     identifiers=data_input_compounds,\n",
    "#     input_species=\"Mouse\",\n",
    "#     input_datasource=\"KEGG Compound\",\n",
    "#     output_datasource=\"PubChem Compound\",\n",
    "# )\n",
    "# bridgdb_df.head(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['5742\\tChemspider', 'DB00171\\tDrugBank', 'HMDB00538\\tHMDB', '5957\\tPubChem-compound', 'CHEBI:15422\\tChEBI', '15422\\tChEBI', 'ZKHQWZAMYRWXGA-KQYNXXCUSA-N\\tInChIKey', 'DTXSID6022559\\tEPA CompTox', 'C00001491\\tKNApSAcK', '56-65-5\\tCAS', 'HMDB0000538\\tHMDB', 'CHEMBL14249\\tChEMBL compound', 'Q80863\\tWikidata', 'C00002\\tKEGG Compound', '1713\\tGuide to Pharmacology', 'D08646\\tKEGG Drug']\n",
      "No metabolite names found.\n"
     ]
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "chebi_id = \"15422\"\n",
    "\n",
    "url = f\"http://webservice.bridgedb.org/Human/xrefs/ChEBI/15422\"\n",
    "response = requests.get(url)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    # Parse the response to extract metabolite name(s)\n",
    "    lines = response.text.splitlines()\n",
    "    print(lines)\n",
    "    names = [line.split(\"\\t\")[2] for line in lines if len(line.split(\"\\t\")) > 2]\n",
    "    print(names if names else \"No metabolite names found.\")\n",
    "else:\n",
    "    print(f\"Error: Unable to retrieve data (status code {response.status_code}).\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reactome pathways from OpenTargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene-pathway interactions: 100%|██████████| 71/71 [00:00<00:00, 405.41it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_98032\\442422480.py:1: UserWarning: The intermediate_df in OpenTargets_reactome annotatur should be checked, please create an issue on https://github.com/BioDataFuse/pyBiodatafuse/issues/.\n",
      "  opentargets_reactome_df, opentargets_reactome_metadata = opentargets.get_gene_reactome_pathways(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_reactome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Platelet degranulation ', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Hyaluronan uptake and degr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Orc1 removal from chromati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Abacavir transmembrane tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Estrogen biosynthesis', 'p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source           target target.source  \\\n",
       "0  ENSG00000072080           Ensembl  ENSG00000072080       Ensembl   \n",
       "1  ENSG00000072571           Ensembl  ENSG00000072571       Ensembl   \n",
       "2  ENSG00000073111           Ensembl  ENSG00000073111       Ensembl   \n",
       "3  ENSG00000085563           Ensembl  ENSG00000085563       Ensembl   \n",
       "4  ENSG00000087076           Ensembl  ENSG00000087076       Ensembl   \n",
       "\n",
       "                                OpenTargets_reactome  \n",
       "0  [{'pathway_label': 'Platelet degranulation ', ...  \n",
       "1  [{'pathway_label': 'Hyaluronan uptake and degr...  \n",
       "2  [{'pathway_label': 'Orc1 removal from chromati...  \n",
       "3  [{'pathway_label': 'Abacavir transmembrane tra...  \n",
       "4  [{'pathway_label': 'Estrogen biosynthesis', 'p...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_reactome_df, opentargets_reactome_metadata = opentargets.get_gene_reactome_pathways(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "opentargets_reactome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pathway_label': 'Platelet degranulation ',\n",
       "  'pathway_id': 'Reactome:R-HSA-114608'},\n",
       " {'pathway_label': 'Regulation of Insulin-like Growth Factor (IGF) transport and uptake by Insulin-like Growth Factor Binding Proteins (IGFBPs)',\n",
       "  'pathway_id': 'Reactome:R-HSA-381426'},\n",
       " {'pathway_label': 'Post-translational protein phosphorylation',\n",
       "  'pathway_id': 'Reactome:R-HSA-8957275'}]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_reactome_df[OPENTARGETS_REACTOME_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Gene Ontology from OpenTargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene annotation: 100%|██████████| 71/71 [00:00<00:00, 325.19it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_98032\\517848960.py:1: UserWarning: The intermediate_df in OpenTargets_go annotatur should be checked, please create an issue on https://github.com/BioDataFuse/pyBiodatafuse/issues/.\n",
      "  opentargets_go_df, opentargets_go_metadata = opentargets.get_gene_go_process(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005788', 'go_name': 'endoplasm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005515', 'go_name': 'protein b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005654', 'go_name': 'nucleopla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0008559', 'go_name': 'ABC-type ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005515', 'go_name': 'protein b...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source           target target.source  \\\n",
       "0  ENSG00000072080           Ensembl  ENSG00000072080       Ensembl   \n",
       "1  ENSG00000072571           Ensembl  ENSG00000072571       Ensembl   \n",
       "2  ENSG00000073111           Ensembl  ENSG00000073111       Ensembl   \n",
       "3  ENSG00000085563           Ensembl  ENSG00000085563       Ensembl   \n",
       "4  ENSG00000087076           Ensembl  ENSG00000087076       Ensembl   \n",
       "\n",
       "                                      OpenTargets_go  \n",
       "0  [{'go_id': 'GO:0005788', 'go_name': 'endoplasm...  \n",
       "1  [{'go_id': 'GO:0005515', 'go_name': 'protein b...  \n",
       "2  [{'go_id': 'GO:0005654', 'go_name': 'nucleopla...  \n",
       "3  [{'go_id': 'GO:0008559', 'go_name': 'ABC-type ...  \n",
       "4  [{'go_id': 'GO:0005515', 'go_name': 'protein b...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_go_df, opentargets_go_metadata = opentargets.get_gene_go_process(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "opentargets_go_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'go_id': 'GO:0005788',\n",
       "  'go_name': 'endoplasmic reticulum lumen',\n",
       "  'go_type': 'C'},\n",
       " {'go_id': 'GO:0031089',\n",
       "  'go_name': 'platelet dense granule lumen',\n",
       "  'go_type': 'C'},\n",
       " {'go_id': 'GO:0004866',\n",
       "  'go_name': 'endopeptidase inhibitor activity',\n",
       "  'go_type': 'F'},\n",
       " {'go_id': 'GO:0046849', 'go_name': 'bone remodeling', 'go_type': 'P'},\n",
       " {'go_id': 'GO:0062023',\n",
       "  'go_name': 'collagen-containing extracellular matrix',\n",
       "  'go_type': 'C'},\n",
       " {'go_id': 'GO:0005576', 'go_name': 'extracellular region', 'go_type': 'C'},\n",
       " {'go_id': 'GO:0001501',\n",
       "  'go_name': 'skeletal system development',\n",
       "  'go_type': 'P'}]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_go_df[OPENTARGETS_GO_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.6. Protein-Protein Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_40024\\3298456429.py:2: UserWarning: There is no interaction between your input list based on StringDB, {'source_version': '12.0'}.\n",
      "  ppi_df, ppi_metadata = stringdb.get_ppi(bridgedb_df=bridgedb_df, species=input_species)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_species = \"Mouse\"\n",
    "ppi_df, ppi_metadata = stringdb.get_ppi(bridgedb_df=bridgedb_df, species=input_species)\n",
    "ppi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_df[STRING_PPI_COL][3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Intact WIP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Error getting IntAct version\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ENSMUSG00000067274': ['Q5M8R8', 'S4R1N1', 'P14869', 'D3YVM5'], 'ENSMUSG00000000001': ['Q9DC51'], 'ENSMUSG00000030619': ['Q921E6', 'A0A140LIG5', 'A0A140LIN6', 'A0A5F8MPX8'], 'ENSMUSG00000027490': ['Q9CYB4', 'Q547J6', 'Q61501'], 'ENSMUSG00000022472': ['F8WJJ8', 'Q9CQT7', 'Q3TCG9', 'D6RDE8', 'E9Q2Y9']}\n",
      "ENSMUSG00000030619 {'ENSMUSG00000030619', 'ENSMUSG00000027490', 'ENSMUSG00000000001', 'ENSMUSG00000067274', 'ENSMUSG00000022472'}\n",
      "ENSMUSG00000027490 {'ENSMUSG00000030619', 'ENSMUSG00000027490', 'ENSMUSG00000000001', 'ENSMUSG00000067274', 'ENSMUSG00000022472'}\n",
      "ENSMUSG00000000001 {'ENSMUSG00000030619', 'ENSMUSG00000027490', 'ENSMUSG00000000001', 'ENSMUSG00000067274', 'ENSMUSG00000022472'}\n",
      "ENSMUSG00000067274 {'ENSMUSG00000030619', 'ENSMUSG00000027490', 'ENSMUSG00000000001', 'ENSMUSG00000067274', 'ENSMUSG00000022472'}\n",
      "ENSMUSG00000022472 {'ENSMUSG00000030619', 'ENSMUSG00000027490', 'ENSMUSG00000000001', 'ENSMUSG00000067274', 'ENSMUSG00000022472'}\n",
      "{'ENSMUSG00000030619': ['EBI-904301'], 'ENSMUSG00000027490': ['EBI-1025536'], 'ENSMUSG00000000001': ['EBI-641852'], 'ENSMUSG00000067274': ['EBI-529949'], 'ENSMUSG00000022472': ['EBI-7768710']}\n",
      "{'EBI-529949', 'EBI-904301', 'EBI-7768710', 'EBI-641852', 'EBI-1025536'}\n",
      "interaction EBI-1379423 EBI-529949\n",
      "interaction EBI-1379423 EBI-529949\n",
      "interaction EBI-6909754 EBI-529949\n",
      "interaction EBI-7768710 EBI-529949\n",
      "interaction EBI-904301 EBI-529949\n",
      "interaction EBI-400825 EBI-529949\n",
      "interaction EBI-3862093 EBI-529949\n",
      "interaction EBI-716699 EBI-529949\n",
      "interaction EBI-298576 EBI-529949\n",
      "interaction EBI-4281130 EBI-529949\n",
      "interaction EBI-641852 EBI-22297545\n",
      "interaction EBI-1379423 EBI-641852\n",
      "interaction EBI-641852 EBI-10981450\n",
      "interaction EBI-1379423 EBI-641852\n",
      "interaction EBI-641852 EBI-7575403\n",
      "interaction EBI-641852 EBI-641864\n",
      "interaction EBI-641852 EBI-641864\n",
      "interaction EBI-641852 EBI-641864\n",
      "interaction EBI-641852 EBI-641864\n",
      "interaction EBI-641852 EBI-644645\n",
      "interaction EBI-912574 EBI-641852\n",
      "interaction EBI-15665134 EBI-904301\n",
      "interaction EBI-904301 EBI-15665134\n",
      "interaction EBI-15665134 EBI-904301\n",
      "interaction EBI-15825247 EBI-904301\n",
      "interaction EBI-904301 EBI-15665134\n",
      "interaction EBI-904301 EBI-6430187\n",
      "interaction EBI-15665134 EBI-904301\n",
      "interaction EBI-16024836 EBI-904301\n",
      "interaction EBI-16024836 EBI-904301\n",
      "interaction EBI-904301 EBI-2526529\n",
      "interaction EBI-904301 EBI-2270972\n",
      "interaction EBI-904301 EBI-2526893\n",
      "interaction EBI-6876582 EBI-904301\n",
      "interaction EBI-904301 EBI-2526952\n",
      "interaction EBI-904301 EBI-2526714\n",
      "interaction EBI-904301 EBI-2527165\n",
      "interaction EBI-904301 EBI-493592\n",
      "interaction EBI-904301 EBI-2526952\n",
      "interaction EBI-904301 EBI-1373622\n",
      "interaction EBI-904301 EBI-299641\n",
      "interaction EBI-7225787 EBI-1025536\n",
      "interaction EBI-1025536 EBI-8873147\n",
      "interaction EBI-5529102 EBI-1025536\n",
      "interaction EBI-1025536 EBI-5529102\n",
      "interaction EBI-1025536 EBI-5529102\n",
      "interaction EBI-1801080 EBI-1025536\n",
      "interaction EBI-1801080 EBI-1025536\n",
      "interaction EBI-1801080 EBI-1025536\n",
      "interaction EBI-5529102 EBI-1025536\n",
      "interaction EBI-1025536 EBI-1801080\n",
      "interaction EBI-1801080 EBI-1025536\n",
      "interaction EBI-5529102 EBI-1025536\n",
      "interaction EBI-1025536 EBI-1801080\n",
      "interaction EBI-645192 EBI-1025536\n",
      "interaction EBI-1025536 EBI-26673036\n",
      "interaction EBI-642213 EBI-1025536\n",
      "interaction EBI-491274 EBI-1025536\n",
      "interaction EBI-7768710 EBI-7768710\n",
      "interaction EBI-7768710 EBI-7768710\n",
      "interaction EBI-7768710 EBI-7768710\n",
      "interaction EBI-7770338 EBI-7768710\n",
      "interaction EBI-7770653 EBI-7768710\n",
      "interaction EBI-7770511 EBI-7768710\n",
      "interaction EBI-911407 EBI-7768710\n",
      "interaction EBI-7768710 EBI-7770727\n",
      "interaction EBI-645649 EBI-7768710\n",
      "interaction EBI-7769654 EBI-7768710\n",
      "interaction EBI-7769720 EBI-7768710\n",
      "interaction EBI-7769967 EBI-7768710\n",
      "interaction EBI-7770144 EBI-7768710\n",
      "interaction EBI-773122 EBI-7768710\n",
      "interaction EBI-7768710 EBI-7769891\n",
      "interaction EBI-7768710 EBI-7770423\n",
      "interaction EBI-6393149 EBI-7768710\n",
      "interaction EBI-7768710 EBI-7768990\n",
      "interaction EBI-7768710 EBI-7768990\n",
      "interaction EBI-772883 EBI-7768710\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>IntAct_interactions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'interaction_id': 'EBI-7770205', 'interactor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000030619</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000030619</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000027490</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027490</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000022472</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000022472</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'interaction_id': 'EBI-7768731', 'interactor...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000067274           Ensembl  ENSMUSG00000067274       Ensembl   \n",
       "1  ENSMUSG00000000001           Ensembl  ENSMUSG00000000001       Ensembl   \n",
       "2  ENSMUSG00000030619           Ensembl  ENSMUSG00000030619       Ensembl   \n",
       "3  ENSMUSG00000027490           Ensembl  ENSMUSG00000027490       Ensembl   \n",
       "4  ENSMUSG00000022472           Ensembl  ENSMUSG00000022472       Ensembl   \n",
       "\n",
       "                                 IntAct_interactions  \n",
       "0  [{'interaction_id': 'EBI-7770205', 'interactor...  \n",
       "1                                                 []  \n",
       "2                                                 []  \n",
       "3                                                 []  \n",
       "4  [{'interaction_id': 'EBI-7768731', 'interactor...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intact_df, intact_metadata = intact.get_interactions(bridgedb_df)\n",
    "intact_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'interaction_id': 'EBI-7770205',\n",
       "  'interactor_id_A': 'EBI-7768710',\n",
       "  'interactor_id_B': 'EBI-529949',\n",
       "  'binary_interaction_id': 12910167,\n",
       "  'confidence_values': ['intact-miscore:0.37'],\n",
       "  'intact_score': 0.37,\n",
       "  'biological_role_A': 'unspecified role',\n",
       "  'biological_role_B': 'unspecified role',\n",
       "  'type': 'physical association',\n",
       "  'stoichiometry_A': '0-0',\n",
       "  'stoichiometry_B': '0-0',\n",
       "  'detection_method': 'two hybrid pooling',\n",
       "  'detection_method_id': 'MI:0398',\n",
       "  'host_organism': 'Saccharomyces cerevisiae',\n",
       "  'interactor_A_name': 'desi1_mouse',\n",
       "  'interactor_B_name': 'rla0_mouse',\n",
       "  'interactor_A_species': 'Mus musculus',\n",
       "  'interactor_B_species': 'Mus musculus',\n",
       "  'molecule_A': 'Desi1',\n",
       "  'molecule_B': 'Rplp0',\n",
       "  'id_A': 'Q9CQT7',\n",
       "  'id_B': 'P14869',\n",
       "  'pubmed_publication_id': '22370726',\n",
       "  'ensembl': 'ENSMUSG00000067274'},\n",
       " {'interaction_id': 'EBI-2526357',\n",
       "  'interactor_id_A': 'EBI-904301',\n",
       "  'interactor_id_B': 'EBI-529949',\n",
       "  'binary_interaction_id': 7267158,\n",
       "  'confidence_values': ['intact-miscore:0.35'],\n",
       "  'intact_score': 0.35,\n",
       "  'biological_role_A': 'unspecified role',\n",
       "  'biological_role_B': 'unspecified role',\n",
       "  'type': 'association',\n",
       "  'stoichiometry_A': '0-0',\n",
       "  'stoichiometry_B': '0-0',\n",
       "  'detection_method': 'anti tag coip',\n",
       "  'detection_method_id': 'MI:0007',\n",
       "  'host_organism': 'Mus musculus embryonic stem cell',\n",
       "  'interactor_A_name': 'eed_mouse',\n",
       "  'interactor_B_name': 'rla0_mouse',\n",
       "  'interactor_A_species': 'Mus musculus',\n",
       "  'interactor_B_species': 'Mus musculus',\n",
       "  'molecule_A': 'Eed',\n",
       "  'molecule_B': 'Rplp0',\n",
       "  'id_A': 'Q921E6',\n",
       "  'id_B': 'P14869',\n",
       "  'pubmed_publication_id': '20064375',\n",
       "  'ensembl': 'ENSMUSG00000067274'}]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "intact_df[\"IntAct_interactions\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "intact_compound_df, intact_compound_metadata = intact.get_compound_interactions(bridgedb_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to query IntAct.\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "intact_data = intact.get_compound_related_interactions()\n",
    "print(intact_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.7 Gene expression edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[8], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m bgee_df, bgee_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mbgee\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_gene_expression\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbridgedb_df\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbridgedb_df\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      2\u001b[0m bgee_df\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\annotators\\bgee.py:145\u001b[0m, in \u001b[0;36mget_gene_expression\u001b[1;34m(bridgedb_df)\u001b[0m\n\u001b[0;32m    142\u001b[0m sparql_query_template_sub \u001b[38;5;241m=\u001b[39m sparql_query_template\u001b[38;5;241m.\u001b[39msubstitute(substit_dict)\n\u001b[0;32m    144\u001b[0m sparql\u001b[38;5;241m.\u001b[39msetQuery(sparql_query_template_sub)\n\u001b[1;32m--> 145\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43msparql\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mqueryAndConvert\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    147\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(res[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresults\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbindings\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m    149\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m col \u001b[38;5;129;01min\u001b[39;00m df:\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\SPARQLWrapper\\Wrapper.py:967\u001b[0m, in \u001b[0;36mSPARQLWrapper.queryAndConvert\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    962\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mqueryAndConvert\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mQueryResult.ConvertResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    963\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Macro like method: issue a query and return the converted results.\u001b[39;00m\n\u001b[0;32m    964\u001b[0m \n\u001b[0;32m    965\u001b[0m \u001b[38;5;124;03m    :return: the converted query result. See the conversion methods for more details.\u001b[39;00m\n\u001b[0;32m    966\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 967\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    968\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m res\u001b[38;5;241m.\u001b[39mconvert()\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\SPARQLWrapper\\Wrapper.py:960\u001b[0m, in \u001b[0;36mSPARQLWrapper.query\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    942\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mquery\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mQueryResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    943\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    944\u001b[0m \u001b[38;5;124;03m    Execute the query.\u001b[39;00m\n\u001b[0;32m    945\u001b[0m \u001b[38;5;124;03m    Exceptions can be raised if either the URI is wrong or the HTTP sends back an error (this is also the\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    958\u001b[0m \u001b[38;5;124;03m    :rtype: :class:`QueryResult` instance\u001b[39;00m\n\u001b[0;32m    959\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 960\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m QueryResult(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_query\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\SPARQLWrapper\\Wrapper.py:926\u001b[0m, in \u001b[0;36mSPARQLWrapper._query\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    924\u001b[0m         response \u001b[38;5;241m=\u001b[39m urlopener(request, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout)\n\u001b[0;32m    925\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 926\u001b[0m         response \u001b[38;5;241m=\u001b[39m \u001b[43murlopener\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    927\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturnFormat\n\u001b[0;32m    928\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m urllib\u001b[38;5;241m.\u001b[39merror\u001b[38;5;241m.\u001b[39mHTTPError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:214\u001b[0m, in \u001b[0;36murlopen\u001b[1;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[0;32m    212\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    213\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[1;32m--> 214\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mopener\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:517\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[1;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[0;32m    514\u001b[0m     req \u001b[38;5;241m=\u001b[39m meth(req)\n\u001b[0;32m    516\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124murllib.Request\u001b[39m\u001b[38;5;124m'\u001b[39m, req\u001b[38;5;241m.\u001b[39mfull_url, req\u001b[38;5;241m.\u001b[39mdata, req\u001b[38;5;241m.\u001b[39mheaders, req\u001b[38;5;241m.\u001b[39mget_method())\n\u001b[1;32m--> 517\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    519\u001b[0m \u001b[38;5;66;03m# post-process response\u001b[39;00m\n\u001b[0;32m    520\u001b[0m meth_name \u001b[38;5;241m=\u001b[39m protocol\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_response\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:534\u001b[0m, in \u001b[0;36mOpenerDirector._open\u001b[1;34m(self, req, data)\u001b[0m\n\u001b[0;32m    531\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[0;32m    533\u001b[0m protocol \u001b[38;5;241m=\u001b[39m req\u001b[38;5;241m.\u001b[39mtype\n\u001b[1;32m--> 534\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_open\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\n\u001b[0;32m    535\u001b[0m \u001b[43m                          \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_open\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    536\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[0;32m    537\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:494\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[1;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[0;32m    492\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[0;32m    493\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[1;32m--> 494\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    495\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    496\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:1389\u001b[0m, in \u001b[0;36mHTTPSHandler.https_open\u001b[1;34m(self, req)\u001b[0m\n\u001b[0;32m   1388\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttps_open\u001b[39m(\u001b[38;5;28mself\u001b[39m, req):\n\u001b[1;32m-> 1389\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhttp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mHTTPSConnection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1390\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_context\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_hostname\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:1350\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[1;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[0;32m   1348\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n\u001b[0;32m   1349\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m URLError(err)\n\u001b[1;32m-> 1350\u001b[0m     r \u001b[38;5;241m=\u001b[39m \u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1351\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[0;32m   1352\u001b[0m     h\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\http\\client.py:1377\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1375\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1376\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 1377\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1378\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[0;32m   1379\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\http\\client.py:320\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    318\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[0;32m    319\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m--> 320\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    321\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[0;32m    322\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\http\\client.py:281\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    280\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m--> 281\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    282\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[0;32m    283\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\socket.py:704\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m    702\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m    703\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 704\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    705\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[0;32m    706\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\ssl.py:1242\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[1;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[0;32m   1238\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m   1239\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m   1240\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[0;32m   1241\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[1;32m-> 1242\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1243\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1244\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\ssl.py:1100\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[1;34m(self, len, buffer)\u001b[0m\n\u001b[0;32m   1098\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1099\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m-> 1100\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1101\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1102\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "bgee_df, bgee_metadata = bgee.get_gene_expression(bridgedb_df=bridgedb_df)\n",
    "bgee_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'anatomical_entity_id': 'UBERON_0004535',\n",
       "  'anatomical_entity_name': 'cardiovascular system',\n",
       "  'expression_level': 50.40476,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001007',\n",
       "  'anatomical_entity_name': 'digestive system',\n",
       "  'expression_level': 74.9911,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0000948',\n",
       "  'anatomical_entity_name': 'heart',\n",
       "  'expression_level': 53.08205,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0002113',\n",
       "  'anatomical_entity_name': 'kidney',\n",
       "  'expression_level': 63.30739,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0002107',\n",
       "  'anatomical_entity_name': 'liver',\n",
       "  'expression_level': 96.7258,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001016',\n",
       "  'anatomical_entity_name': 'nervous system',\n",
       "  'expression_level': 38.55995,\n",
       "  'confidence_level_id': 'CIO_0000030',\n",
       "  'confidence_level_name': 'medium confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001264',\n",
       "  'anatomical_entity_name': 'pancreas',\n",
       "  'expression_level': 42.49478,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001987',\n",
       "  'anatomical_entity_name': 'placenta',\n",
       "  'expression_level': 42.2012,\n",
       "  'confidence_level_id': 'CIO_0000030',\n",
       "  'confidence_level_name': 'medium confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'}]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bgee_df[BGEE_GENE_EXPRESSION_LEVELS_COL][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.8 Transporter Inhibitors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inhibitor_df, inhibitor_metadata = molmedb.get_gene_compound_inhibitor(bridgedb_df=bridgedb_df_hl)\n",
    "inhibitor_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Generating Graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Combing all the results into single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "      <th>StringDB_ppi</th>\n",
       "      <th>WikiPathways</th>\n",
       "      <th>Ensembl_homologs</th>\n",
       "      <th>homolog</th>\n",
       "      <th>OpenTargets_reactome</th>\n",
       "      <th>OpenTargets_go</th>\n",
       "      <th>OpenTargets_gene_compounds</th>\n",
       "      <th>DISGENET_diseases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'ge...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': 'ENSG00000177103'}]</td>\n",
       "      <td>ENSG00000177103</td>\n",
       "      <td>[{'pathway_label': 'DSCAM interactions', 'path...</td>\n",
       "      <td>[{'go_id': 'GO:0007409', 'go_name': 'axonogene...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': 'Alcoholism', 'HPO': 'HPO:HP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00830', 'pathway_labe...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000027452', 'E...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': 'ENSG00000165092'}]</td>\n",
       "      <td>ENSG00000165092</td>\n",
       "      <td>[{'pathway_label': 'Fructose catabolism', 'pat...</td>\n",
       "      <td>[{'go_id': 'GO:0051287', 'go_name': 'NAD bindi...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': 'Melanoma', 'HPO': 'HPO:HP:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00983', 'pathway_labe...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': nan}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00480', 'pathway_labe...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>[{'pathway_id': 'WP:WP4466', 'pathway_label': ...</td>\n",
       "      <td>[{'homolog': 'ENSG00000213366'}]</td>\n",
       "      <td>ENSG00000213366</td>\n",
       "      <td>[{'pathway_label': 'Glutathione conjugation', ...</td>\n",
       "      <td>[{'go_id': 'GO:0043295', 'go_name': 'glutathio...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': 'Carcinoma, Non Small Cell L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00340', 'pathway_labe...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000028356', 'E...</td>\n",
       "      <td>[{'pathway_id': 'WP:WP435', 'pathway_label': '...</td>\n",
       "      <td>[{'homolog': 'ENSG00000160282'}]</td>\n",
       "      <td>ENSG00000160282</td>\n",
       "      <td>[{'pathway_label': 'Histidine catabolism', 'pa...</td>\n",
       "      <td>[{'go_id': 'GO:0005515', 'go_name': 'protein b...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': 'Glutamate formiminotransfer...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'ge...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000026205', 'E...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': 'ENSG00000146039'}]</td>\n",
       "      <td>ENSG00000146039</td>\n",
       "      <td>[{'pathway_label': nan, 'pathway_id': nan}]</td>\n",
       "      <td>[{'go_id': 'GO:0035725', 'go_name': 'sodium io...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': nan, 'HPO': nan, 'NCI': nan,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'ge...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000006398', 'E...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': 'ENSG00000128944'}]</td>\n",
       "      <td>ENSG00000128944</td>\n",
       "      <td>[{'pathway_label': nan, 'pathway_id': nan}]</td>\n",
       "      <td>[{'go_id': 'GO:0005737', 'go_name': 'cytoplasm...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': 'Squamous cell carcinoma of ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'ge...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000006398', 'E...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': 'ENSG00000134690'}]</td>\n",
       "      <td>ENSG00000134690</td>\n",
       "      <td>[{'pathway_label': 'Resolution of Sister Chrom...</td>\n",
       "      <td>[{'go_id': 'GO:0000775', 'go_name': 'chromosom...</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "      <td>[{'disease_name': 'Liver cell carcinoma', 'HPO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'path:mmu04512', 'pathway_labe...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000022415', 'E...</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'homolog': 'ENSG00000159164'}]</td>\n",
       "      <td>ENSG00000159164</td>\n",
       "      <td>[{'pathway_label': 'Toxicity of botulinum toxi...</td>\n",
       "      <td>[{'go_id': 'GO:0030672', 'go_name': 'synaptic ...</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1286', 'drugbank_...</td>\n",
       "      <td>[{'disease_name': 'Schizophrenias', 'HPO': 'HP...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000100426           Ensembl  ENSMUSG00000100426       Ensembl   \n",
       "1  ENSMUSG00000032087           Ensembl  ENSMUSG00000032087       Ensembl   \n",
       "2  ENSMUSG00000024747           Ensembl  ENSMUSG00000024747       Ensembl   \n",
       "3  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813       Ensembl   \n",
       "4  ENSMUSG00000004035           Ensembl  ENSMUSG00000004035       Ensembl   \n",
       "5  ENSMUSG00000001155           Ensembl  ENSMUSG00000001155       Ensembl   \n",
       "6  ENSMUSG00000021336           Ensembl  ENSMUSG00000021336       Ensembl   \n",
       "7  ENSMUSG00000027331           Ensembl  ENSMUSG00000027331       Ensembl   \n",
       "8  ENSMUSG00000028873           Ensembl  ENSMUSG00000028873       Ensembl   \n",
       "9  ENSMUSG00000038486           Ensembl  ENSMUSG00000038486       Ensembl   \n",
       "\n",
       "                                       KEGG_pathways  \\\n",
       "0                                                NaN   \n",
       "1  [{'pathway_id': nan, 'pathway_label': nan, 'ge...   \n",
       "2  [{'pathway_id': 'path:mmu00830', 'pathway_labe...   \n",
       "3  [{'pathway_id': 'path:mmu00983', 'pathway_labe...   \n",
       "4  [{'pathway_id': 'path:mmu00480', 'pathway_labe...   \n",
       "5  [{'pathway_id': 'path:mmu00340', 'pathway_labe...   \n",
       "6  [{'pathway_id': nan, 'pathway_label': nan, 'ge...   \n",
       "7  [{'pathway_id': nan, 'pathway_label': nan, 'ge...   \n",
       "8  [{'pathway_id': nan, 'pathway_label': nan, 'ge...   \n",
       "9  [{'pathway_id': 'path:mmu04512', 'pathway_labe...   \n",
       "\n",
       "                                        StringDB_ppi  \\\n",
       "0  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "1  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "2  [{'stringdb_link_to': 'ENSMUSG00000027452', 'E...   \n",
       "3  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "4  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "5  [{'stringdb_link_to': 'ENSMUSG00000028356', 'E...   \n",
       "6  [{'stringdb_link_to': 'ENSMUSG00000026205', 'E...   \n",
       "7  [{'stringdb_link_to': 'ENSMUSG00000006398', 'E...   \n",
       "8  [{'stringdb_link_to': 'ENSMUSG00000006398', 'E...   \n",
       "9  [{'stringdb_link_to': 'ENSMUSG00000022415', 'E...   \n",
       "\n",
       "                                        WikiPathways  \\\n",
       "0                                                NaN   \n",
       "1  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "2  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "3  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "4  [{'pathway_id': 'WP:WP4466', 'pathway_label': ...   \n",
       "5  [{'pathway_id': 'WP:WP435', 'pathway_label': '...   \n",
       "6  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "7  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "8  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "9  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "\n",
       "                   Ensembl_homologs          homolog  \\\n",
       "0                [{'homolog': nan}]              NaN   \n",
       "1  [{'homolog': 'ENSG00000177103'}]  ENSG00000177103   \n",
       "2  [{'homolog': 'ENSG00000165092'}]  ENSG00000165092   \n",
       "3                [{'homolog': nan}]              NaN   \n",
       "4  [{'homolog': 'ENSG00000213366'}]  ENSG00000213366   \n",
       "5  [{'homolog': 'ENSG00000160282'}]  ENSG00000160282   \n",
       "6  [{'homolog': 'ENSG00000146039'}]  ENSG00000146039   \n",
       "7  [{'homolog': 'ENSG00000128944'}]  ENSG00000128944   \n",
       "8  [{'homolog': 'ENSG00000134690'}]  ENSG00000134690   \n",
       "9  [{'homolog': 'ENSG00000159164'}]  ENSG00000159164   \n",
       "\n",
       "                                OpenTargets_reactome  \\\n",
       "0                                                NaN   \n",
       "1  [{'pathway_label': 'DSCAM interactions', 'path...   \n",
       "2  [{'pathway_label': 'Fructose catabolism', 'pat...   \n",
       "3                                                NaN   \n",
       "4  [{'pathway_label': 'Glutathione conjugation', ...   \n",
       "5  [{'pathway_label': 'Histidine catabolism', 'pa...   \n",
       "6        [{'pathway_label': nan, 'pathway_id': nan}]   \n",
       "7        [{'pathway_label': nan, 'pathway_id': nan}]   \n",
       "8  [{'pathway_label': 'Resolution of Sister Chrom...   \n",
       "9  [{'pathway_label': 'Toxicity of botulinum toxi...   \n",
       "\n",
       "                                      OpenTargets_go  \\\n",
       "0                                                NaN   \n",
       "1  [{'go_id': 'GO:0007409', 'go_name': 'axonogene...   \n",
       "2  [{'go_id': 'GO:0051287', 'go_name': 'NAD bindi...   \n",
       "3                                                NaN   \n",
       "4  [{'go_id': 'GO:0043295', 'go_name': 'glutathio...   \n",
       "5  [{'go_id': 'GO:0005515', 'go_name': 'protein b...   \n",
       "6  [{'go_id': 'GO:0035725', 'go_name': 'sodium io...   \n",
       "7  [{'go_id': 'GO:0005737', 'go_name': 'cytoplasm...   \n",
       "8  [{'go_id': 'GO:0000775', 'go_name': 'chromosom...   \n",
       "9  [{'go_id': 'GO:0030672', 'go_name': 'synaptic ...   \n",
       "\n",
       "                          OpenTargets_gene_compounds  \\\n",
       "0                                                NaN   \n",
       "1  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "2  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "3                                                NaN   \n",
       "4  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "5  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "6  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "7  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "8  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...   \n",
       "9  [{'chembl_id': 'CHEMBL:CHEMBL1286', 'drugbank_...   \n",
       "\n",
       "                                   DISGENET_diseases  \n",
       "0                                                NaN  \n",
       "1  [{'disease_name': 'Alcoholism', 'HPO': 'HPO:HP...  \n",
       "2  [{'disease_name': 'Melanoma', 'HPO': 'HPO:HP:0...  \n",
       "3                                                NaN  \n",
       "4  [{'disease_name': 'Carcinoma, Non Small Cell L...  \n",
       "5  [{'disease_name': 'Glutamate formiminotransfer...  \n",
       "6  [{'disease_name': nan, 'HPO': nan, 'NCI': nan,...  \n",
       "7  [{'disease_name': 'Squamous cell carcinoma of ...  \n",
       "8  [{'disease_name': 'Liver cell carcinoma', 'HPO...  \n",
       "9  [{'disease_name': 'Schizophrenias', 'HPO': 'HP...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#        bgee_df,\n",
    "#        disgenet_df,\n",
    "#        minerva_df,\n",
    "#        opentargets_reactome_df,\n",
    "#        opentargets_go_df,\n",
    "#        opentargets_compound_df,\n",
    "#        inhibitor_df,\n",
    "#        kegg_df,\n",
    "#        wikipathways_df,\n",
    "#        ppi_df,\n",
    "#        ensembl_homologs_df,\n",
    "\n",
    "combined_df = combine_sources(\n",
    "    bridgedb_df,\n",
    "    [kegg_df, ppi_df, wikipathways_df, ensembl_homologs_df],\n",
    ")\n",
    "\n",
    "combined_df = combine_with_homologs(\n",
    "    combined_df,\n",
    "    [\n",
    "        opentargets_reactome_df,\n",
    "        opentargets_go_df,\n",
    "        opentargets_compound_df,\n",
    "        disgenet_df,\n",
    "    ],\n",
    ")\n",
    "\n",
    "\n",
    "combined_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_df = combine_sources(\n",
    "    bridgedb_df,\n",
    "    [intact_df],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5, 5)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Exporting the database in pickle format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"combined_df.pkl\", \"wb\") as out:\n",
    "    pickle.dump(combined_df, out)\n",
    "# with open(\"opentargets_disease_compound_df.pkl\", \"wb\") as out:\n",
    "#     pickle.dump(opentargets_disease_compound_df, out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Creating a graph from the annotated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>IntAct_interactions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'interaction_id': 'EBI-6909752', 'interactor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'interaction_id': 'EBI-22300555', 'interacto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000084349</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000084349</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000027490</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027490</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'interaction_id': 'EBI-1186759', 'interactor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000046840</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000046840</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000067274           Ensembl  ENSMUSG00000067274       Ensembl   \n",
       "1  ENSMUSG00000000001           Ensembl  ENSMUSG00000000001       Ensembl   \n",
       "2  ENSMUSG00000084349           Ensembl  ENSMUSG00000084349       Ensembl   \n",
       "3  ENSMUSG00000027490           Ensembl  ENSMUSG00000027490       Ensembl   \n",
       "4  ENSMUSG00000046840           Ensembl  ENSMUSG00000046840       Ensembl   \n",
       "\n",
       "                                 IntAct_interactions  \n",
       "0  [{'interaction_id': 'EBI-6909752', 'interactor...  \n",
       "1  [{'interaction_id': 'EBI-22300555', 'interacto...  \n",
       "2                                                 []  \n",
       "3  [{'interaction_id': 'EBI-1186759', 'interactor...  \n",
       "4                                                 []  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combined_df = generator.load_dataframe_from_pickle(\"combined_df.pkl\")\n",
    "# opentargets_disease_compound_df = generator.load_dataframe_from_pickle(\n",
    "#     \"opentargets_disease_compound_df.pkl\"\n",
    "# )\n",
    "\n",
    "combined_df.head(15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pathway_id': nan,\n",
       "  'pathway_label': nan,\n",
       "  'gene_count': nan,\n",
       "  'compounds': [{'KEGG_identifier': None, 'name': None}]}]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df[\"KEGG_pathways\"][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_disease_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UMLS_C0000786</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_1001255</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1276308', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UMLS_C0001175</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000765</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL704', 'drugbank_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UMLS_C0002103</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0005854</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1201353', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UMLS_C0002171</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0004192</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1200963', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UMLS_C0002940</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0004264</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1076903', 'drugba...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      identifier identifier.source       target target.source  \\\n",
       "0  UMLS_C0000786              UMLS  EFO_1001255           EFO   \n",
       "1  UMLS_C0001175              UMLS  EFO_0000765           EFO   \n",
       "2  UMLS_C0002103              UMLS  EFO_0005854           EFO   \n",
       "3  UMLS_C0002171              UMLS  EFO_0004192           EFO   \n",
       "4  UMLS_C0002940              UMLS  EFO_0004264           EFO   \n",
       "\n",
       "                       OpenTargets_disease_compounds  \n",
       "0  [{'chembl_id': 'CHEMBL:CHEMBL1276308', 'drugba...  \n",
       "1  [{'chembl_id': 'CHEMBL:CHEMBL704', 'drugbank_i...  \n",
       "2  [{'chembl_id': 'CHEMBL:CHEMBL1201353', 'drugba...  \n",
       "3  [{'chembl_id': 'CHEMBL:CHEMBL1200963', 'drugba...  \n",
       "4  [{'chembl_id': 'CHEMBL:CHEMBL1076903', 'drugba...  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_disease_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building graph:   0%|          | 0/81 [00:00<?, ?it/s]"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building graph: 100%|██████████| 81/81 [00:00<00:00, 500.39it/s]\n",
      "100%|██████████| 81/81 [00:00<00:00, 1945.18it/s]\n"
     ]
    }
   ],
   "source": [
    "pygraph = generator.build_networkx_graph(\n",
    "    combined_df,\n",
    "    disease_compound=opentargets_disease_compound_df,\n",
    "    pathway_compound=kegg_compound_df,\n",
    "    homolog_df_list=[\n",
    "        opentargets_reactome_df,\n",
    "        opentargets_compound_df,\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building graph: 100%|██████████| 5/5 [00:00<00:00, 2984.84it/s]\n",
      "Building graph: 100%|██████████| 5/5 [00:00<00:00, 3236.85it/s]\n"
     ]
    }
   ],
   "source": [
    "pygraph = generator.build_networkx_graph(combined_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MultiDiGraph with 5 nodes and 0 edges\n"
     ]
    }
   ],
   "source": [
    "print(pygraph)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Store the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"networkx_graph_test.pkl\", \"wb\") as out:\n",
    "    pickle.dump(pygraph, out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 Visualize the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pos = nx.circular_layout(pygraph)\n",
    "\n",
    "# plt.figure(3, figsize=(30, 30))\n",
    "# nx.draw(pygraph, pos)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Exporting Graph to external sources"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1 Cytoscape\n",
    "Make sure that the Cytoscape is open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'source'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3608\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3607\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 3608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   3609\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\pandas\\_libs\\index.pyx:147\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\pandas\\_libs\\index.pyx:176\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:7080\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:7088\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'source'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[11], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpyBiodatafuse\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mgraph\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m cytoscape\n\u001b[1;32m----> 3\u001b[0m \u001b[43mcytoscape\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mload_graph\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpygraph\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mnetwork_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mTest network\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\graph\\cytoscape.py:110\u001b[0m, in \u001b[0;36mload_graph\u001b[1;34m(g, network_name)\u001b[0m\n\u001b[0;32m    100\u001b[0m default \u001b[38;5;241m=\u001b[39m {\n\u001b[0;32m    101\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtitle\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mBioDataFuse_style\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[0;32m    102\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdefaults\u001b[39m\u001b[38;5;124m\"\u001b[39m: [\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    106\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmappings\u001b[39m\u001b[38;5;124m\"\u001b[39m: [],\n\u001b[0;32m    107\u001b[0m }\n\u001b[0;32m    109\u001b[0m \u001b[38;5;66;03m# Create the network in Cytoscape\u001b[39;00m\n\u001b[1;32m--> 110\u001b[0m \u001b[43mp4c\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnetworks\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_network_from_networkx\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m    111\u001b[0m \u001b[43m    \u001b[49m\u001b[43madj_g\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    112\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtitle\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mnetwork_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    113\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcollection\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mBioDataFuse\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[0;32m    114\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    116\u001b[0m \u001b[38;5;66;03m# Apply the visual style\u001b[39;00m\n\u001b[0;32m    117\u001b[0m p4c\u001b[38;5;241m.\u001b[39mstyles\u001b[38;5;241m.\u001b[39mcreate_visual_style(default)\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\py4cytoscape\\py4cytoscape_logger.py:133\u001b[0m, in \u001b[0;36mcy_log.<locals>.wrapper_log\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    131\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m log_return(func, value)\n\u001b[0;32m    132\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m--> 133\u001b[0m     \u001b[43mlog_exception\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43me\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    134\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[0;32m    135\u001b[0m     log_finally()\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\py4cytoscape\\py4cytoscape_logger.py:130\u001b[0m, in \u001b[0;36mcy_log.<locals>.wrapper_log\u001b[1;34m(*args, **kwargs)\u001b[0m\n\u001b[0;32m    128\u001b[0m log_incoming(func, \u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m    129\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 130\u001b[0m     value \u001b[38;5;241m=\u001b[39m func(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;66;03m# Call function being logged\u001b[39;00m\n\u001b[0;32m    131\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m log_return(func, value)\n\u001b[0;32m    132\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\py4cytoscape\\networks.py:1028\u001b[0m, in \u001b[0;36mcreate_network_from_networkx\u001b[1;34m(netx, title, collection, base_url)\u001b[0m\n\u001b[0;32m   1024\u001b[0m \u001b[38;5;66;03m# TODO: This will blow if there are no edges or no nodes ... so will create_network_from_igraph() ... will R blow, too?\u001b[39;00m\n\u001b[0;32m   1025\u001b[0m \n\u001b[0;32m   1026\u001b[0m \u001b[38;5;66;03m# Make sure critical attributes are strings\u001b[39;00m\n\u001b[0;32m   1027\u001b[0m node_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m node_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m)\n\u001b[1;32m-> 1028\u001b[0m edge_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124msource\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[43medge_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msource\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m)\n\u001b[0;32m   1029\u001b[0m edge_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m edge_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtarget\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m)\n\u001b[0;32m   1030\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124minteraction\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;129;01min\u001b[39;00m edge_df\u001b[38;5;241m.\u001b[39mcolumns: edge_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124minteraction\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m edge_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124minteraction\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\pandas\\core\\frame.py:3716\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3714\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m   3715\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 3716\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   3717\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   3718\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3610\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3608\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3609\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[1;32m-> 3610\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[0;32m   3611\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[0;32m   3612\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3613\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3614\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3615\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'source'"
     ]
    }
   ],
   "source": [
    "from pyBiodatafuse.graph import cytoscape\n",
    "\n",
    "cytoscape.load_graph(pygraph, network_name=\"Test network\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2 Neo4j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pygraph' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[17], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpyBiodatafuse\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mgraph\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m neo4j\n\u001b[1;32m----> 3\u001b[0m neo4j\u001b[38;5;241m.\u001b[39msave_graph_to_graphml(\u001b[43mpygraph\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnetworkx_graph_usecase_with_homologs.graphml\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'pygraph' is not defined"
     ]
    }
   ],
   "source": [
    "from pyBiodatafuse.graph import neo4j\n",
    "\n",
    "neo4j.save_graph_to_graphml(pygraph, \"networkx_graph_usecase_with_homologs.graphml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps to load the graph in Neo4j"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `.graphml` file in **import** subfolder of the DBMS folder\n",
    "- Install apoc plugin\n",
    "- Create `apoc.conf` file:\n",
    "    ```\n",
    "    apoc.trigger.enabled=true\n",
    "    apoc.import.file.enabled=true\n",
    "    apoc.export.file.enabled=true\n",
    "    apoc.import.file.use_neo4j_config=true\n",
    "    ```\n",
    "- Add `apoc.conf` file to **conf** subfolder of the DBMS folder\n",
    "- Open Neo4j Browser\n",
    "- (Optionl, only run if you have imported a graph  before) Remove all the nodes before importing `.graphml` file\n",
    "\n",
    "    ```MATCH (n) DETACH DELETE n```\n",
    "\n",
    "- Import `.graphml` file\n",
    "\n",
    "    ```call apoc.import.graphml('file:///networkx_graph_test.graphml',{readLabels:TRUE})```\n",
    "\n",
    "- Add indexes after importing the graph for improving the performance of queries\n",
    "\n",
    "    ```\n",
    "    create index Gene for (n:Gene) on (n.node_type)\n",
    "    create index Pathway for (n:Pathway) on (n.node_type)\n",
    "    create index `Biological Process` for (n:`Biological Process`) on (n.node_type)\n",
    "    create index `Molecular Function` for (n:`Molecular Function`) on (n.node_type)\n",
    "    create index `Cellular Component` for (n:`Cellular Component`) on (n.node_type)\n",
    "    create index Disease for (n:Disease) on (n.node_type)\n",
    "    create index Compound for (n:Compound) on (n.node_type)\n",
    "    create index `Side Effect` for (n:`Side Effect`) on (n.node_type)\n",
    "    ```\n",
    "    \n",
    "\n",
    "- Count the number of each node type\n",
    "    - total (```MATCH (n) RETURN count(n)```) \n",
    "        - Gene (```MATCH (n:Gene) RETURN count(n)```)\n",
    "        - Pathway (```MATCH (n:Pathway) RETURN count(n)```)\n",
    "            - WikiPathways (```MATCH (n:Pathway {source: \"WikiPathways\"}) RETURN count(n)```) \n",
    "            - OpenTargets, Reactome (```MATCH (n:Pathway {source: \"OpenTargets\"}) RETURN count(n)```) \n",
    "            - MINERVA (```MATCH (n:Pathway {source: \"MINERVA\"}) RETURN count(n)```) \n",
    "        - Biological Process (```MATCH (n:`Biological Process`) RETURN count(n)```) \n",
    "        - Molecular Function (```MATCH (n:`Molecular Function`) RETURN count(n)```) \n",
    "        - Cellular Component (```MATCH (n:`Cellular Component`) RETURN count(n)```) \n",
    "        - Disease (```MATCH (n:Disease) RETURN count(n)```) \n",
    "        - Compound (```MATCH (n:Compound) RETURN count(n)```)\n",
    "        - Side Effect (```MATCH (n:`Side Effect`) RETURN count(n)```) \n",
    "- Count the number of each edge type\n",
    "    - total (```MATCH ()-[r]->() RETURN count(r)```) \n",
    "        - interacts_with (```MATCH ()-[r:interacts_with]->() RETURN count(r)```) \n",
    "        - part_of (```MATCH ()-[r:part_of]->() RETURN count(r)```) \n",
    "            - WikiPathways (```MATCH ()-[r:part_of {source: \"WikiPathways\"}]->() RETURN count(r)```) \n",
    "            - OpenTargets, Reactome (```MATCH ()-[r:part_of {source: \"OpenTargets\"}]->() RETURN count(r)```) \n",
    "            - MINERVA (```MATCH ()-[r:part_of {source: \"MINERVA\"}]->() RETURN count(r)```) \n",
    "        - activates (```MATCH ()-[r:activates]->() RETURN count(r)```) \n",
    "        - treats (```MATCH ()-[r:treats]->() RETURN count(r)```) \n",
    "        - has_side_effect (```MATCH ()-[r:has_side_effect]->() RETURN count(r)```) \n",
    "        - inhibits (```MATCH ()-[r:inhibits]->() RETURN count(r)```) = 71\n",
    "        - associated_with (```MATCH ()-[r:associated_with]->() RETURN count(r)```) \n",
    "\n",
    "- Export the graph as a `.csv` file\n",
    "\n",
    "    ```call apoc.export.csv.all(\"networkx_graph_test.csv\",{})```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
