{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example: Polysistice Kidney Disease use case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39\\\\python39.zip', 'C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39\\\\DLLs', 'C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39\\\\lib', 'C:\\\\Users\\\\are10\\\\AppData\\\\Local\\\\Programs\\\\Python\\\\Python39', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv', '', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages\\\\win32', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages\\\\win32\\\\lib', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\.venv\\\\lib\\\\site-packages\\\\Pythonwin', 'c:\\\\Users\\\\are10\\\\Documents\\\\BAFSTU\\\\code\\\\pyBioDatafusemain\\\\pyBiodatafuse\\\\src']\n"
     ]
    }
   ],
   "source": [
    "new_path = r\"C:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\examples\\usecases\"\n",
    "\n",
    "# Setting up the working directory\n",
    "import os\n",
    "import sys\n",
    "\n",
    "src_path = os.path.abspath(os.path.join(\"..\", \"..\", \"..\", \"src\"))\n",
    "if src_path not in sys.path:\n",
    "    sys.path.append(src_path)\n",
    "\n",
    "print(sys.path)  # Verify the correct src path is included"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import modules\n",
    "import pickle\n",
    "\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "\n",
    "from pyBiodatafuse import id_mapper\n",
    "from pyBiodatafuse import human_homologs\n",
    "from pyBiodatafuse.annotators import (\n",
    "    bgee,\n",
    "    disgenet,\n",
    "    kegg,\n",
    "    minerva,\n",
    "    molmedb,\n",
    "    opentargets,\n",
    "    pubchem,\n",
    "    stringdb,\n",
    "    wikipathways,\n",
    ")\n",
    "from pyBiodatafuse.constants import (\n",
    "    BGEE_GENE_EXPRESSION_LEVELS_COL,\n",
    "    DISGENET_DISEASE_COL,\n",
    "    ENSEMBL_HOMOLOG_COL,\n",
    "    KEGG_COL,\n",
    "    MINERVA,\n",
    "    MOLMEDB_PROTEIN_COMPOUND_COL,\n",
    "    OPENTARGETS_DISEASE_COMPOUND_COL,\n",
    "    OPENTARGETS_GENE_COMPOUND_COL,\n",
    "    OPENTARGETS_GO_COL,\n",
    "    OPENTARGETS_REACTOME_COL,\n",
    "    PUBCHEM_COMPOUND_ASSAYS_COL,\n",
    "    STRING_PPI_COL,\n",
    "    WIKIPATHWAYS,\n",
    ")\n",
    "from pyBiodatafuse.graph import generator\n",
    "from pyBiodatafuse.utils import combine_sources, create_harmonized_input_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Entity resolution using BridgeDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1. Load the input list and convert it to a dataframe\n",
    "\n",
    "Small set of mouse data used for debugging."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEST Mice Ensembl\n",
    "genes_of_interest = \"\"\"ENSMUSG00000067274\n",
    "ENSMUSG00000000001\n",
    "ENSMUSG00000084349\n",
    "ENSMUSG00000025428\n",
    "ENSMUSG00000044533\"\"\"\n",
    "\n",
    "# TEST Mice use case = \n",
    "genes_of_interest = \"\"\"ENSMUSG00000026295\n",
    "ENSMUSG00000022877\n",
    "ENSMUSG00000020914\n",
    "ENSMUSG00000024747\n",
    "ENSMUSG00000032081\n",
    "ENSMUSG00000004035\n",
    "ENSMUSG00000072949\n",
    "ENSMUSG00000028970\n",
    "ENSMUSG00000028937\n",
    "ENSMUSG00000075044\n",
    "\"\"\"\n",
    "\n",
    "# TEST Rat Ensembl\n",
    "# genes_of_interest = \"\"\"ENSRNOG00060027926\n",
    "# ENSRNOG00055005387\n",
    "# ENSRNOG00060018596\n",
    "# ENSRNOG00060011358\n",
    "# ENSRNOG00055009275 \n",
    "# \"\"\"\n",
    "\n",
    "# TEST Human HGNC\n",
    "# genes_of_interest = \"\"\"CHRNG\n",
    "# DMD\n",
    "# AHR\n",
    "# SCN4A\n",
    "# LC25A1\n",
    "# HTR3A\"\"\"\n",
    "\n",
    "# TEST Human Ensembl\n",
    "# genes_of_interest = \"\"\"ENSG00000072080\n",
    "# ENSG00000113905\n",
    "# ENSG00000131747\n",
    "# ENSG00000165092\n",
    "# ENSG00000110245\n",
    "# ENSG00000213366\n",
    "# ENSG00000184227\n",
    "# ENSG00000085563\n",
    "# ENSG00000097021\n",
    "# ENSG00000149742\"\"\"\n",
    "\n",
    "gene_list = genes_of_interest.split(\"\\n\")\n",
    "len(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Mouse use case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "81\n"
     ]
    }
   ],
   "source": [
    "from pyBiodatafuse import data_loader, id_mapper\n",
    "\n",
    "data_input = data_loader.create_df_from_dea(\"data/full_de_genes_treated_vs_untreated_plus_cpm_fc.csv\")\n",
    "data_filtered = data_input[data_input['DE'].isin([1, -1])]\n",
    "features_filtered = data_filtered['identifier']\n",
    "\n",
    "gene_list = features_filtered.tolist()\n",
    "print(len(gene_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000044816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000003038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000015451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000026295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000056749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000016756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000027452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000027577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000029368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000026205</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            identifier\n",
       "0   ENSMUSG00000100426\n",
       "1   ENSMUSG00000032087\n",
       "2   ENSMUSG00000024747\n",
       "3   ENSMUSG00000091813\n",
       "4   ENSMUSG00000004035\n",
       "5   ENSMUSG00000001155\n",
       "6   ENSMUSG00000021336\n",
       "7   ENSMUSG00000027331\n",
       "8   ENSMUSG00000028873\n",
       "9   ENSMUSG00000038486\n",
       "10  ENSMUSG00000044816\n",
       "11  ENSMUSG00000003038\n",
       "12  ENSMUSG00000015451\n",
       "13  ENSMUSG00000026295\n",
       "14  ENSMUSG00000056749\n",
       "15  ENSMUSG00000016756\n",
       "16  ENSMUSG00000027452\n",
       "17  ENSMUSG00000027577\n",
       "18  ENSMUSG00000029368\n",
       "19  ENSMUSG00000026205"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_input = pd.DataFrame(gene_list, columns=[\"identifier\"])\n",
    "data_input.head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSMUSG00000100426', 'ENSMUSG00000032087', 'ENSMUSG00000024747', 'ENSMUSG00000091813', 'ENSMUSG00000004035', 'ENSMUSG00000001155', 'ENSMUSG00000021336', 'ENSMUSG00000027331', 'ENSMUSG00000028873', 'ENSMUSG00000038486', 'ENSMUSG00000044816', 'ENSMUSG00000003038', 'ENSMUSG00000015451', 'ENSMUSG00000026295', 'ENSMUSG00000056749', 'ENSMUSG00000016756', 'ENSMUSG00000027452', 'ENSMUSG00000027577', 'ENSMUSG00000029368', 'ENSMUSG00000026205', 'ENSMUSG00000027379', 'ENSMUSG00000028715', 'ENSMUSG00000035186', 'ENSMUSG00000043439', 'ENSMUSG00000055629', 'ENSMUSG00000079494', 'ENSMUSG00000045328', 'ENSMUSG00000075044', 'ENSMUSG00000020914', 'ENSMUSG00000001334', 'ENSMUSG00000006398', 'ENSMUSG00000022415', 'ENSMUSG00000021213', 'ENSMUSG00000002870', 'ENSMUSG00000061959', 'ENSMUSG00000001313', 'ENSMUSG00000007950', 'ENSMUSG00000017390', 'ENSMUSG00000025983', 'ENSMUSG00000026049', 'ENSMUSG00000026415', 'ENSMUSG00000030825', 'ENSMUSG00000037628', 'ENSMUSG00000048489', 'ENSMUSG00000051483', 'ENSMUSG00000025196', 'ENSMUSG00000032113', 'ENSMUSG00000037887', 'ENSMUSG00000054612', 'ENSMUSG00000080115', 'ENSMUSG00000004356', 'ENSMUSG00000022877', 'ENSMUSG00000034855', 'ENSMUSG00000020330', 'ENSMUSG00000027490', 'ENSMUSG00000020083', 'ENSMUSG00000026981', 'ENSMUSG00000022026', 'ENSMUSG00000037440', 'ENSMUSG00000071547', 'ENSMUSG00000028937', 'ENSMUSG00000051262', 'ENSMUSG00000046840', 'ENSMUSG00000048572', 'ENSMUSG00000097927', 'ENSMUSG00000023015', 'ENSMUSG00000041939', 'ENSMUSG00000019942', 'ENSMUSG00000000934', 'ENSMUSG00000027306', 'ENSMUSG00000028356', 'ENSMUSG00000028970', 'ENSMUSG00000032218', 'ENSMUSG00000042029', 'ENSMUSG00000070661', 'ENSMUSG00000072949', 'ENSMUSG00000023505', 'ENSMUSG00000032081', 'ENSMUSG00000024521', 'ENSMUSG00000042041', 'ENSMUSG00000031016']\n"
     ]
    }
   ],
   "source": [
    "print(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2. Query BridgeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:3782384</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:2150309</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>114873</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>A0A1L1SQZ7</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E9QPR7</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q4VA61</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>A0A1L1SQ53</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>26358</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>O35945</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>B2RTL5</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:1347050</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>F6Z9B9</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:3648740</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>436059</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>68312</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YVP5</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YVP6</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YZ29</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E9QAC8</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q80W21</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:1915562</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            identifier identifier.source              target   target.source\n",
       "0   ENSMUSG00000100426           Ensembl         MGI:3782384             MGI\n",
       "1   ENSMUSG00000100426           Ensembl  ENSMUSG00000100426         Ensembl\n",
       "2   ENSMUSG00000032087           Ensembl         MGI:2150309             MGI\n",
       "3   ENSMUSG00000032087           Ensembl              114873       NCBI Gene\n",
       "4   ENSMUSG00000032087           Ensembl          A0A1L1SQZ7  Uniprot-TrEMBL\n",
       "5   ENSMUSG00000032087           Ensembl              E9QPR7  Uniprot-TrEMBL\n",
       "6   ENSMUSG00000032087           Ensembl              Q4VA61  Uniprot-TrEMBL\n",
       "7   ENSMUSG00000032087           Ensembl          A0A1L1SQ53  Uniprot-TrEMBL\n",
       "8   ENSMUSG00000032087           Ensembl  ENSMUSG00000032087         Ensembl\n",
       "9   ENSMUSG00000024747           Ensembl  ENSMUSG00000024747         Ensembl\n",
       "10  ENSMUSG00000024747           Ensembl               26358       NCBI Gene\n",
       "11  ENSMUSG00000024747           Ensembl              O35945  Uniprot-TrEMBL\n",
       "12  ENSMUSG00000024747           Ensembl              B2RTL5  Uniprot-TrEMBL\n",
       "13  ENSMUSG00000024747           Ensembl         MGI:1347050             MGI\n",
       "14  ENSMUSG00000091813           Ensembl              F6Z9B9  Uniprot-TrEMBL\n",
       "15  ENSMUSG00000091813           Ensembl         MGI:3648740             MGI\n",
       "16  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813         Ensembl\n",
       "17  ENSMUSG00000091813           Ensembl              436059       NCBI Gene\n",
       "18  ENSMUSG00000004035           Ensembl               68312       NCBI Gene\n",
       "19  ENSMUSG00000004035           Ensembl              D3YVP5  Uniprot-TrEMBL\n",
       "20  ENSMUSG00000004035           Ensembl              D3YVP6  Uniprot-TrEMBL\n",
       "21  ENSMUSG00000004035           Ensembl              D3YZ29  Uniprot-TrEMBL\n",
       "22  ENSMUSG00000004035           Ensembl              E9QAC8  Uniprot-TrEMBL\n",
       "23  ENSMUSG00000004035           Ensembl              Q80W21  Uniprot-TrEMBL\n",
       "24  ENSMUSG00000004035           Ensembl         MGI:1915562             MGI"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mouse usecase\n",
    "input_species=\"Mouse\"\n",
    "\n",
    "bridgedb_df, bridgedb_metadata = id_mapper.bridgedb_xref(\n",
    "   identifiers=data_input,\n",
    "    input_species=input_species,\n",
    "    input_datasource=\"Ensembl\",\n",
    "    output_datasource=\"All\",\n",
    ")\n",
    "\n",
    "# TEST Human data\n",
    "# bridgedb_df, bridgedb_metadata = id_mapper.bridgedb_xref(\n",
    "#     identifiers=data_input,\n",
    "#     input_species=\"Human\",\n",
    "#     input_datasource=\"Ensembl\",\n",
    "#     output_datasource=\"All\",\n",
    "# )\n",
    "\n",
    "bridgedb_df.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Homologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"release\":\"15.9\"}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>Ensembl_homologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000177103'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000165092'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'homolog': 'ENSG00000213366'}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000100426           Ensembl  ENSMUSG00000100426       Ensembl   \n",
       "1  ENSMUSG00000032087           Ensembl  ENSMUSG00000032087       Ensembl   \n",
       "2  ENSMUSG00000024747           Ensembl  ENSMUSG00000024747       Ensembl   \n",
       "3  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813       Ensembl   \n",
       "4  ENSMUSG00000004035           Ensembl  ENSMUSG00000004035       Ensembl   \n",
       "\n",
       "                   Ensembl_homologs  \n",
       "0                              None  \n",
       "1  [{'homolog': 'ENSG00000177103'}]  \n",
       "2  [{'homolog': 'ENSG00000165092'}]  \n",
       "3                              None  \n",
       "4  [{'homolog': 'ENSG00000213366'}]  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensembl_homologs_df, ensembl_metadata = human_homologs.get_homologs(bridgedb_df=bridgedb_df)\n",
    "ensembl_homologs_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ENSG00000177103', 'ENSG00000165092', 'ENSG00000213366', 'ENSG00000160282', 'ENSG00000146039', 'ENSG00000128944', 'ENSG00000134690', 'ENSG00000159164', 'ENSG00000188674', 'ENSG00000224389', 'ENSG00000072080', 'ENSG00000165030', 'ENSG00000154930', 'ENSG00000101204', 'ENSG00000163631', 'ENSG00000213901', 'ENSG00000169679', 'ENSG00000213886', 'ENSG00000273604', 'ENSG00000182272', 'ENSG00000144035', 'ENSG00000138778', 'ENSG00000149742', 'ENSG00000131747', 'ENSG00000160097', 'ENSG00000117399', 'ENSG00000100321', 'ENSG00000073111', 'ENSG00000108830', 'ENSG00000127220', 'ENSG00000109107', 'ENSG00000144395', 'ENSG00000151287', 'ENSG00000162897', 'ENSG00000087076', 'ENSG00000100526', 'ENSG00000165507', 'ENSG00000159228', 'ENSG00000120054', 'ENSG00000149554', 'ENSG00000184545', 'ENSG00000170430', 'ENSG00000123427', 'ENSG00000120800', 'ENSG00000113905', 'ENSG00000169245', 'ENSG00000072571', 'ENSG00000101412', 'ENSG00000136689', 'ENSG00000102837', 'ENSG00000112299', 'ENSG00000168268', 'ENSG00000097021', 'ENSG00000144035', 'ENSG00000181778', 'ENSG00000161800', 'ENSG00000110921', 'ENSG00000170312', 'ENSG00000184428', 'ENSG00000137804', 'ENSG00000106927', 'ENSG00000085563', 'ENSG00000157456', 'ENSG00000146918', 'ENSG00000178828', 'ENSG00000184227', 'ENSG00000111665', 'ENSG00000110245', 'ENSG00000141682', 'ENSG00000173237', 'ENSG00000166483']\n"
     ]
    }
   ],
   "source": [
    "homologs = ensembl_homologs_df[ENSEMBL_HOMOLOG_COL].apply(\n",
    "    lambda x: x[0]['homolog'] if isinstance(x, list) and len(x) > 0 and 'homolog' in x[0] else None\n",
    ").dropna().tolist()\n",
    "\n",
    "print(homologs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Query homologs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:3782384</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:2150309</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>114873</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>A0A1L1SQZ7</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E9QPR7</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q4VA61</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>A0A1L1SQ53</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>26358</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>O35945</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>B2RTL5</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:1347050</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>F6Z9B9</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:3648740</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>436059</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>68312</td>\n",
       "      <td>NCBI Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YVP5</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YVP6</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>D3YZ29</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E9QAC8</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q80W21</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>MGI:1915562</td>\n",
       "      <td>MGI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            identifier identifier.source              target   target.source\n",
       "0   ENSMUSG00000100426           Ensembl         MGI:3782384             MGI\n",
       "1   ENSMUSG00000100426           Ensembl  ENSMUSG00000100426         Ensembl\n",
       "2   ENSMUSG00000032087           Ensembl         MGI:2150309             MGI\n",
       "3   ENSMUSG00000032087           Ensembl              114873       NCBI Gene\n",
       "4   ENSMUSG00000032087           Ensembl          A0A1L1SQZ7  Uniprot-TrEMBL\n",
       "5   ENSMUSG00000032087           Ensembl              E9QPR7  Uniprot-TrEMBL\n",
       "6   ENSMUSG00000032087           Ensembl              Q4VA61  Uniprot-TrEMBL\n",
       "7   ENSMUSG00000032087           Ensembl          A0A1L1SQ53  Uniprot-TrEMBL\n",
       "8   ENSMUSG00000032087           Ensembl  ENSMUSG00000032087         Ensembl\n",
       "9   ENSMUSG00000024747           Ensembl  ENSMUSG00000024747         Ensembl\n",
       "10  ENSMUSG00000024747           Ensembl               26358       NCBI Gene\n",
       "11  ENSMUSG00000024747           Ensembl              O35945  Uniprot-TrEMBL\n",
       "12  ENSMUSG00000024747           Ensembl              B2RTL5  Uniprot-TrEMBL\n",
       "13  ENSMUSG00000024747           Ensembl         MGI:1347050             MGI\n",
       "14  ENSMUSG00000091813           Ensembl              F6Z9B9  Uniprot-TrEMBL\n",
       "15  ENSMUSG00000091813           Ensembl         MGI:3648740             MGI\n",
       "16  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813         Ensembl\n",
       "17  ENSMUSG00000091813           Ensembl              436059       NCBI Gene\n",
       "18  ENSMUSG00000004035           Ensembl               68312       NCBI Gene\n",
       "19  ENSMUSG00000004035           Ensembl              D3YVP5  Uniprot-TrEMBL\n",
       "20  ENSMUSG00000004035           Ensembl              D3YVP6  Uniprot-TrEMBL\n",
       "21  ENSMUSG00000004035           Ensembl              D3YZ29  Uniprot-TrEMBL\n",
       "22  ENSMUSG00000004035           Ensembl              E9QAC8  Uniprot-TrEMBL\n",
       "23  ENSMUSG00000004035           Ensembl              Q80W21  Uniprot-TrEMBL\n",
       "24  ENSMUSG00000004035           Ensembl         MGI:1915562             MGI"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_species=\"Human\"\n",
    "\n",
    "data_input_hl = pd.DataFrame(homologs, columns=[\"identifier\"])\n",
    "\n",
    "bridgedb_df_hl, bridgedb_metadata_hl = id_mapper.bridgedb_xref(\n",
    "    identifiers=data_input_hl,\n",
    "    input_species=\"Human\",\n",
    "    input_datasource=\"Ensembl\",\n",
    "    output_datasource=\"All\",\n",
    ")\n",
    "\n",
    "bridgedb_df.head(25)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Step-by-step graph generation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1. Gene-Disease edges\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15eadd18-9b50-466d-a41d-3deb5fad122c\n"
     ]
    }
   ],
   "source": [
    "load_dotenv('disgenet.env')\n",
    "\n",
    "disgenet_api_key = os.getenv(\"DISGENET_API_KEY\")\n",
    "print(disgenet_api_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying DisGeNET: 100%|██████████| 71/71 [01:10<00:00,  1.01it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_22080\\374358601.py:1: UserWarning: The intermediate_df in DISGENET annotatur should be checked, please create an issue on https://github.com/BioDataFuse/pyBiodatafuse/issues/.\n",
      "  disgenet_df, disgenet_metadata = disgenet.get_gene_disease(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>DISGENET_diseases</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>6694</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Retinal dystrophy', 'HPO': ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>3161</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Cancer, Breast', 'HPO': Non...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>4171</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Liver cell carcinoma', 'HPO...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>5243</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Breast Neoplasms', 'HPO': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>51171</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'disease_name': 'Severe myopia (&gt; -6.00 diop...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source target target.source  \\\n",
       "0  ENSG00000072080           Ensembl   6694     NCBI Gene   \n",
       "1  ENSG00000072571           Ensembl   3161     NCBI Gene   \n",
       "2  ENSG00000073111           Ensembl   4171     NCBI Gene   \n",
       "3  ENSG00000085563           Ensembl   5243     NCBI Gene   \n",
       "4  ENSG00000087076           Ensembl  51171     NCBI Gene   \n",
       "\n",
       "                                   DISGENET_diseases  \n",
       "0  [{'disease_name': 'Retinal dystrophy', 'HPO': ...  \n",
       "1  [{'disease_name': 'Cancer, Breast', 'HPO': Non...  \n",
       "2  [{'disease_name': 'Liver cell carcinoma', 'HPO...  \n",
       "3  [{'disease_name': 'Breast Neoplasms', 'HPO': '...  \n",
       "4  [{'disease_name': 'Severe myopia (> -6.00 diop...  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disgenet_df, disgenet_metadata = disgenet.get_gene_disease(\n",
    "    api_key=disgenet_api_key, bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "disgenet_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'disease_name': 'ARCND1',\n",
       "  'HPO': None,\n",
       "  'NCI': None,\n",
       "  'OMIM': 'MIM:602483, MIM:139370',\n",
       "  'MONDO': 'MONDO:0000107, MONDO:0011234',\n",
       "  'ORDO': 'ORDO:137888',\n",
       "  'EFO': None,\n",
       "  'DO': None,\n",
       "  'MESH': 'MESH:C538270',\n",
       "  'UMLS': 'UMLS:C4551996',\n",
       "  'disease_type': 'disease',\n",
       "  'disease_umlscui': 'C4551996',\n",
       "  'score': 0.7,\n",
       "  'ei': 1.0,\n",
       "  'el': None},\n",
       " {'disease_name': 'Auriculo-condylar syndrome',\n",
       "  'HPO': None,\n",
       "  'NCI': None,\n",
       "  'OMIM': 'MIM:602483',\n",
       "  'MONDO': 'MONDO:0000107',\n",
       "  'ORDO': 'ORDO:137888',\n",
       "  'EFO': None,\n",
       "  'DO': None,\n",
       "  'MESH': 'MESH:C538270',\n",
       "  'UMLS': 'UMLS:C1865295',\n",
       "  'disease_type': 'disease',\n",
       "  'disease_umlscui': 'C1865295',\n",
       "  'score': 0.65,\n",
       "  'ei': 0.8,\n",
       "  'el': None},\n",
       " {'disease_name': 'Amnestic disorders',\n",
       "  'HPO': None,\n",
       "  'NCI': None,\n",
       "  'OMIM': None,\n",
       "  'MONDO': 'MONDO:0001152',\n",
       "  'ORDO': None,\n",
       "  'EFO': None,\n",
       "  'DO': 'DOID:10914',\n",
       "  'MESH': None,\n",
       "  'UMLS': 'UMLS:C0002625',\n",
       "  'disease_type': 'disease',\n",
       "  'disease_umlscui': 'C0002625',\n",
       "  'score': 0.4,\n",
       "  'ei': 1.0,\n",
       "  'el': None},\n",
       " {'disease_name': 'Amnesia',\n",
       "  'HPO': None,\n",
       "  'NCI': 'NCI:C2867',\n",
       "  'OMIM': None,\n",
       "  'MONDO': 'MONDO:0001152',\n",
       "  'ORDO': None,\n",
       "  'EFO': 'EFO:0001072, EFO:1001454',\n",
       "  'DO': 'DOID:10914, DOID:11037',\n",
       "  'MESH': 'MESH:D000647',\n",
       "  'UMLS': 'UMLS:C0002622',\n",
       "  'disease_type': 'disease',\n",
       "  'disease_umlscui': 'C0002622',\n",
       "  'score': 0.4,\n",
       "  'ei': 1.0,\n",
       "  'el': None}]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disgenet_df[DISGENET_DISEASE_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Disease-Compound edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UMLS_C0006142</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000305</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UMLS_C0006142</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0003869</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UMLS_C1458155</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0003869</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UMLS_C2239176</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000182</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UMLS_C2239176</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000762</td>\n",
       "      <td>EFO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      identifier identifier.source       target target.source\n",
       "0  UMLS_C0006142              UMLS  EFO_0000305           EFO\n",
       "1  UMLS_C0006142              UMLS  EFO_0003869           EFO\n",
       "2  UMLS_C1458155              UMLS  EFO_0003869           EFO\n",
       "3  UMLS_C2239176              UMLS  EFO_0000182           EFO\n",
       "4  UMLS_C2239176              UMLS  EFO_0000762           EFO"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Prepare the input to use DISGENET output as seed for OpenTargets\n",
    "disease_mapping_df = create_harmonized_input_file(disgenet_df, DISGENET_DISEASE_COL, \"EFO\", \"UMLS\")\n",
    "disease_mapping_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing diseases-drug interactions: 100%|██████████| 342/342 [00:05<00:00, 68.06it/s]\n",
      "Mapping PubChem: 100%|██████████| 2433/2433 [06:38<00:00,  6.11it/s]\n",
      "c:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\annotators\\opentargets.py:1133: UserWarning: Not all values in column 'adverse_effect' have the correct type '<class 'list'>'.\n",
      "  check_columns_against_constants(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_disease_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UMLS_C0000786</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_1001255</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1276308', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UMLS_C0001175</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0000765</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL704', 'drugbank_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UMLS_C0002103</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0005854</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1201353', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UMLS_C0002171</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0004192</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1200963', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UMLS_C0002940</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0004264</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1076903', 'drugba...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      identifier identifier.source       target target.source  \\\n",
       "0  UMLS_C0000786              UMLS  EFO_1001255           EFO   \n",
       "1  UMLS_C0001175              UMLS  EFO_0000765           EFO   \n",
       "2  UMLS_C0002103              UMLS  EFO_0005854           EFO   \n",
       "3  UMLS_C0002171              UMLS  EFO_0004192           EFO   \n",
       "4  UMLS_C0002940              UMLS  EFO_0004264           EFO   \n",
       "\n",
       "                       OpenTargets_disease_compounds  \n",
       "0  [{'chembl_id': 'CHEMBL:CHEMBL1276308', 'drugba...  \n",
       "1  [{'chembl_id': 'CHEMBL:CHEMBL704', 'drugbank_i...  \n",
       "2  [{'chembl_id': 'CHEMBL:CHEMBL1201353', 'drugba...  \n",
       "3  [{'chembl_id': 'CHEMBL:CHEMBL1200963', 'drugba...  \n",
       "4  [{'chembl_id': 'CHEMBL:CHEMBL1076903', 'drugba...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(\n",
    "    opentargets_disease_compound_df,\n",
    "    opentargets_disease_compound_metadata,\n",
    ") = opentargets.get_disease_compound_interactions(disease_mapping_df)\n",
    "opentargets_disease_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "opentargets_disease_compound_df[OPENTARGETS_DISEASE_COMPOUND_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Gene-Compound edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene-drug interactions: 100%|██████████| 71/71 [00:00<00:00, 318.26it/s]\n",
      "Mapping PubChem: 100%|██████████| 37/37 [00:15<00:00,  2.41it/s]\n",
      "c:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\annotators\\opentargets.py:607: UserWarning: Not all values in column 'adverse_effect_count' have the correct type '<class 'int'>'.\n",
      "  check_columns_against_constants(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_gene_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL1086218', 'drugba...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'chembl_id': nan, 'drugbank_id': nan, 'compo...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source           target target.source  \\\n",
       "0  ENSG00000072080           Ensembl  ENSG00000072080       Ensembl   \n",
       "1  ENSG00000072571           Ensembl  ENSG00000072571       Ensembl   \n",
       "2  ENSG00000073111           Ensembl  ENSG00000073111       Ensembl   \n",
       "3  ENSG00000085563           Ensembl  ENSG00000085563       Ensembl   \n",
       "4  ENSG00000087076           Ensembl  ENSG00000087076       Ensembl   \n",
       "\n",
       "                          OpenTargets_gene_compounds  \n",
       "0  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  \n",
       "1  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  \n",
       "2  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  \n",
       "3  [{'chembl_id': 'CHEMBL:CHEMBL1086218', 'drugba...  \n",
       "4  [{'chembl_id': nan, 'drugbank_id': nan, 'compo...  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_compound_df, opentargets_compound_metadata = opentargets.get_gene_compound_interactions(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "opentargets_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "opentargets_compound_df[OPENTARGETS_GENE_COMPOUND_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 Gene-Pathways edges"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pathways from WikiPathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Querying WikiPathways: 100%|██████████| 4/4 [02:05<00:00, 31.28s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>WikiPathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000934</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>72960</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>14317</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'WP:WP435', 'pathway_label': '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000001313</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>11858</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000001334</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>384061</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000002870</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>17216</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'WP:WP413', 'pathway_label': '...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source  target target.source  \\\n",
       "0  ENSMUSG00000000934           Ensembl   72960     NCBI Gene   \n",
       "1  ENSMUSG00000001155           Ensembl   14317     NCBI Gene   \n",
       "2  ENSMUSG00000001313           Ensembl   11858     NCBI Gene   \n",
       "3  ENSMUSG00000001334           Ensembl  384061     NCBI Gene   \n",
       "4  ENSMUSG00000002870           Ensembl   17216     NCBI Gene   \n",
       "\n",
       "                                        WikiPathways  \n",
       "0  [{'pathway_id': nan, 'pathway_label': nan, 'pa...  \n",
       "1  [{'pathway_id': 'WP:WP435', 'pathway_label': '...  \n",
       "2  [{'pathway_id': nan, 'pathway_label': nan, 'pa...  \n",
       "3  [{'pathway_id': nan, 'pathway_label': nan, 'pa...  \n",
       "4  [{'pathway_id': 'WP:WP413', 'pathway_label': '...  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipathways_df, wikipathways_metadata = wikipathways.get_gene_wikipathways(bridgedb_df=bridgedb_df)\n",
    "wikipathways_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pathway_id': 'WP:WP5224',\n",
       "  'pathway_label': '2q37 copy number variation syndrome',\n",
       "  'pathway_gene_count': 153.0}]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wikipathways_df[WIKIPATHWAYS][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pathways from KEGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>KEGG_pathways</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>114873</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>26358</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00830', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>436059</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00983', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>68312</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00480', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>14317</td>\n",
       "      <td>NCBI Gene</td>\n",
       "      <td>[{'pathway_id': 'path:mmu00340', 'pathway_labe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source  target target.source  \\\n",
       "0  ENSMUSG00000032087           Ensembl  114873     NCBI Gene   \n",
       "1  ENSMUSG00000024747           Ensembl   26358     NCBI Gene   \n",
       "2  ENSMUSG00000091813           Ensembl  436059     NCBI Gene   \n",
       "3  ENSMUSG00000004035           Ensembl   68312     NCBI Gene   \n",
       "4  ENSMUSG00000001155           Ensembl   14317     NCBI Gene   \n",
       "\n",
       "                                       KEGG_pathways  \n",
       "0                                               None  \n",
       "1  [{'pathway_id': 'path:mmu00830', 'pathway_labe...  \n",
       "2  [{'pathway_id': 'path:mmu00983', 'pathway_labe...  \n",
       "3  [{'pathway_id': 'path:mmu00480', 'pathway_labe...  \n",
       "4  [{'pathway_id': 'path:mmu00340', 'pathway_labe...  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df, kegg_metadata = kegg.get_pathways(bridgedb_df)\n",
    "kegg_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'pathway_id': 'path:mmu04015',\n",
       "  'pathway_name': 'Rap1 signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 213,\n",
       "  'compounds': ['C00035', 'C00044', 'C00076', 'C00165', 'C00575']},\n",
       " {'pathway_id': 'path:mmu04022',\n",
       "  'pathway_name': 'cGMP-PKG signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 171,\n",
       "  'compounds': ['C00020',\n",
       "   'C00027',\n",
       "   'C00076',\n",
       "   'C00144',\n",
       "   'C00212',\n",
       "   'C00238',\n",
       "   'C00533',\n",
       "   'C00575',\n",
       "   'C00942',\n",
       "   'C01245']},\n",
       " {'pathway_id': 'path:mmu04024',\n",
       "  'pathway_name': 'cAMP signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 223,\n",
       "  'compounds': ['C00020',\n",
       "   'C00042',\n",
       "   'C00076',\n",
       "   'C00080',\n",
       "   'C00165',\n",
       "   'C00186',\n",
       "   'C00212',\n",
       "   'C00238',\n",
       "   'C00288',\n",
       "   'C00334',\n",
       "   'C00416',\n",
       "   'C00547',\n",
       "   'C00575',\n",
       "   'C00584',\n",
       "   'C00698',\n",
       "   'C00780',\n",
       "   'C00788',\n",
       "   'C01089',\n",
       "   'C01245',\n",
       "   'C01312',\n",
       "   'C01330',\n",
       "   'C01996',\n",
       "   'C03758',\n",
       "   'C20792',\n",
       "   'C20793']},\n",
       " {'pathway_id': 'path:mmu04062',\n",
       "  'pathway_name': 'Chemokine signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 192,\n",
       "  'compounds': ['C00076', 'C00165', 'C00575', 'C01245', 'C05981']},\n",
       " {'pathway_id': 'path:mmu04071',\n",
       "  'pathway_name': 'Sphingolipid signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 125,\n",
       "  'compounds': ['C00065',\n",
       "   'C00154',\n",
       "   'C00165',\n",
       "   'C00195',\n",
       "   'C00212',\n",
       "   'C00319',\n",
       "   'C00346',\n",
       "   'C00533',\n",
       "   'C00550',\n",
       "   'C00836',\n",
       "   'C01245',\n",
       "   'C05981',\n",
       "   'C06124',\n",
       "   'C12126',\n",
       "   'C16845']},\n",
       " {'pathway_id': 'path:mmu04261',\n",
       "  'pathway_name': 'Adrenergic signaling in cardiomyocytes - Mus musculus (house mouse)',\n",
       "  'gene_count': 155,\n",
       "  'compounds': ['C00076',\n",
       "   'C00080',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00547',\n",
       "   'C00575',\n",
       "   'C00788',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C07056']},\n",
       " {'pathway_id': 'path:mmu04360',\n",
       "  'pathway_name': 'Axon guidance - Mus musculus (house mouse)',\n",
       "  'gene_count': 180,\n",
       "  'compounds': ['C00076']},\n",
       " {'pathway_id': 'path:mmu04371',\n",
       "  'pathway_name': 'Apelin signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 138,\n",
       "  'compounds': ['C00076',\n",
       "   'C00080',\n",
       "   'C00165',\n",
       "   'C00290',\n",
       "   'C00533',\n",
       "   'C00575',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C06124']},\n",
       " {'pathway_id': 'path:mmu04540',\n",
       "  'pathway_name': 'Gap junction - Mus musculus (house mouse)',\n",
       "  'gene_count': 85,\n",
       "  'compounds': ['C00025',\n",
       "   'C00076',\n",
       "   'C00165',\n",
       "   'C00533',\n",
       "   'C00547',\n",
       "   'C00575',\n",
       "   'C00681',\n",
       "   'C00780',\n",
       "   'C00942',\n",
       "   'C01245',\n",
       "   'C03758']},\n",
       " {'pathway_id': 'path:mmu04611',\n",
       "  'pathway_name': 'Platelet activation - Mus musculus (house mouse)',\n",
       "  'gene_count': 125,\n",
       "  'compounds': ['C00002',\n",
       "   'C00008',\n",
       "   'C00076',\n",
       "   'C00165',\n",
       "   'C00219',\n",
       "   'C00427',\n",
       "   'C00533',\n",
       "   'C00575',\n",
       "   'C00942',\n",
       "   'C01245',\n",
       "   'C01312',\n",
       "   'C02198',\n",
       "   'C05956',\n",
       "   'C05981']},\n",
       " {'pathway_id': 'path:mmu04670',\n",
       "  'pathway_name': 'Leukocyte transendothelial migration - Mus musculus (house mouse)',\n",
       "  'gene_count': 118,\n",
       "  'compounds': ['C00027', 'C00076', 'C00575', 'C04637']},\n",
       " {'pathway_id': 'path:mmu04713',\n",
       "  'pathway_name': 'Circadian entrainment - Mus musculus (house mouse)',\n",
       "  'gene_count': 98,\n",
       "  'compounds': ['C00025',\n",
       "   'C00076',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00533',\n",
       "   'C00575',\n",
       "   'C00942',\n",
       "   'C01330',\n",
       "   'C01598']},\n",
       " {'pathway_id': 'path:mmu04723',\n",
       "  'pathway_name': 'Retrograde endocannabinoid signaling - Mus musculus (house mouse)',\n",
       "  'gene_count': 152,\n",
       "  'compounds': ['C00025',\n",
       "   'C00076',\n",
       "   'C00116',\n",
       "   'C00157',\n",
       "   'C00165',\n",
       "   'C00189',\n",
       "   'C00219',\n",
       "   'C00238',\n",
       "   'C00334',\n",
       "   'C00350',\n",
       "   'C00427',\n",
       "   'C00575',\n",
       "   'C00689',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C11695',\n",
       "   'C13856',\n",
       "   'C19912',\n",
       "   'C19913']},\n",
       " {'pathway_id': 'path:mmu04724',\n",
       "  'pathway_name': 'Glutamatergic synapse - Mus musculus (house mouse)',\n",
       "  'gene_count': 115,\n",
       "  'compounds': ['C00025',\n",
       "   'C00064',\n",
       "   'C00076',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00575',\n",
       "   'C01245',\n",
       "   'C01330']},\n",
       " {'pathway_id': 'path:mmu04725',\n",
       "  'pathway_name': 'Cholinergic synapse - Mus musculus (house mouse)',\n",
       "  'gene_count': 112,\n",
       "  'compounds': ['C00010',\n",
       "   'C00024',\n",
       "   'C00033',\n",
       "   'C00076',\n",
       "   'C00114',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00575',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C01996',\n",
       "   'C04637']},\n",
       " {'pathway_id': 'path:mmu04726',\n",
       "  'pathway_name': 'Serotonergic synapse - Mus musculus (house mouse)',\n",
       "  'gene_count': 135,\n",
       "  'compounds': ['C00076',\n",
       "   'C00078',\n",
       "   'C00165',\n",
       "   'C00219',\n",
       "   'C00238',\n",
       "   'C00427',\n",
       "   'C00575',\n",
       "   'C00584',\n",
       "   'C00639',\n",
       "   'C00643',\n",
       "   'C00696',\n",
       "   'C00780',\n",
       "   'C00909',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C02165',\n",
       "   'C02166',\n",
       "   'C4',\n",
       "   'C02198',\n",
       "   'C04822',\n",
       "   'C05356',\n",
       "   'C05635',\n",
       "   'C05953',\n",
       "   'C05954',\n",
       "   'C05955',\n",
       "   'C2',\n",
       "   'C05956',\n",
       "   'C05957',\n",
       "   'C05963',\n",
       "   'C05964',\n",
       "   'C05965',\n",
       "   'C05966',\n",
       "   'C14768',\n",
       "   'C14769',\n",
       "   'C14770',\n",
       "   'C14771',\n",
       "   'C14772',\n",
       "   'C14773',\n",
       "   'C14774',\n",
       "   'C14775',\n",
       "   'C14812',\n",
       "   'C14820',\n",
       "   'C14821',\n",
       "   'C14823']},\n",
       " {'pathway_id': 'path:mmu04727',\n",
       "  'pathway_name': 'GABAergic synapse - Mus musculus (house mouse)',\n",
       "  'gene_count': 89,\n",
       "  'compounds': ['C00025',\n",
       "   'C00026',\n",
       "   'C00042',\n",
       "   'C00064',\n",
       "   'C00076',\n",
       "   'C00238',\n",
       "   'C00334',\n",
       "   'C00575',\n",
       "   'C00698']},\n",
       " {'pathway_id': 'path:mmu04728',\n",
       "  'pathway_name': 'Dopaminergic synapse - Mus musculus (house mouse)',\n",
       "  'gene_count': 135,\n",
       "  'compounds': ['C00076',\n",
       "   'C00082',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00355',\n",
       "   'C00575',\n",
       "   'C01161',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C03758',\n",
       "   'C05582',\n",
       "   'C05587']},\n",
       " {'pathway_id': 'path:mmu04730',\n",
       "  'pathway_name': 'Long-term depression - Mus musculus (house mouse)',\n",
       "  'gene_count': 59,\n",
       "  'compounds': ['C00025',\n",
       "   'C00076',\n",
       "   'C00165',\n",
       "   'C00219',\n",
       "   'C00533',\n",
       "   'C00641',\n",
       "   'C00942',\n",
       "   'C01245',\n",
       "   'C01330']},\n",
       " {'pathway_id': 'path:mmu04914',\n",
       "  'pathway_name': 'Progesterone-mediated oocyte maturation - Mus musculus (house mouse)',\n",
       "  'gene_count': 92,\n",
       "  'compounds': ['C00410', 'C00575', 'C05981', 'C11554']},\n",
       " {'pathway_id': 'path:mmu04915',\n",
       "  'pathway_name': 'Estrogen signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 133,\n",
       "  'compounds': ['C00076',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00334',\n",
       "   'C00533',\n",
       "   'C00575',\n",
       "   'C00951',\n",
       "   'C01245']},\n",
       " {'pathway_id': 'path:mmu04916',\n",
       "  'pathway_name': 'Melanogenesis - Mus musculus (house mouse)',\n",
       "  'gene_count': 99,\n",
       "  'compounds': ['C00076', 'C00082', 'C00165', 'C00575', 'C01245', 'C05606']},\n",
       " {'pathway_id': 'path:mmu04921',\n",
       "  'pathway_name': 'Oxytocin signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 152,\n",
       "  'compounds': ['C00076',\n",
       "   'C00165',\n",
       "   'C00219',\n",
       "   'C00238',\n",
       "   'C00427',\n",
       "   'C00533',\n",
       "   'C00575',\n",
       "   'C00584',\n",
       "   'C00639',\n",
       "   'C00942',\n",
       "   'C01245',\n",
       "   'C13050']},\n",
       " {'pathway_id': 'path:mmu04923',\n",
       "  'pathway_name': 'Regulation of lipolysis in adipocytes - Mus musculus (house mouse)',\n",
       "  'gene_count': 56,\n",
       "  'compounds': ['C00020',\n",
       "   'C00116',\n",
       "   'C00162',\n",
       "   'C00165',\n",
       "   'C00212',\n",
       "   'C00219',\n",
       "   'C00422',\n",
       "   'C00547',\n",
       "   'C00575',\n",
       "   'C00584',\n",
       "   'C00788',\n",
       "   'C00942',\n",
       "   'C01885',\n",
       "   'C02140']},\n",
       " {'pathway_id': 'path:mmu04924',\n",
       "  'pathway_name': 'Renin secretion - Mus musculus (house mouse)',\n",
       "  'gene_count': 75,\n",
       "  'compounds': ['C00001',\n",
       "   'C00002',\n",
       "   'C00008',\n",
       "   'C00020',\n",
       "   'C00076',\n",
       "   'C00212',\n",
       "   'C00238',\n",
       "   'C00533',\n",
       "   'C00547',\n",
       "   'C00575',\n",
       "   'C00584',\n",
       "   'C00698',\n",
       "   'C00788',\n",
       "   'C00873',\n",
       "   'C00942',\n",
       "   'C01245',\n",
       "   'C02135']},\n",
       " {'pathway_id': 'path:mmu04926',\n",
       "  'pathway_name': 'Relaxin signaling pathway - Mus musculus (house mouse)',\n",
       "  'gene_count': 129,\n",
       "  'compounds': ['C00076', 'C00165', 'C00533', 'C00575', 'C01245', 'C05981']},\n",
       " {'pathway_id': 'path:mmu04928',\n",
       "  'pathway_name': 'Parathyroid hormone synthesis, secretion and action - Mus musculus (house mouse)',\n",
       "  'gene_count': 115,\n",
       "  'compounds': ['C00009',\n",
       "   'C00020',\n",
       "   'C00076',\n",
       "   'C00165',\n",
       "   'C00575',\n",
       "   'C00698',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C01561',\n",
       "   'C01673',\n",
       "   'C18231']},\n",
       " {'pathway_id': 'path:mmu04934',\n",
       "  'pathway_name': 'Cushing syndrome - Mus musculus (house mouse)',\n",
       "  'gene_count': 161,\n",
       "  'compounds': ['C00020',\n",
       "   'C00076',\n",
       "   'C00187',\n",
       "   'C00238',\n",
       "   'C00410',\n",
       "   'C00575',\n",
       "   'C00735',\n",
       "   'C01176',\n",
       "   'C01245',\n",
       "   'C01953',\n",
       "   'C05138',\n",
       "   'C05488',\n",
       "   'C07557']},\n",
       " {'pathway_id': 'path:mmu04935',\n",
       "  'pathway_name': 'Growth hormone synthesis, secretion and action - Mus musculus (house mouse)',\n",
       "  'gene_count': 116,\n",
       "  'compounds': ['C00076', 'C00165', 'C00575', 'C01245']},\n",
       " {'pathway_id': 'path:mmu04971',\n",
       "  'pathway_name': 'Gastric acid secretion - Mus musculus (house mouse)',\n",
       "  'gene_count': 74,\n",
       "  'compounds': ['C00001',\n",
       "   'C00011',\n",
       "   'C00076',\n",
       "   'C00080',\n",
       "   'C00165',\n",
       "   'C00238',\n",
       "   'C00288',\n",
       "   'C00388',\n",
       "   'C00575',\n",
       "   'C00698',\n",
       "   'C01245',\n",
       "   'C01330',\n",
       "   'C01353',\n",
       "   'C01996']},\n",
       " {'pathway_id': 'path:mmu05012',\n",
       "  'pathway_name': 'Parkinson disease - Mus musculus (house mouse)',\n",
       "  'gene_count': 273,\n",
       "  'compounds': ['C00002',\n",
       "   'C00007',\n",
       "   'C00008',\n",
       "   'C00009',\n",
       "   'C00013',\n",
       "   'C00020',\n",
       "   'C00027',\n",
       "   'C00034',\n",
       "   'C00076',\n",
       "   'C00082',\n",
       "   'C00165',\n",
       "   'C00212',\n",
       "   'C00355',\n",
       "   'C00575',\n",
       "   'C00704',\n",
       "   'C01245',\n",
       "   'C03758',\n",
       "   'C04043',\n",
       "   'C04599',\n",
       "   'C07593',\n",
       "   'C11310',\n",
       "   'C14701',\n",
       "   'C15231',\n",
       "   'C16844',\n",
       "   'C17755',\n",
       "   'C22381']},\n",
       " {'pathway_id': 'path:mmu05030',\n",
       "  'pathway_name': 'Cocaine addiction - Mus musculus (house mouse)',\n",
       "  'gene_count': 47,\n",
       "  'compounds': ['C00025',\n",
       "   'C00027',\n",
       "   'C00082',\n",
       "   'C00355',\n",
       "   'C00575',\n",
       "   'C01161',\n",
       "   'C03758']},\n",
       " {'pathway_id': 'path:mmu05032',\n",
       "  'pathway_name': 'Morphine addiction - Mus musculus (house mouse)',\n",
       "  'gene_count': 91,\n",
       "  'compounds': ['C00020',\n",
       "   'C00076',\n",
       "   'C00212',\n",
       "   'C00238',\n",
       "   'C00334',\n",
       "   'C00575',\n",
       "   'C00698',\n",
       "   'C03758']},\n",
       " {'pathway_id': 'path:mmu05034',\n",
       "  'pathway_name': 'Alcoholism - Mus musculus (house mouse)',\n",
       "  'gene_count': 206,\n",
       "  'compounds': ['C00025',\n",
       "   'C00027',\n",
       "   'C00076',\n",
       "   'C00082',\n",
       "   'C00212',\n",
       "   'C00355',\n",
       "   'C00469',\n",
       "   'C00575',\n",
       "   'C01161',\n",
       "   'C03758']},\n",
       " {'pathway_id': 'path:mmu05133',\n",
       "  'pathway_name': 'Pertussis - Mus musculus (house mouse)',\n",
       "  'gene_count': 76,\n",
       "  'compounds': ['C00002',\n",
       "   'C00076',\n",
       "   'C00238',\n",
       "   'C00253',\n",
       "   'C00338',\n",
       "   'C00374',\n",
       "   'C00575',\n",
       "   'C05005',\n",
       "   'C12505',\n",
       "   'C19596']},\n",
       " {'pathway_id': 'path:mmu05142',\n",
       "  'pathway_name': 'Chagas disease - Mus musculus (house mouse)',\n",
       "  'gene_count': 102,\n",
       "  'compounds': ['C00062', 'C00076', 'C00533', 'C00575', 'C01245', 'C05981']},\n",
       " {'pathway_id': 'path:mmu05145',\n",
       "  'pathway_name': 'Toxoplasmosis - Mus musculus (house mouse)',\n",
       "  'gene_count': 108,\n",
       "  'compounds': ['C00076', 'C05981', 'C06314']},\n",
       " {'pathway_id': 'path:mmu05163',\n",
       "  'pathway_name': 'Human cytomegalovirus infection - Mus musculus (house mouse)',\n",
       "  'gene_count': 253,\n",
       "  'compounds': ['C00076', 'C00165', 'C00584', 'C01245', 'C05981', 'C20640']},\n",
       " {'pathway_id': 'path:mmu05170',\n",
       "  'pathway_name': 'Human immunodeficiency virus 1 infection - Mus musculus (house mouse)',\n",
       "  'gene_count': 240,\n",
       "  'compounds': ['C00076', 'C00165', 'C01245', 'C05981', 'C20640']},\n",
       " {'pathway_id': 'path:mmu05200',\n",
       "  'pathway_name': 'Pathways in cancer - Mus musculus (house mouse)',\n",
       "  'gene_count': 547,\n",
       "  'compounds': ['C00027',\n",
       "   'C00039',\n",
       "   'C00076',\n",
       "   'C00122',\n",
       "   'C00149',\n",
       "   'C00165',\n",
       "   'C00187',\n",
       "   'C00280',\n",
       "   'C00410',\n",
       "   'C00416',\n",
       "   'C00533',\n",
       "   'C00535',\n",
       "   'C00575',\n",
       "   'C00584',\n",
       "   'C00681',\n",
       "   'C00704',\n",
       "   'C00735',\n",
       "   'C00762',\n",
       "   'C00777',\n",
       "   'C00951',\n",
       "   'C01227',\n",
       "   'C01245',\n",
       "   'C01471',\n",
       "   'C03917',\n",
       "   'C05981',\n",
       "   'C07653',\n",
       "   'C15493',\n",
       "   'C16038',\n",
       "   'C16844',\n",
       "   'C21641',\n",
       "   'C21642']},\n",
       " {'pathway_id': 'path:mmu05207',\n",
       "  'pathway_name': 'Chemical carcinogenesis - receptor activation - Mus musculus (house mouse)',\n",
       "  'gene_count': 227,\n",
       "  'compounds': ['C00076',\n",
       "   'C00410',\n",
       "   'C00547',\n",
       "   'C00575',\n",
       "   'C00745',\n",
       "   'C00780',\n",
       "   'C00788',\n",
       "   'C00951',\n",
       "   'C03690',\n",
       "   'C04623',\n",
       "   'C05981',\n",
       "   'C06697',\n",
       "   'C06790',\n",
       "   'C06916',\n",
       "   'C07020',\n",
       "   'C07434',\n",
       "   'C07535',\n",
       "   'C07557',\n",
       "   'C08150',\n",
       "   'C11042',\n",
       "   'C11150',\n",
       "   'C11371',\n",
       "   'C11906',\n",
       "   'C13624',\n",
       "   'C14240',\n",
       "   'C14573',\n",
       "   'C15995',\n",
       "   'C16452',\n",
       "   'C16453']}]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_df[KEGG_COL][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reactome pathways from OpenTargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene-pathway interactions: 100%|██████████| 71/71 [00:00<00:00, 571.39it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_22080\\442422480.py:1: UserWarning: The intermediate_df in OpenTargets_reactome annotatur should be checked, please create an issue on https://github.com/BioDataFuse/pyBiodatafuse/issues/.\n",
      "  opentargets_reactome_df, opentargets_reactome_metadata = opentargets.get_gene_reactome_pathways(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_reactome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Platelet degranulation ', ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Hyaluronan uptake and degr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Orc1 removal from chromati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Abacavir transmembrane tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_label': 'Estrogen biosynthesis', 'p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source           target target.source  \\\n",
       "0  ENSG00000072080           Ensembl  ENSG00000072080       Ensembl   \n",
       "1  ENSG00000072571           Ensembl  ENSG00000072571       Ensembl   \n",
       "2  ENSG00000073111           Ensembl  ENSG00000073111       Ensembl   \n",
       "3  ENSG00000085563           Ensembl  ENSG00000085563       Ensembl   \n",
       "4  ENSG00000087076           Ensembl  ENSG00000087076       Ensembl   \n",
       "\n",
       "                                OpenTargets_reactome  \n",
       "0  [{'pathway_label': 'Platelet degranulation ', ...  \n",
       "1  [{'pathway_label': 'Hyaluronan uptake and degr...  \n",
       "2  [{'pathway_label': 'Orc1 removal from chromati...  \n",
       "3  [{'pathway_label': 'Abacavir transmembrane tra...  \n",
       "4  [{'pathway_label': 'Estrogen biosynthesis', 'p...  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_reactome_df, opentargets_reactome_metadata = opentargets.get_gene_reactome_pathways(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "opentargets_reactome_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "opentargets_reactome_df[OPENTARGETS_REACTOME_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Gene Ontology from OpenTargets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing gene annotation: 100%|██████████| 71/71 [00:00<00:00, 537.84it/s]\n",
      "C:\\Users\\are10\\AppData\\Local\\Temp\\ipykernel_22080\\517848960.py:1: UserWarning: The intermediate_df in OpenTargets_go annotatur should be checked, please create an issue on https://github.com/BioDataFuse/pyBiodatafuse/issues/.\n",
      "  opentargets_go_df, opentargets_go_metadata = opentargets.get_gene_go_process(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_go</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005788', 'go_name': 'endoplasm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005515', 'go_name': 'protein b...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000073111</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005654', 'go_name': 'nucleopla...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000085563</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0008559', 'go_name': 'ABC-type ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSG00000087076</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'go_id': 'GO:0005515', 'go_name': 'protein b...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source           target target.source  \\\n",
       "0  ENSG00000072080           Ensembl  ENSG00000072080       Ensembl   \n",
       "1  ENSG00000072571           Ensembl  ENSG00000072571       Ensembl   \n",
       "2  ENSG00000073111           Ensembl  ENSG00000073111       Ensembl   \n",
       "3  ENSG00000085563           Ensembl  ENSG00000085563       Ensembl   \n",
       "4  ENSG00000087076           Ensembl  ENSG00000087076       Ensembl   \n",
       "\n",
       "                                      OpenTargets_go  \n",
       "0  [{'go_id': 'GO:0005788', 'go_name': 'endoplasm...  \n",
       "1  [{'go_id': 'GO:0005515', 'go_name': 'protein b...  \n",
       "2  [{'go_id': 'GO:0005654', 'go_name': 'nucleopla...  \n",
       "3  [{'go_id': 'GO:0008559', 'go_name': 'ABC-type ...  \n",
       "4  [{'go_id': 'GO:0005515', 'go_name': 'protein b...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_go_df, opentargets_go_metadata = opentargets.get_gene_go_process(\n",
    "    bridgedb_df=bridgedb_df_hl\n",
    ")\n",
    "opentargets_go_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'go_id': 'GO:0005788',\n",
       "  'go_name': 'endoplasmic reticulum lumen',\n",
       "  'go_type': 'C'},\n",
       " {'go_id': 'GO:0031089',\n",
       "  'go_name': 'platelet dense granule lumen',\n",
       "  'go_type': 'C'},\n",
       " {'go_id': 'GO:0004866',\n",
       "  'go_name': 'endopeptidase inhibitor activity',\n",
       "  'go_type': 'F'},\n",
       " {'go_id': 'GO:0046849', 'go_name': 'bone remodeling', 'go_type': 'P'},\n",
       " {'go_id': 'GO:0062023',\n",
       "  'go_name': 'collagen-containing extracellular matrix',\n",
       "  'go_type': 'C'},\n",
       " {'go_id': 'GO:0005576', 'go_name': 'extracellular region', 'go_type': 'C'},\n",
       " {'go_id': 'GO:0001501',\n",
       "  'go_name': 'skeletal system development',\n",
       "  'go_type': 'P'}]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_go_df[OPENTARGETS_GO_COL][0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.6. Protein-Protein Interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>StringDB_ppi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000027452', 'E...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000100426           Ensembl  ENSMUSG00000100426       Ensembl   \n",
       "1  ENSMUSG00000032087           Ensembl  ENSMUSG00000032087       Ensembl   \n",
       "2  ENSMUSG00000024747           Ensembl  ENSMUSG00000024747       Ensembl   \n",
       "3  ENSMUSG00000091813           Ensembl  ENSMUSG00000091813       Ensembl   \n",
       "4  ENSMUSG00000004035           Ensembl  ENSMUSG00000004035       Ensembl   \n",
       "\n",
       "                                        StringDB_ppi  \n",
       "0  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...  \n",
       "1  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...  \n",
       "2  [{'stringdb_link_to': 'ENSMUSG00000027452', 'E...  \n",
       "3  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...  \n",
       "4  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...  "
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_species = \"Mouse\"\n",
    "ppi_df, ppi_metadata = stringdb.get_ppi(bridgedb_df=bridgedb_df, species=input_species)\n",
    "ppi_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'stringdb_link_to': 'Rps2',\n",
       "  'Ensembl': 'Ensembl:ENSMUSP00000092502',\n",
       "  'score': 0.902,\n",
       "  'Uniprot-TrEMBL': 'Atp5a1'},\n",
       " {'stringdb_link_to': 'Rplp0',\n",
       "  'Ensembl': 'Ensembl:ENSMUSP00000083705',\n",
       "  'score': 0.903,\n",
       "  'Uniprot-TrEMBL': 'Atp5a1'}]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ppi_df[STRING_PPI_COL][3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.7 Gene expression edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[19], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m bgee_df, bgee_metadata \u001b[38;5;241m=\u001b[39m \u001b[43mbgee\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_gene_expression\u001b[49m\u001b[43m(\u001b[49m\u001b[43mbridgedb_df\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbridgedb_df\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      2\u001b[0m bgee_df\u001b[38;5;241m.\u001b[39mhead()\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\src\\pyBiodatafuse\\annotators\\bgee.py:145\u001b[0m, in \u001b[0;36mget_gene_expression\u001b[1;34m(bridgedb_df)\u001b[0m\n\u001b[0;32m    142\u001b[0m sparql_query_template_sub \u001b[38;5;241m=\u001b[39m sparql_query_template\u001b[38;5;241m.\u001b[39msubstitute(substit_dict)\n\u001b[0;32m    144\u001b[0m sparql\u001b[38;5;241m.\u001b[39msetQuery(sparql_query_template_sub)\n\u001b[1;32m--> 145\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[43msparql\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mqueryAndConvert\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    147\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(res[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mresults\u001b[39m\u001b[38;5;124m\"\u001b[39m][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbindings\u001b[39m\u001b[38;5;124m\"\u001b[39m])\n\u001b[0;32m    149\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m col \u001b[38;5;129;01min\u001b[39;00m df:\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\SPARQLWrapper\\Wrapper.py:967\u001b[0m, in \u001b[0;36mSPARQLWrapper.queryAndConvert\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    962\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mqueryAndConvert\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mQueryResult.ConvertResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    963\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Macro like method: issue a query and return the converted results.\u001b[39;00m\n\u001b[0;32m    964\u001b[0m \n\u001b[0;32m    965\u001b[0m \u001b[38;5;124;03m    :return: the converted query result. See the conversion methods for more details.\u001b[39;00m\n\u001b[0;32m    966\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 967\u001b[0m     res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mquery\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    968\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m res\u001b[38;5;241m.\u001b[39mconvert()\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\SPARQLWrapper\\Wrapper.py:960\u001b[0m, in \u001b[0;36mSPARQLWrapper.query\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    942\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mquery\u001b[39m(\u001b[38;5;28mself\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mQueryResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[0;32m    943\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m    944\u001b[0m \u001b[38;5;124;03m    Execute the query.\u001b[39;00m\n\u001b[0;32m    945\u001b[0m \u001b[38;5;124;03m    Exceptions can be raised if either the URI is wrong or the HTTP sends back an error (this is also the\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m    958\u001b[0m \u001b[38;5;124;03m    :rtype: :class:`QueryResult` instance\u001b[39;00m\n\u001b[0;32m    959\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m--> 960\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m QueryResult(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_query\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[1;32mc:\\Users\\are10\\Documents\\BAFSTU\\code\\pyBioDatafusemain\\pyBiodatafuse\\.venv\\lib\\site-packages\\SPARQLWrapper\\Wrapper.py:926\u001b[0m, in \u001b[0;36mSPARQLWrapper._query\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    924\u001b[0m         response \u001b[38;5;241m=\u001b[39m urlopener(request, timeout\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout)\n\u001b[0;32m    925\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m--> 926\u001b[0m         response \u001b[38;5;241m=\u001b[39m \u001b[43murlopener\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    927\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m response, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturnFormat\n\u001b[0;32m    928\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m urllib\u001b[38;5;241m.\u001b[39merror\u001b[38;5;241m.\u001b[39mHTTPError \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:214\u001b[0m, in \u001b[0;36murlopen\u001b[1;34m(url, data, timeout, cafile, capath, cadefault, context)\u001b[0m\n\u001b[0;32m    212\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m    213\u001b[0m     opener \u001b[38;5;241m=\u001b[39m _opener\n\u001b[1;32m--> 214\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mopener\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mopen\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:517\u001b[0m, in \u001b[0;36mOpenerDirector.open\u001b[1;34m(self, fullurl, data, timeout)\u001b[0m\n\u001b[0;32m    514\u001b[0m     req \u001b[38;5;241m=\u001b[39m meth(req)\n\u001b[0;32m    516\u001b[0m sys\u001b[38;5;241m.\u001b[39maudit(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124murllib.Request\u001b[39m\u001b[38;5;124m'\u001b[39m, req\u001b[38;5;241m.\u001b[39mfull_url, req\u001b[38;5;241m.\u001b[39mdata, req\u001b[38;5;241m.\u001b[39mheaders, req\u001b[38;5;241m.\u001b[39mget_method())\n\u001b[1;32m--> 517\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    519\u001b[0m \u001b[38;5;66;03m# post-process response\u001b[39;00m\n\u001b[0;32m    520\u001b[0m meth_name \u001b[38;5;241m=\u001b[39m protocol\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_response\u001b[39m\u001b[38;5;124m\"\u001b[39m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:534\u001b[0m, in \u001b[0;36mOpenerDirector._open\u001b[1;34m(self, req, data)\u001b[0m\n\u001b[0;32m    531\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n\u001b[0;32m    533\u001b[0m protocol \u001b[38;5;241m=\u001b[39m req\u001b[38;5;241m.\u001b[39mtype\n\u001b[1;32m--> 534\u001b[0m result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_call_chain\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mhandle_open\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprotocol\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\n\u001b[0;32m    535\u001b[0m \u001b[43m                          \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_open\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    536\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m result:\n\u001b[0;32m    537\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:494\u001b[0m, in \u001b[0;36mOpenerDirector._call_chain\u001b[1;34m(self, chain, kind, meth_name, *args)\u001b[0m\n\u001b[0;32m    492\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m handler \u001b[38;5;129;01min\u001b[39;00m handlers:\n\u001b[0;32m    493\u001b[0m     func \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(handler, meth_name)\n\u001b[1;32m--> 494\u001b[0m     result \u001b[38;5;241m=\u001b[39m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    495\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    496\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m result\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:1389\u001b[0m, in \u001b[0;36mHTTPSHandler.https_open\u001b[1;34m(self, req)\u001b[0m\n\u001b[0;32m   1388\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mhttps_open\u001b[39m(\u001b[38;5;28mself\u001b[39m, req):\n\u001b[1;32m-> 1389\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_open\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhttp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mclient\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mHTTPSConnection\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m   1390\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_context\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_check_hostname\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\urllib\\request.py:1350\u001b[0m, in \u001b[0;36mAbstractHTTPHandler.do_open\u001b[1;34m(self, http_class, req, **http_conn_args)\u001b[0m\n\u001b[0;32m   1348\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err: \u001b[38;5;66;03m# timeout error\u001b[39;00m\n\u001b[0;32m   1349\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m URLError(err)\n\u001b[1;32m-> 1350\u001b[0m     r \u001b[38;5;241m=\u001b[39m \u001b[43mh\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresponse\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1351\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:\n\u001b[0;32m   1352\u001b[0m     h\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\http\\client.py:1377\u001b[0m, in \u001b[0;36mHTTPConnection.getresponse\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1375\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1376\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 1377\u001b[0m         \u001b[43mresponse\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbegin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1378\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mConnectionError\u001b[39;00m:\n\u001b[0;32m   1379\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\http\\client.py:320\u001b[0m, in \u001b[0;36mHTTPResponse.begin\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    318\u001b[0m \u001b[38;5;66;03m# read until we get a non-100 response\u001b[39;00m\n\u001b[0;32m    319\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m--> 320\u001b[0m     version, status, reason \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_read_status\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    321\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m status \u001b[38;5;241m!=\u001b[39m CONTINUE:\n\u001b[0;32m    322\u001b[0m         \u001b[38;5;28;01mbreak\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\http\\client.py:281\u001b[0m, in \u001b[0;36mHTTPResponse._read_status\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    280\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_read_status\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m--> 281\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mreadline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m_MAXLINE\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124miso-8859-1\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m    282\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(line) \u001b[38;5;241m>\u001b[39m _MAXLINE:\n\u001b[0;32m    283\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m LineTooLong(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstatus line\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\socket.py:704\u001b[0m, in \u001b[0;36mSocketIO.readinto\u001b[1;34m(self, b)\u001b[0m\n\u001b[0;32m    702\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[0;32m    703\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m--> 704\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrecv_into\u001b[49m\u001b[43m(\u001b[49m\u001b[43mb\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m    705\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m timeout:\n\u001b[0;32m    706\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_timeout_occurred \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\ssl.py:1242\u001b[0m, in \u001b[0;36mSSLSocket.recv_into\u001b[1;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[0;32m   1238\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m flags \u001b[38;5;241m!=\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m   1239\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[0;32m   1240\u001b[0m           \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnon-zero flags not allowed in calls to recv_into() on \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m\n\u001b[0;32m   1241\u001b[0m           \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__class__\u001b[39m)\n\u001b[1;32m-> 1242\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnbytes\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1243\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1244\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28msuper\u001b[39m()\u001b[38;5;241m.\u001b[39mrecv_into(buffer, nbytes, flags)\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python39\\lib\\ssl.py:1100\u001b[0m, in \u001b[0;36mSSLSocket.read\u001b[1;34m(self, len, buffer)\u001b[0m\n\u001b[0;32m   1098\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m   1099\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m buffer \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m-> 1100\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mlen\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mbuffer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m   1101\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m   1102\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_sslobj\u001b[38;5;241m.\u001b[39mread(\u001b[38;5;28mlen\u001b[39m)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "bgee_df, bgee_metadata = bgee.get_gene_expression(bridgedb_df=bridgedb_df)\n",
    "bgee_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'anatomical_entity_id': 'UBERON_0002371',\n",
       "  'anatomical_entity_name': 'bone marrow',\n",
       "  'expression_level': 98.45455,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0000955',\n",
       "  'anatomical_entity_name': 'brain',\n",
       "  'expression_level': 99.51974,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0004535',\n",
       "  'anatomical_entity_name': 'cardiovascular system',\n",
       "  'expression_level': 99.83003,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001007',\n",
       "  'anatomical_entity_name': 'digestive system',\n",
       "  'expression_level': 99.18479,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0000948',\n",
       "  'anatomical_entity_name': 'heart',\n",
       "  'expression_level': 99.83252,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0005057',\n",
       "  'anatomical_entity_name': 'immune organ',\n",
       "  'expression_level': 98.96394,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0002113',\n",
       "  'anatomical_entity_name': 'kidney',\n",
       "  'expression_level': 99.73472,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0002107',\n",
       "  'anatomical_entity_name': 'liver',\n",
       "  'expression_level': 99.07897,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0002048',\n",
       "  'anatomical_entity_name': 'lung',\n",
       "  'expression_level': 98.90879,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001016',\n",
       "  'anatomical_entity_name': 'nervous system',\n",
       "  'expression_level': 99.5071,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001264',\n",
       "  'anatomical_entity_name': 'pancreas',\n",
       "  'expression_level': 98.99588000000001,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001987',\n",
       "  'anatomical_entity_name': 'placenta',\n",
       "  'expression_level': 98.83242000000001,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0000990',\n",
       "  'anatomical_entity_name': 'reproductive system',\n",
       "  'expression_level': 97.54951,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001004',\n",
       "  'anatomical_entity_name': 'respiratory system',\n",
       "  'expression_level': 98.90427,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'},\n",
       " {'anatomical_entity_id': 'UBERON_0001434',\n",
       "  'anatomical_entity_name': 'skeletal system',\n",
       "  'expression_level': 98.01689,\n",
       "  'confidence_level_id': 'CIO_0000029',\n",
       "  'confidence_level_name': 'high confidence level',\n",
       "  'developmental_stage_id': 'UBERON_0000104',\n",
       "  'developmental_stage_name': 'life cycle'}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bgee_df[BGEE_GENE_EXPRESSION_LEVELS_COL][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.8 Transporter Inhibitors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>MolMeDB_transporter_inhibitor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>C9J6K0</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "      <td>[{'compound_name': nan, 'inchikey': nan, 'smil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000072080</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>Q13103</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "      <td>[{'compound_name': nan, 'inchikey': nan, 'smil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E5RI30</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "      <td>[{'compound_name': nan, 'inchikey': nan, 'smil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>E5RIH2</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "      <td>[{'compound_name': nan, 'inchikey': nan, 'smil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000072571</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>O75330</td>\n",
       "      <td>Uniprot-TrEMBL</td>\n",
       "      <td>[{'compound_name': nan, 'inchikey': nan, 'smil...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        identifier identifier.source  target   target.source  \\\n",
       "0  ENSG00000072080           Ensembl  C9J6K0  Uniprot-TrEMBL   \n",
       "1  ENSG00000072080           Ensembl  Q13103  Uniprot-TrEMBL   \n",
       "2  ENSG00000072571           Ensembl  E5RI30  Uniprot-TrEMBL   \n",
       "3  ENSG00000072571           Ensembl  E5RIH2  Uniprot-TrEMBL   \n",
       "4  ENSG00000072571           Ensembl  O75330  Uniprot-TrEMBL   \n",
       "\n",
       "                       MolMeDB_transporter_inhibitor  \n",
       "0  [{'compound_name': nan, 'inchikey': nan, 'smil...  \n",
       "1  [{'compound_name': nan, 'inchikey': nan, 'smil...  \n",
       "2  [{'compound_name': nan, 'inchikey': nan, 'smil...  \n",
       "3  [{'compound_name': nan, 'inchikey': nan, 'smil...  \n",
       "4  [{'compound_name': nan, 'inchikey': nan, 'smil...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inhibitor_df, inhibitor_metadata = molmedb.get_gene_compound_inhibitor(bridgedb_df=bridgedb_df_hl)\n",
    "inhibitor_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'compound_name': nan, 'inchikey': nan, 'smiles': nan, 'compound_cid': nan, 'molmedb_id': nan, 'source_pmid': nan, 'chebi_id': nan, 'drugbank_id': nan, 'uniprot_trembl_id': nan}]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>MolMeDB_transporter_inhibitor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [identifier, identifier.source, target, target.source, MolMeDB_transporter_inhibitor]\n",
       "Index: []"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(inhibitor_df[MOLMEDB_PROTEIN_COMPOUND_COL][47])\n",
    "inhibitor_df[inhibitor_df[\"target\"] == \"P35499\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Generating Graph"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Combing all the results into single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>WikiPathways</th>\n",
       "      <th>StringDB_ppi</th>\n",
       "      <th>Ensembl_homologs</th>\n",
       "      <th>OpenTargets_go</th>\n",
       "      <th>DISGENET_diseases</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000100426</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>[{'homolog': 'ENSG00000177103'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000032087</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000027452', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000165092'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000024747</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000091813</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>[{'pathway_id': 'WP:WP4466', 'pathway_label': ...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>[{'homolog': 'ENSG00000213366'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000004035</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>[{'pathway_id': 'WP:WP435', 'pathway_label': '...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000028356', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000160282'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000001155</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000026205', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000146039'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000021336</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000006398', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000128944'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000027331</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000006398', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000134690'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000028873</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "      <td>[{'pathway_id': nan, 'pathway_label': nan, 'pa...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000022415', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000159164'}]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000038486</td>\n",
       "      <td>Ensembl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier                                       WikiPathways  \\\n",
       "0  ENSMUSG00000100426                                                NaN   \n",
       "1  ENSMUSG00000032087  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "2  ENSMUSG00000024747  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "3  ENSMUSG00000091813  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "4  ENSMUSG00000004035  [{'pathway_id': 'WP:WP4466', 'pathway_label': ...   \n",
       "5  ENSMUSG00000001155  [{'pathway_id': 'WP:WP435', 'pathway_label': '...   \n",
       "6  ENSMUSG00000021336  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "7  ENSMUSG00000027331  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "8  ENSMUSG00000028873  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "9  ENSMUSG00000038486  [{'pathway_id': nan, 'pathway_label': nan, 'pa...   \n",
       "\n",
       "                                        StringDB_ppi  \\\n",
       "0  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "1  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "2  [{'stringdb_link_to': 'ENSMUSG00000027452', 'E...   \n",
       "3  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "4  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "5  [{'stringdb_link_to': 'ENSMUSG00000028356', 'E...   \n",
       "6  [{'stringdb_link_to': 'ENSMUSG00000026205', 'E...   \n",
       "7  [{'stringdb_link_to': 'ENSMUSG00000006398', 'E...   \n",
       "8  [{'stringdb_link_to': 'ENSMUSG00000006398', 'E...   \n",
       "9  [{'stringdb_link_to': 'ENSMUSG00000022415', 'E...   \n",
       "\n",
       "                   Ensembl_homologs OpenTargets_go DISGENET_diseases  \\\n",
       "0                              None            NaN               NaN   \n",
       "1  [{'homolog': 'ENSG00000177103'}]            NaN               NaN   \n",
       "2  [{'homolog': 'ENSG00000165092'}]            NaN               NaN   \n",
       "3                              None            NaN               NaN   \n",
       "4  [{'homolog': 'ENSG00000213366'}]            NaN               NaN   \n",
       "5  [{'homolog': 'ENSG00000160282'}]            NaN               NaN   \n",
       "6  [{'homolog': 'ENSG00000146039'}]            NaN               NaN   \n",
       "7  [{'homolog': 'ENSG00000128944'}]            NaN               NaN   \n",
       "8  [{'homolog': 'ENSG00000134690'}]            NaN               NaN   \n",
       "9  [{'homolog': 'ENSG00000159164'}]            NaN               NaN   \n",
       "\n",
       "  identifier.source              target target.source  \n",
       "0           Ensembl  ENSMUSG00000100426       Ensembl  \n",
       "1           Ensembl  ENSMUSG00000032087       Ensembl  \n",
       "2           Ensembl  ENSMUSG00000024747       Ensembl  \n",
       "3           Ensembl  ENSMUSG00000091813       Ensembl  \n",
       "4           Ensembl  ENSMUSG00000004035       Ensembl  \n",
       "5           Ensembl  ENSMUSG00000001155       Ensembl  \n",
       "6           Ensembl  ENSMUSG00000021336       Ensembl  \n",
       "7           Ensembl  ENSMUSG00000027331       Ensembl  \n",
       "8           Ensembl  ENSMUSG00000028873       Ensembl  \n",
       "9           Ensembl  ENSMUSG00000038486       Ensembl  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#        bgee_df,\n",
    "#        disgenet_df,\n",
    "#        minerva_df,\n",
    "#        opentargets_reactome_df,\n",
    "#        opentargets_go_df,\n",
    "#        opentargets_compound_df,\n",
    "#        inhibitor_df,\n",
    "#        kegg_df,\n",
    "\n",
    "combined_df = combine_sources(\n",
    "    bridgedb_df,\n",
    "    [\n",
    "        wikipathways_df,\n",
    "        ppi_df,\n",
    "        ensembl_homologs_df\n",
    "    ],\n",
    ")\n",
    "\n",
    "combined_df_hl = combine_sources(\n",
    "    bridgedb_df_hl,\n",
    "    [\n",
    "        opentargets_go_df,\n",
    "        disgenet_df,\n",
    "    ],\n",
    ")\n",
    "\n",
    "combined_df_hl.head(10)\n",
    "\n",
    "merged_df = pd.merge(combined_df, combined_df_hl, on='identifier', how='outer')\n",
    "\n",
    "for col in merged_df.columns:\n",
    "    if col.endswith('_x'):\n",
    "        base_col = col[:-2] \n",
    "        merged_df[base_col] = merged_df[col].combine_first(merged_df.get(f'{base_col}_y'))\n",
    "        merged_df.drop([col, f'{base_col}_y'], axis=1, inplace=True)\n",
    "\n",
    "merged_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 9)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Exporting the database in pickle format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"combined_df.pkl\", \"wb\") as out:\n",
    "    pickle.dump(merged_df, out)\n",
    "# with open(\"opentargets_disease_compound_df.pkl\", \"wb\") as out:\n",
    "#     pickle.dump(opentargets_disease_compound_df, out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Creating a graph from the annotated dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>WikiPathways</th>\n",
       "      <th>StringDB_ppi</th>\n",
       "      <th>Ensembl_homologs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000067274</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'WP:WP163', 'pathway_label': '...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000025428', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000089157'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'WP:WP2292', 'pathway_label': ...</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>[{'homolog': 'ENSG00000065135'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000084349</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000084349</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000025428</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000025428</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'WP:WP295', 'pathway_label': '...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000044533', 'E...</td>\n",
       "      <td>[{'homolog': 'ENSG00000152234'}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000044533</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>ENSMUSG00000044533</td>\n",
       "      <td>Ensembl</td>\n",
       "      <td>[{'pathway_id': 'WP:WP163', 'pathway_label': '...</td>\n",
       "      <td>[{'stringdb_link_to': 'ENSMUSG00000025428', 'E...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           identifier identifier.source              target target.source  \\\n",
       "0  ENSMUSG00000067274           Ensembl  ENSMUSG00000067274       Ensembl   \n",
       "1  ENSMUSG00000000001           Ensembl  ENSMUSG00000000001       Ensembl   \n",
       "2  ENSMUSG00000084349           Ensembl  ENSMUSG00000084349       Ensembl   \n",
       "3  ENSMUSG00000025428           Ensembl  ENSMUSG00000025428       Ensembl   \n",
       "4  ENSMUSG00000044533           Ensembl  ENSMUSG00000044533       Ensembl   \n",
       "\n",
       "                                        WikiPathways  \\\n",
       "0  [{'pathway_id': 'WP:WP163', 'pathway_label': '...   \n",
       "1  [{'pathway_id': 'WP:WP2292', 'pathway_label': ...   \n",
       "2                                                NaN   \n",
       "3  [{'pathway_id': 'WP:WP295', 'pathway_label': '...   \n",
       "4  [{'pathway_id': 'WP:WP163', 'pathway_label': '...   \n",
       "\n",
       "                                        StringDB_ppi  \\\n",
       "0  [{'stringdb_link_to': 'ENSMUSG00000025428', 'E...   \n",
       "1  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "2  [{'stringdb_link_to': nan, 'Ensembl': nan, 'sc...   \n",
       "3  [{'stringdb_link_to': 'ENSMUSG00000044533', 'E...   \n",
       "4  [{'stringdb_link_to': 'ENSMUSG00000025428', 'E...   \n",
       "\n",
       "                   Ensembl_homologs  \n",
       "0  [{'homolog': 'ENSG00000089157'}]  \n",
       "1  [{'homolog': 'ENSG00000065135'}]  \n",
       "2                              None  \n",
       "3  [{'homolog': 'ENSG00000152234'}]  \n",
       "4                              None  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# combined_df = generator.load_dataframe_from_pickle(\"combined_df.pkl\")\n",
    "# opentargets_disease_compound_df = generator.load_dataframe_from_pickle(\n",
    "#     \"opentargets_disease_compound_df.pkl\"\n",
    "# )\n",
    "\n",
    "combined_df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>identifier.source</th>\n",
       "      <th>target</th>\n",
       "      <th>target.source</th>\n",
       "      <th>OpenTargets_disease_compounds</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>UMLS_C0001125</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_1000036</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL306823', 'drugban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UMLS_C0002622</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_0001072</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL526', 'drugbank_i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UMLS_C0002622</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>EFO_1001454</td>\n",
       "      <td>EFO</td>\n",
       "      <td>[{'chembl_id': 'CHEMBL:CHEMBL655', 'drugbank_i...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      identifier identifier.source       target target.source  \\\n",
       "0  UMLS_C0001125              UMLS  EFO_1000036           EFO   \n",
       "1  UMLS_C0002622              UMLS  EFO_0001072           EFO   \n",
       "2  UMLS_C0002622              UMLS  EFO_1001454           EFO   \n",
       "\n",
       "                       OpenTargets_disease_compounds  \n",
       "0  [{'chembl_id': 'CHEMBL:CHEMBL306823', 'drugban...  \n",
       "1  [{'chembl_id': 'CHEMBL:CHEMBL526', 'drugbank_i...  \n",
       "2  [{'chembl_id': 'CHEMBL:CHEMBL655', 'drugbank_i...  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "opentargets_disease_compound_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Building graph: 100%|██████████| 152/152 [00:00<00:00, 2745.70it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000100426\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000032087\n",
      "[{'stringdb_link_to': 'ENSMUSG00000027452', 'Ensembl': 'Ensembl:ENSMUSP00000028944', 'score': 0.636, 'Uniprot-TrEMBL': 'Aldh1a7'}, {'stringdb_link_to': 'ENSMUSG00000028715', 'Ensembl': 'Ensembl:ENSMUSP00000030487', 'score': 0.738, 'Uniprot-TrEMBL': 'Aldh1a7'}]\n",
      "Adding PPI for: ENSMUSG00000024747\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000091813\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000004035\n",
      "[{'stringdb_link_to': 'ENSMUSG00000028356', 'Ensembl': 'Ensembl:ENSMUSP00000030041', 'score': 0.47, 'Uniprot-TrEMBL': 'Ftcd'}, {'stringdb_link_to': 'ENSMUSG00000022877', 'Ensembl': 'Ensembl:ENSMUSP00000023590', 'score': 0.567, 'Uniprot-TrEMBL': 'Ftcd'}]\n",
      "Adding PPI for: ENSMUSG00000001155\n",
      "[{'stringdb_link_to': 'ENSMUSG00000026205', 'Ensembl': 'Ensembl:ENSMUSP00000027405', 'score': 0.549, 'Uniprot-TrEMBL': 'Slc17a4'}]\n",
      "Adding PPI for: ENSMUSG00000021336\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.752, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.753, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.617, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.868, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.721, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.834, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.898, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.683, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.634, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'ENSMUSP00000070575', 'score': 0.593, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'ENSMUSP00000081319', 'score': 0.834, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'ENSMUSP00000081889', 'score': 0.425, 'Uniprot-TrEMBL': 'Knstrn'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.697, 'Uniprot-TrEMBL': 'Knstrn'}]\n",
      "Adding PPI for: ENSMUSG00000027331\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.987, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.98, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.604, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.973, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.945, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.976, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.931, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.591, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.926, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.901, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'ENSMUSP00000070575', 'score': 0.78, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000003038', 'Ensembl': 'Ensembl:ENSMUSP00000099613', 'score': 0.441, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.517, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.616, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.813, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.834, 'Uniprot-TrEMBL': 'Cdca8'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.909, 'Uniprot-TrEMBL': 'Cdca8'}]\n",
      "Adding PPI for: ENSMUSG00000028873\n",
      "[{'stringdb_link_to': 'ENSMUSG00000022415', 'Ensembl': 'ENSMUSP00000009727', 'score': 0.498, 'Uniprot-TrEMBL': 'Sv2a'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'ENSMUSP00000031314', 'score': 0.411, 'Uniprot-TrEMBL': 'Sv2a'}]\n",
      "Adding PPI for: ENSMUSG00000038486\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000044816\n",
      "[{'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.467, 'Uniprot-TrEMBL': 'Hmgn2'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'ENSMUSP00000081319', 'score': 0.441, 'Uniprot-TrEMBL': 'Hmgn2'}]\n",
      "Adding PPI for: ENSMUSG00000003038\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000015451\n",
      "[{'stringdb_link_to': 'ENSMUSG00000022877', 'Ensembl': 'ENSMUSP00000023590', 'score': 0.543, 'Uniprot-TrEMBL': 'Spp2'}, {'stringdb_link_to': 'ENSMUSG00000028356', 'Ensembl': 'Ensembl:ENSMUSP00000030041', 'score': 0.513, 'Uniprot-TrEMBL': 'Spp2'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'Ensembl:ENSMUSP00000031314', 'score': 0.585, 'Uniprot-TrEMBL': 'Spp2'}]\n",
      "Adding PPI for: ENSMUSG00000026295\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000056749\n",
      "[{'stringdb_link_to': 'ENSMUSG00000055629', 'Ensembl': 'ENSMUSP00000039758', 'score': 0.516, 'Uniprot-TrEMBL': 'Cmah'}]\n",
      "Adding PPI for: ENSMUSG00000016756\n",
      "[{'stringdb_link_to': 'ENSMUSG00000024747', 'Ensembl': 'ENSMUSP00000025656', 'score': 0.636, 'Uniprot-TrEMBL': 'Acss1'}]\n",
      "Adding PPI for: ENSMUSG00000027452\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000027577\n",
      "[{'stringdb_link_to': 'ENSMUSG00000028970', 'Ensembl': 'ENSMUSP00000009058', 'score': 0.482, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.516, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000022877', 'Ensembl': 'ENSMUSP00000023590', 'score': 0.732, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000026295', 'Ensembl': 'ENSMUSP00000027518', 'score': 0.585, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000028356', 'Ensembl': 'ENSMUSP00000030041', 'score': 0.919, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.411, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000038486', 'Ensembl': 'Ensembl:ENSMUSP00000037576', 'score': 0.411, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.424, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000034855', 'Ensembl': 'Ensembl:ENSMUSP00000047646', 'score': 0.606, 'Uniprot-TrEMBL': 'Alb'}, {'stringdb_link_to': 'ENSMUSG00000032081', 'Ensembl': 'Ensembl:ENSMUSP00000113126', 'score': 0.775, 'Uniprot-TrEMBL': 'Alb'}]\n",
      "Adding PPI for: ENSMUSG00000029368\n",
      "[{'stringdb_link_to': 'ENSMUSG00000021336', 'Ensembl': 'ENSMUSP00000021769', 'score': 0.549, 'Uniprot-TrEMBL': 'Slc23a3'}]\n",
      "Adding PPI for: ENSMUSG00000026205\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.999, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.992, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.731, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.854, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'Ensembl:ENSMUSP00000061923', 'score': 0.506, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.525, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'Ensembl:ENSMUSP00000033326', 'score': 0.545, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.721, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.756, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.801, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.818, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.876, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.929, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.945, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.959, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'Ensembl:ENSMUSP00000034742', 'score': 0.972, 'Uniprot-TrEMBL': 'Bub1'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'Ensembl:ENSMUSP00000057938', 'score': 0.985, 'Uniprot-TrEMBL': 'Bub1'}]\n",
      "Adding PPI for: ENSMUSG00000027379\n",
      "[{'stringdb_link_to': 'ENSMUSG00000024747', 'Ensembl': 'ENSMUSP00000025656', 'score': 0.738, 'Uniprot-TrEMBL': 'Cyp4a14'}, {'stringdb_link_to': 'ENSMUSG00000037440', 'Ensembl': 'Ensembl:ENSMUSP00000040599', 'score': 0.557, 'Uniprot-TrEMBL': 'Cyp4a14'}, {'stringdb_link_to': 'ENSMUSG00000072949', 'Ensembl': 'Ensembl:ENSMUSP00000126448', 'score': 0.657, 'Uniprot-TrEMBL': 'Cyp4a14'}]\n",
      "Adding PPI for: ENSMUSG00000028715\n",
      "[{'stringdb_link_to': 'ENSMUSG00000026415', 'Ensembl': 'Ensembl:ENSMUSP00000108096', 'score': 0.471, 'Uniprot-TrEMBL': 'Ubd'}, {'stringdb_link_to': 'ENSMUSG00000034855', 'Ensembl': 'Ensembl:ENSMUSP00000047646', 'score': 0.54, 'Uniprot-TrEMBL': 'Ubd'}]\n",
      "Adding PPI for: ENSMUSG00000035186\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000043439\n",
      "[{'stringdb_link_to': 'ENSMUSG00000016756', 'Ensembl': 'Ensembl:ENSMUSP00000106021', 'score': 0.516, 'Uniprot-TrEMBL': 'B4galnt4'}]\n",
      "Adding PPI for: ENSMUSG00000055629\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000079494\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.978, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.947, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.803, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.693, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.985, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.91, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'Ensembl:ENSMUSP00000061923', 'score': 0.402, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.552, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.707, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.748, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.838, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.898, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.922, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.931, 'Uniprot-TrEMBL': 'Cenpe'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.935, 'Uniprot-TrEMBL': 'Cenpe'}]\n",
      "Adding PPI for: ENSMUSG00000045328\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000075044\n",
      "[{'stringdb_link_to': 'ENSMUSG00000000934', 'Ensembl': 'ENSMUSP00000000958', 'score': 0.929, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.935, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.99, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.693, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.866, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.959, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'ENSMUSP00000033326', 'score': 0.428, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.868, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.922, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.64, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.992, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000003038', 'Ensembl': 'Ensembl:ENSMUSP00000099613', 'score': 0.467, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.577, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.634, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.652, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.8, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.817, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.863, 'Uniprot-TrEMBL': 'Top2a'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.901, 'Uniprot-TrEMBL': 'Top2a'}]\n",
      "Adding PPI for: ENSMUSG00000020914\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000001334\n",
      "[{'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'Ensembl:ENSMUSP00000020579', 'score': 0.647, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.681, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.685, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.713, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'Ensembl:ENSMUSP00000061923', 'score': 0.714, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'Ensembl:ENSMUSP00000033326', 'score': 0.716, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.752, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.797, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.858, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.911, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.935, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'Ensembl:ENSMUSP00000024270', 'score': 0.978, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'Ensembl:ENSMUSP00000057938', 'score': 0.978, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.987, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'Ensembl:ENSMUSP00000020099', 'score': 0.995, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'Ensembl:ENSMUSP00000034742', 'score': 0.999, 'Uniprot-TrEMBL': 'Cdc20'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'Ensembl:ENSMUSP00000028858', 'score': 0.999, 'Uniprot-TrEMBL': 'Cdc20'}]\n",
      "Adding PPI for: ENSMUSG00000006398\n",
      "[{'stringdb_link_to': 'ENSMUSG00000038486', 'Ensembl': 'Ensembl:ENSMUSP00000037576', 'score': 0.498, 'Uniprot-TrEMBL': 'Syngr1'}]\n",
      "Adding PPI for: ENSMUSG00000022415\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000021213\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.714, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.699, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.528, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.506, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.696, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.402, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.413, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.448, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.502, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.548, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.591, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.64, 'Uniprot-TrEMBL': 'Mcm2'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.81, 'Uniprot-TrEMBL': 'Mcm2'}]\n",
      "Adding PPI for: ENSMUSG00000002870\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000061959\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000001313\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000007950\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000017390\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000025983\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000026049\n",
      "[{'stringdb_link_to': 'ENSMUSG00000035186', 'Ensembl': 'ENSMUSP00000035808', 'score': 0.471, 'Uniprot-TrEMBL': 'Fcamr'}]\n",
      "Adding PPI for: ENSMUSG00000026415\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000030825\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.858, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.909, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.617, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.847, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.818, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'ENSMUSP00000033326', 'score': 0.649, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.848, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.748, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.785, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.8, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.412, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.504, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.573, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.593, 'Uniprot-TrEMBL': 'Cdkn3'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.78, 'Uniprot-TrEMBL': 'Cdkn3'}]\n",
      "Adding PPI for: ENSMUSG00000037628\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000048489\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000051483\n",
      "[{'stringdb_link_to': 'ENSMUSG00000022877', 'Ensembl': 'ENSMUSP00000023590', 'score': 0.401, 'Uniprot-TrEMBL': 'Cpn1'}]\n",
      "Adding PPI for: ENSMUSG00000025196\n",
      "[{'stringdb_link_to': 'ENSMUSG00000000934', 'Ensembl': 'ENSMUSP00000000958', 'score': 0.604, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.681, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.782, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.407, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000024521', 'Ensembl': 'ENSMUSP00000025399', 'score': 0.49, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.876, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'ENSMUSP00000031314', 'score': 0.424, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'ENSMUSP00000033326', 'score': 0.969, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.596, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.552, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.81, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.527, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.817, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'ENSMUSP00000070575', 'score': 0.412, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000054612', 'Ensembl': 'ENSMUSP00000080224', 'score': 0.531, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'ENSMUSP00000081319', 'score': 0.517, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'ENSMUSP00000081889', 'score': 0.449, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'ENSMUSP00000099434', 'score': 0.696, 'Uniprot-TrEMBL': 'Chek1'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'ENSMUSP00000126417', 'score': 0.472, 'Uniprot-TrEMBL': 'Chek1'}]\n",
      "Adding PPI for: ENSMUSG00000032113\n",
      "[{'stringdb_link_to': 'ENSMUSG00000054612', 'Ensembl': 'Ensembl:ENSMUSP00000080224', 'score': 0.515, 'Uniprot-TrEMBL': 'Dusp8'}]\n",
      "Adding PPI for: ENSMUSG00000037887\n",
      "[{'stringdb_link_to': 'ENSMUSG00000037887', 'Ensembl': 'ENSMUSP00000049414', 'score': 0.515, 'Uniprot-TrEMBL': 'Mgmt'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.531, 'Uniprot-TrEMBL': 'Mgmt'}]\n",
      "Adding PPI for: ENSMUSG00000054612\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000080115\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000004356\n",
      "[{'stringdb_link_to': 'ENSMUSG00000001155', 'Ensembl': 'ENSMUSP00000001183', 'score': 0.567, 'Uniprot-TrEMBL': 'Hrg'}, {'stringdb_link_to': 'ENSMUSG00000025196', 'Ensembl': 'Ensembl:ENSMUSP00000026210', 'score': 0.401, 'Uniprot-TrEMBL': 'Hrg'}, {'stringdb_link_to': 'ENSMUSG00000028356', 'Ensembl': 'Ensembl:ENSMUSP00000030041', 'score': 0.475, 'Uniprot-TrEMBL': 'Hrg'}, {'stringdb_link_to': 'ENSMUSG00000026295', 'Ensembl': 'Ensembl:ENSMUSP00000027518', 'score': 0.543, 'Uniprot-TrEMBL': 'Hrg'}, {'stringdb_link_to': 'ENSMUSG00000032081', 'Ensembl': 'Ensembl:ENSMUSP00000113126', 'score': 0.678, 'Uniprot-TrEMBL': 'Hrg'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'Ensembl:ENSMUSP00000031314', 'score': 0.732, 'Uniprot-TrEMBL': 'Hrg'}]\n",
      "Adding PPI for: ENSMUSG00000022877\n",
      "[{'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'ENSMUSP00000031314', 'score': 0.606, 'Uniprot-TrEMBL': 'Cxcl10'}, {'stringdb_link_to': 'ENSMUSG00000035186', 'Ensembl': 'ENSMUSP00000035808', 'score': 0.54, 'Uniprot-TrEMBL': 'Cxcl10'}, {'stringdb_link_to': 'ENSMUSG00000026981', 'Ensembl': 'Ensembl:ENSMUSP00000110126', 'score': 0.666, 'Uniprot-TrEMBL': 'Cxcl10'}]\n",
      "Adding PPI for: ENSMUSG00000034855\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.647, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.639, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.485, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'Ensembl:ENSMUSP00000024270', 'score': 0.528, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.604, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.617, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.617, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.693, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'Ensembl:ENSMUSP00000028858', 'score': 0.731, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.735, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'Ensembl:ENSMUSP00000034742', 'score': 0.754, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.764, 'Uniprot-TrEMBL': 'Hmmr'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'Ensembl:ENSMUSP00000057938', 'score': 0.803, 'Uniprot-TrEMBL': 'Hmmr'}]\n",
      "Adding PPI for: ENSMUSG00000020330\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.685, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.81, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.467, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000024521', 'Ensembl': 'ENSMUSP00000025399', 'score': 0.49, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.525, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'ENSMUSP00000031314', 'score': 0.411, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'ENSMUSP00000033326', 'score': 0.571, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.781, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.448, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.417, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.652, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'ENSMUSP00000081319', 'score': 0.616, 'Uniprot-TrEMBL': 'E2f1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.696, 'Uniprot-TrEMBL': 'E2f1'}]\n",
      "Adding PPI for: ENSMUSG00000027490\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000020083\n",
      "[{'stringdb_link_to': 'ENSMUSG00000034855', 'Ensembl': 'ENSMUSP00000047646', 'score': 0.666, 'Uniprot-TrEMBL': 'Il1rn'}]\n",
      "Adding PPI for: ENSMUSG00000026981\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000022026\n",
      "[{'stringdb_link_to': 'ENSMUSG00000028715', 'Ensembl': 'ENSMUSP00000030487', 'score': 0.557, 'Uniprot-TrEMBL': 'Vnn1'}]\n",
      "Adding PPI for: ENSMUSG00000037440\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000071547\n",
      "[{'stringdb_link_to': 'ENSMUSG00000072949', 'Ensembl': 'ENSMUSP00000126448', 'score': 0.962, 'Uniprot-TrEMBL': 'Acot7'}]\n",
      "Adding PPI for: ENSMUSG00000028937\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000051262\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000046840\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000048572\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000097927\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.797, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.872, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.735, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.594, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.801, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.793, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.838, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.548, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.815, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.863, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'ENSMUSP00000070575', 'score': 0.504, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'ENSMUSP00000081319', 'score': 0.909, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'ENSMUSP00000081889', 'score': 0.606, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'ENSMUSP00000115860', 'score': 0.697, 'Uniprot-TrEMBL': 'Racgap1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.472, 'Uniprot-TrEMBL': 'Racgap1'}]\n",
      "Adding PPI for: ENSMUSG00000023015\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000041939\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.995, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000024521', 'Ensembl': 'Ensembl:ENSMUSP00000025399', 'score': 0.451, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'Ensembl:ENSMUSP00000031314', 'score': 0.516, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'Ensembl:ENSMUSP00000020579', 'score': 0.639, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'Ensembl:ENSMUSP00000061923', 'score': 0.699, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.753, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.782, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.791, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.81, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.872, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.901, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.909, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'Ensembl:ENSMUSP00000057938', 'score': 0.947, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'Ensembl:ENSMUSP00000024270', 'score': 0.953, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'Ensembl:ENSMUSP00000033326', 'score': 0.979, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.98, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.99, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'Ensembl:ENSMUSP00000028858', 'score': 0.992, 'Uniprot-TrEMBL': 'Cdk1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'Ensembl:ENSMUSP00000034742', 'score': 0.999, 'Uniprot-TrEMBL': 'Cdk1'}]\n",
      "Adding PPI for: ENSMUSG00000019942\n",
      "[{'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.604, 'Uniprot-TrEMBL': 'Top1mt'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.929, 'Uniprot-TrEMBL': 'Top1mt'}]\n",
      "Adding PPI for: ENSMUSG00000000934\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.911, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.901, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.764, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.889, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.929, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.881, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.935, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.413, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.417, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.527, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.683, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.703, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.785, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.815, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.926, 'Uniprot-TrEMBL': 'Nusap1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.992, 'Uniprot-TrEMBL': 'Nusap1'}]\n",
      "Adding PPI for: ENSMUSG00000027306\n",
      "[{'stringdb_link_to': 'ENSMUSG00000001155', 'Ensembl': 'ENSMUSP00000001183', 'score': 0.47, 'Uniprot-TrEMBL': 'Ambp'}, {'stringdb_link_to': 'ENSMUSG00000022877', 'Ensembl': 'ENSMUSP00000023590', 'score': 0.475, 'Uniprot-TrEMBL': 'Ambp'}, {'stringdb_link_to': 'ENSMUSG00000026295', 'Ensembl': 'ENSMUSP00000027518', 'score': 0.513, 'Uniprot-TrEMBL': 'Ambp'}, {'stringdb_link_to': 'ENSMUSG00000032081', 'Ensembl': 'Ensembl:ENSMUSP00000113126', 'score': 0.783, 'Uniprot-TrEMBL': 'Ambp'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'Ensembl:ENSMUSP00000031314', 'score': 0.919, 'Uniprot-TrEMBL': 'Ambp'}]\n",
      "Adding PPI for: ENSMUSG00000028356\n",
      "[{'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'Ensembl:ENSMUSP00000031314', 'score': 0.482, 'Uniprot-TrEMBL': 'Abcb1b'}]\n",
      "Adding PPI for: ENSMUSG00000028970\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.999, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.999, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.754, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.968, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.972, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'ENSMUSP00000033326', 'score': 0.741, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.596, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.642, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'Ensembl:ENSMUSP00000061923', 'score': 0.696, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.781, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.793, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.834, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.848, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.868, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.881, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'Ensembl:ENSMUSP00000057938', 'score': 0.91, 'Uniprot-TrEMBL': 'Ccnb2'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.976, 'Uniprot-TrEMBL': 'Ccnb2'}]\n",
      "Adding PPI for: ENSMUSG00000032218\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.713, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.791, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.485, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.41, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.756, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'ENSMUSP00000034742', 'score': 0.642, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'ENSMUSP00000057938', 'score': 0.707, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'ENSMUSP00000061923', 'score': 0.502, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'ENSMUSP00000068713', 'score': 0.703, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'ENSMUSP00000068896', 'score': 0.577, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'ENSMUSP00000070575', 'score': 0.573, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'ENSMUSP00000081319', 'score': 0.813, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.425, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.449, 'Uniprot-TrEMBL': 'Ncapg2'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.606, 'Uniprot-TrEMBL': 'Ncapg2'}]\n",
      "Adding PPI for: ENSMUSG00000042029\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000070661\n",
      "[{'stringdb_link_to': 'ENSMUSG00000028715', 'Ensembl': 'ENSMUSP00000030487', 'score': 0.657, 'Uniprot-TrEMBL': 'Acot1'}, {'stringdb_link_to': 'ENSMUSG00000028937', 'Ensembl': 'Ensembl:ENSMUSP00000129121', 'score': 0.962, 'Uniprot-TrEMBL': 'Acot1'}]\n",
      "Adding PPI for: ENSMUSG00000072949\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.978, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.953, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000020330', 'Ensembl': 'ENSMUSP00000020579', 'score': 0.528, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.407, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000042029', 'Ensembl': 'Ensembl:ENSMUSP00000081889', 'score': 0.41, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.467, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000002870', 'Ensembl': 'Ensembl:ENSMUSP00000061923', 'score': 0.528, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000031016', 'Ensembl': 'Ensembl:ENSMUSP00000033326', 'score': 0.589, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000023015', 'Ensembl': 'Ensembl:ENSMUSP00000126417', 'score': 0.594, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000045328', 'Ensembl': 'Ensembl:ENSMUSP00000057938', 'score': 0.693, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.847, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'Ensembl:ENSMUSP00000028858', 'score': 0.854, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.866, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000027331', 'Ensembl': 'Ensembl:ENSMUSP00000115860', 'score': 0.868, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000027306', 'Ensembl': 'Ensembl:ENSMUSP00000068713', 'score': 0.889, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'Ensembl:ENSMUSP00000034742', 'score': 0.968, 'Uniprot-TrEMBL': 'Cdca3'}, {'stringdb_link_to': 'ENSMUSG00000028873', 'Ensembl': 'Ensembl:ENSMUSP00000081319', 'score': 0.973, 'Uniprot-TrEMBL': 'Cdca3'}]\n",
      "Adding PPI for: ENSMUSG00000023505\n",
      "[{'stringdb_link_to': 'ENSMUSG00000022877', 'Ensembl': 'ENSMUSP00000023590', 'score': 0.678, 'Uniprot-TrEMBL': 'Apoc3'}, {'stringdb_link_to': 'ENSMUSG00000028356', 'Ensembl': 'ENSMUSP00000030041', 'score': 0.783, 'Uniprot-TrEMBL': 'Apoc3'}, {'stringdb_link_to': 'ENSMUSG00000029368', 'Ensembl': 'ENSMUSP00000031314', 'score': 0.775, 'Uniprot-TrEMBL': 'Apoc3'}]\n",
      "Adding PPI for: ENSMUSG00000032081\n",
      "[{'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.451, 'Uniprot-TrEMBL': 'Pmaip1'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.49, 'Uniprot-TrEMBL': 'Pmaip1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.49, 'Uniprot-TrEMBL': 'Pmaip1'}]\n",
      "Adding PPI for: ENSMUSG00000024521\n",
      "[{'stringdb_link_to': nan, 'Ensembl': nan, 'score': nan}]\n",
      "Skipping row with no valid PPI for: ENSMUSG00000042041\n",
      "[{'stringdb_link_to': 'ENSMUSG00000006398', 'Ensembl': 'ENSMUSP00000006565', 'score': 0.716, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000019942', 'Ensembl': 'ENSMUSP00000020099', 'score': 0.979, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000023505', 'Ensembl': 'ENSMUSP00000024270', 'score': 0.589, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000027379', 'Ensembl': 'ENSMUSP00000028858', 'score': 0.545, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000020914', 'Ensembl': 'Ensembl:ENSMUSP00000068896', 'score': 0.428, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000027490', 'Ensembl': 'Ensembl:ENSMUSP00000099434', 'score': 0.571, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000037628', 'Ensembl': 'Ensembl:ENSMUSP00000070575', 'score': 0.649, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000032218', 'Ensembl': 'Ensembl:ENSMUSP00000034742', 'score': 0.741, 'Uniprot-TrEMBL': 'Wee1'}, {'stringdb_link_to': 'ENSMUSG00000032113', 'Ensembl': 'Ensembl:ENSMUSP00000134388', 'score': 0.969, 'Uniprot-TrEMBL': 'Wee1'}]\n",
      "Adding PPI for: ENSMUSG00000031016\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000177103\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000165092\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000213366\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000160282\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000146039\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000128944\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000134690\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000159164\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000188674\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000224389\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000072080\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000165030\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000154930\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000101204\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000163631\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000213901\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000169679\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000213886\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000273604\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000182272\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000144035\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000144035\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000138778\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000149742\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000131747\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000160097\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000117399\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000100321\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000073111\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000108830\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000127220\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000109107\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000144395\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000151287\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000162897\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000087076\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000100526\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000165507\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000159228\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000120054\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000149554\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000184545\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000170430\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000123427\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000120800\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000113905\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000169245\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000072571\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000101412\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000136689\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000102837\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000112299\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000168268\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000097021\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000181778\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000161800\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000110921\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000170312\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000184428\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000137804\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000106927\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000085563\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000157456\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000146918\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000178828\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000184227\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000111665\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000110245\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000141682\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000173237\n",
      "nan\n",
      "Skipping row with empty or invalid PPI list for: ENSG00000166483\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "pygraph = generator.build_networkx_graph(merged_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 Store the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"networkx_graph_test.pkl\", \"wb\") as out:\n",
    "    pickle.dump(pygraph, out)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 Visualize the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pos = nx.circular_layout(pygraph)\n",
    "\n",
    "# plt.figure(3, figsize=(30, 30))\n",
    "# nx.draw(pygraph, pos)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Exporting Graph to external sources"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.1 Cytoscape\n",
    "Make sure that the Cytoscape is open"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying default style...\n",
      "Applying preferred layout\n"
     ]
    }
   ],
   "source": [
    "from pyBiodatafuse.graph import cytoscape\n",
    "\n",
    "cytoscape.load_graph(pygraph, network_name=\"Test network\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4.2 Neo4j"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyBiodatafuse.graph import neo4j\n",
    "\n",
    "neo4j.save_graph_to_graphml(pygraph, \"networkx_graph_test.graphml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Steps to load the graph in Neo4j"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `.graphml` file in **import** subfolder of the DBMS folder\n",
    "- Install apoc plugin\n",
    "- Create `apoc.conf` file:\n",
    "    ```\n",
    "    apoc.trigger.enabled=true\n",
    "    apoc.import.file.enabled=true\n",
    "    apoc.export.file.enabled=true\n",
    "    apoc.import.file.use_neo4j_config=true\n",
    "    ```\n",
    "- Add `apoc.conf` file to **conf** subfolder of the DBMS folder\n",
    "- Open Neo4j Browser\n",
    "- (Optionl, only run if you have imported a graph  before) Remove all the nodes before importing `.graphml` file\n",
    "\n",
    "    ```MATCH (n) DETACH DELETE n```\n",
    "\n",
    "- Import `.graphml` file\n",
    "\n",
    "    ```call apoc.import.graphml('file:///networkx_graph_test.graphml',{readLabels:TRUE})```\n",
    "\n",
    "- Add indexes after importing the graph for improving the performance of queries\n",
    "\n",
    "    ```\n",
    "    create index Gene for (n:Gene) on (n.node_type)\n",
    "    create index Pathway for (n:Pathway) on (n.node_type)\n",
    "    create index `Biological Process` for (n:`Biological Process`) on (n.node_type)\n",
    "    create index `Molecular Function` for (n:`Molecular Function`) on (n.node_type)\n",
    "    create index `Cellular Component` for (n:`Cellular Component`) on (n.node_type)\n",
    "    create index Disease for (n:Disease) on (n.node_type)\n",
    "    create index Compound for (n:Compound) on (n.node_type)\n",
    "    create index `Side Effect` for (n:`Side Effect`) on (n.node_type)\n",
    "    ```\n",
    "    \n",
    "\n",
    "- Count the number of each node type\n",
    "    - total (```MATCH (n) RETURN count(n)```) \n",
    "        - Gene (```MATCH (n:Gene) RETURN count(n)```)\n",
    "        - Pathway (```MATCH (n:Pathway) RETURN count(n)```)\n",
    "            - WikiPathways (```MATCH (n:Pathway {source: \"WikiPathways\"}) RETURN count(n)```) \n",
    "            - OpenTargets, Reactome (```MATCH (n:Pathway {source: \"OpenTargets\"}) RETURN count(n)```) \n",
    "            - MINERVA (```MATCH (n:Pathway {source: \"MINERVA\"}) RETURN count(n)```) \n",
    "        - Biological Process (```MATCH (n:`Biological Process`) RETURN count(n)```) \n",
    "        - Molecular Function (```MATCH (n:`Molecular Function`) RETURN count(n)```) \n",
    "        - Cellular Component (```MATCH (n:`Cellular Component`) RETURN count(n)```) \n",
    "        - Disease (```MATCH (n:Disease) RETURN count(n)```) \n",
    "        - Compound (```MATCH (n:Compound) RETURN count(n)```)\n",
    "        - Side Effect (```MATCH (n:`Side Effect`) RETURN count(n)```) \n",
    "- Count the number of each edge type\n",
    "    - total (```MATCH ()-[r]->() RETURN count(r)```) \n",
    "        - interacts_with (```MATCH ()-[r:interacts_with]->() RETURN count(r)```) \n",
    "        - part_of (```MATCH ()-[r:part_of]->() RETURN count(r)```) \n",
    "            - WikiPathways (```MATCH ()-[r:part_of {source: \"WikiPathways\"}]->() RETURN count(r)```) \n",
    "            - OpenTargets, Reactome (```MATCH ()-[r:part_of {source: \"OpenTargets\"}]->() RETURN count(r)```) \n",
    "            - MINERVA (```MATCH ()-[r:part_of {source: \"MINERVA\"}]->() RETURN count(r)```) \n",
    "        - activates (```MATCH ()-[r:activates]->() RETURN count(r)```) \n",
    "        - treats (```MATCH ()-[r:treats]->() RETURN count(r)```) \n",
    "        - has_side_effect (```MATCH ()-[r:has_side_effect]->() RETURN count(r)```) \n",
    "        - inhibits (```MATCH ()-[r:inhibits]->() RETURN count(r)```) = 71\n",
    "        - associated_with (```MATCH ()-[r:associated_with]->() RETURN count(r)```) \n",
    "\n",
    "- Export the graph as a `.csv` file\n",
    "\n",
    "    ```call apoc.export.csv.all(\"networkx_graph_test.csv\",{})```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
